---
title: 
author: Tina, Pablo, Christa & Marc
date: "`r Sys.Date()`"
bibliography: /Users/admin/Desktop/nwork/Empirical/Manuscript_3/TPCM.bib
csl: /Users/admin/Desktop/nwork/simulation/styles-master/styles-master/molecular-biology-and-evolution.csl 
output:
  pdf_document: default
  html_notebook:
    fig_caption: yes
    keep_tex: yes
    highlight: espresso
    number_sections: no
  html_document:
    df_print: paged
  word_document: default
urlcolor: blue
fontsize: 10pt
geometry: margin=1in
editor_options:
  chunk_output_type: console
header-includes:
- \usepackage{caption}
- \usepackage{wasysym}
- \captionsetup[figure]{labelformat=empty}
- \usepackage{float}
- \usepackage{color}
---

<style>
body {
text-align: justify}
</style>


```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.pos = 'H')
```


```{r setup, include=FALSE}
  knitr::opts_chunk$set(echo = FALSE,message=FALSE, warning=FALSE, dev="png",dpi = 500,cache = TRUE)
```



```{r Initial settings, echo=F, message=F, warning=F}
library(ape)
library(geiger)
library(phytools)
library(treeio)
library(stringr)
library(ggrepel)
library(dplyr)
library(ggtree)
library (magrittr) 
library(parallel) ##Needed to run read.nhx from treeio 
library(tidyverse)
library(digest)
library(ggplot2)
library(caper)
library(cowplot)
library(easyGgplot2)
library(gridExtra)
library(R.utils)
library(hash)
library(rphylopic)
library(repr)
library(grid)

setwd("/Users/admin/Desktop/nwork/Empirical/Manuscript_3/")
folder1 <- c("~/Desktop/nwork/Empirical/Manuscript_3/processed_data_files/")
folder2 <- c("~/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/")
folder3<-c("~/Desktop/nwork/Empirical/Manuscript_3/kallisto_output/")
folder4<-c("~/Desktop/nwork/Empirical/Manuscript_3/2021-01-06_Selectome_Results/")

## Setting seed number for reproducibility 
set.seed(786)

## Codes information
#"BGEE_TPM_data_QNPIC_largescale.R" is used to extract and process TPM data of tissues for zebrafish.
#"tau.R" was used to process TPM data for rest of the four species and to calculate tau for our study

## To perform parallel computation we need to detect core numbers
cores <- detectCores()

```

 
# **Phylogenetic modeling provides evidence for sudden shifts in expression after small-scale duplication in vertebrates and strong support for the ortholog conjecture**

**Tina Begum**^1,2^, **Pablo Duchen**^3^, **Christabel Floi Bucao**^1,2^, **Marc Robinson-Rechavi**^1,2^* 

^1^Department of Ecology and Evolution, University of Lausanne, Lausanne, Switzerland <br />  
^2^SIB Swiss Institute of Bioinformatics, Lausanne, Switzerland <br />  
^3^Department of Computational Biology, University of Lausanne, Lausanne <br />  
^4^Institute for Organismal and Molecular Evolutionary Biology, Johannes Gutenberg University of Mainz, Germany <br />



\*Corresponding author email: marc.robinson-rechavi@unil.ch

## **Abstract**

Gene duplication is a potential source of innovation, but the evolutionary dynamics of functional change are still poorly understood. The "ortholog conjecture" posits that such innovation or other functional change is more common after duplication than in its absence, but is still debated. We have applied a Lévy model of evolutionary trait jumps to two phenotypes of gene function: average gene expression levels, and tissue specificity. Such instantaneous jumps in trait values typically occur during the transition of organisms into new adaptive zones, and thus provide insights into the role of duplication in adaptation. Using time-calibrated phylogenies of 15 vertebrates, including 5 teleost fishes, we show for the first time that such a mechanism of trait divergence strongly affects paralogs, in addition to other modes of functional evolution. We find that at least 25% of teleost fish small-scale duplicates follow rapid evolutionary rate shift model for both traits, much more than after speciations. We show that such trait jumps play a major role in the evolution of duplicates. While there is some evidence for more positive selection at the protein-coding level after duplication, this does not appear strongly linked to the presence of jumps in expression. This helps to explain higher phylogenetic independent contrasts between paralogs, supporting the ortholog conjecture. However, genome-wide duplicates (ohnologs) do not support such a trait jump model, and thus follow a different evolutionary dynamic, while also supporting the ortholog conjecture. 


## **Introduction**  

Gene duplication is a potential source of functional innovation, and plays an important role in the evolution of genomes [@Ohno1970; @Chen2012; @Guschanski2017]. Such innovation can lead to overall function gain or to specialization, and can be symmetric or not between paralogs [@Assis2013; @Assis2015; @Braasch2016; @Lien2016; @Guschanski2017; @Sandve2018]. Paralogs can also be retained without any functional change, due to dosage constraints [@Gout2015; Thompson et al. 2016]. In this context, there has been a debate about the long standing hypothesis that paralogs diverge generally more in function than orthologs, the "ortholog conjecture" [@Koonin2005; @Studer2009; @Nehrt2011; @Altenhoff2012; @Chen2012; @Gabaldon2013; @Rogozin2014; @Kryuchkova-Mostacci2016; @Dunn2018; @Fukushima2020; @Stamboulian2020; @Begum2021a].This ortholog conjecture is of interest both in practice, for functional annotation of homologs, and in principle, to understand how gene function evolves. Thus contrasting the evolution of genes after duplication or speciation provides a clear framework to investigate the evolution of function.

\vspace{0.5cm}
Since the retention rates and evolutionary dynamics of duplicates vary among lineages and between genes, they need to be studied using a phylogenetic framework with proper care [@Pagel1999; @Blomberg2003; @Dunn2018; @Begum2021a]. A phylogenetic approach also provides the opportunity to explicitly test different models of functional evolution. In most studies, the implicit model is that function evolves gradually, with changes in evolutionary rate restricted to specific branches of the phylogeny. These approaches, while important, remain less explored than simpler pairwise comparisons. Contrasting results from other vertebrates with teleosts thus can provide insights into the difference in evolutionary innovations of types of retained duplicates in teleosts with respect to non-teleost vertebrates.  

\vspace{0.5cm}
There is abundant evidence for an asymmetric steady acceleration of sequence-level evolutionary rates (e.g. dN/dS) after duplication [@Conant2003; @Gu2005; @Brunet2006; @Kim2006; @Cusack2007; @Scannell2008; @Studer2009; @Panchin2010; @Pegueroles2013; @Roselló2014; @Braasch2016; @Lien2016; @Holland2017; @Sandve2018]. However, functional evolution following gene duplication might not be limited to such gradual mode of evolution; it might not even be the main mode of divergence. Studies of phenotype evolution have found that there can be a period of instantaneous rapid bursts or jumps in trait values at the bases of clades that transitioned into new adaptive zones, due to the appearance of an evolutionary novelty [@Simpson1944; @Bokma2008; @Landis2013; @Duchen2017; @Landis2017]. This pattern of evolutionary innovations is called "pulsed evolution", where the background trait evolutionary rates $\sigma^2$ remain unaltered following a jump with a strength $\lambda$ [@Duchen2017; @Gao2022]. It is an open question whether such pulsed evolution exists for gene and genome duplicates [@Begum2021b]. We propose here to examine this question in light of the ortholog conjecture.   

\vspace{0.5cm}
Since phenotypic evolution may arise through mutations that largely affect gene expression patterns [@Guschanski2017], we have performed large-scale phylogenetic comparative analyses considering two phenotypic traits which characterize gene expression: average expression levels, and tissue specificity ($\tau$). We focused on teleost fishes since they have undergone a whole genome duplication (the "FishWGD") around 320 million years ago (Mya), followed by a spectacular evolutionary radiation leading to more than half of all vertebrate species [@Pasquier2017; @Davesne2021; @Parey2022a]. This allowed us to address the following questions using a phylogenetic framework: Do we find a global support for the ortholog conjecture for different types of duplicates and different expression traits? What is the contribution of instantaneous trait jump mechanisms to paralog evolution? We also explored whether changes in expression of paralogs are linked to positive selection at the coding-sequence level.   


```{r loading_data, echo=F, message=FALSE, warning=FALSE, cache=T}

 ## Loading preprocessed data "Paper3_preprocess_script.R"
 load("/Users/admin/Desktop/nwork/Empirical/Manuscript_3/norm_exp_15spe_4tips_new.RData") 
 ## Functions required for this study
 source("/Users/admin/Desktop/nwork/Empirical/Manuscript_3/functions_TPCM.R")

```

\pagebreak
## **Results**

### **An improved phylogenetic test supports the ortholog conjecture** 


```{r Fig_1, dpi=500, fig.width=8, fig.height=5, fig.cap= paste("Fig 1: The vertebrate species phylogeny of this study. The ‘brownish’ color block highlights 5 teleost fishes, and the ‘off-white’ color block highlights the 10 tetrapods.")}

 ## Loading previously generated plot
 img <- png::readPNG("~/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/Figures/Fig1_rphylo_latest3_300dpi.png")
 grid::grid.raster(img)

```


```{r Fig_2, dpi=400,fig.width=8, fig.height=8, fig.cap= paste0("**Fig 2**: Phylogenetic independent contrast ortholog conjecture (OC) tests. PICs: Phylogenetic independent contrasts, exp: expression. Plots are based on ", length(Calibrated.all.standardized.tau), " and ", length(Calibrated.all.standardized.exp), " trees for $\\tau$ and for average gene expression levels, respecively. For (A)-(B), we considered the PICs of teleosts specific speciation and duplication nodes for our analyses, while PICs of all the speciation and duplication nodes are used in (C)-(D). *P* values are from Wilcoxon two-tailed tests.")}


##Collecting silhouettes with rphylopic
#fish <- name_search(text = "teleost", options = "namebankID")[[1]]
#fish_id <- name_images(uuid = fish$uid[1])$same[[2]]$uid
#fish_pic <- image_data(fish_id, size = 256)[[1]]
fish <- get_uuid(name = "teleostei", n = 500) # list images 
fish_pic <- get_phylopic(uuid =  "0b4e39a1-3bca-4b58-8805-ebaae22cd860")  # get individual image id
fish_id <- fish[2]

#tetrapods <- name_search(text = "tetrapoda", options = "namebankID")[[1]]
#tetrapod_id <- name_images(uuid = tetrapods$uid[2])$same[[2]]$uid
#tetrapods_pic <- image_data(tetrapod_id, size = 256)[[1]] 
tetrapods <- get_uuid(name ="tetrapoda", n = 500) # list images 
tetrapods_pic <- get_phylopic(uuid =  "24c49c6a-0052-4004-b571-b0ba2cb1d073") # get individual image id
tetrapod_id <- tetrapods[15]


#vertebrates <- name_search(text = "vertebrates", options = "namebankID")[[1]]
#vertebrate_id <- name_images(uuid = vertebrates$uid[1])$same[[1]]$uid
#vertebrate_pic <- image_data(vertebrate_id, size = 256)[[1]] 

##Identifying teleost specific clades
nodes.contrast.tau.teleosts<-nodes.contrast.tau.all[which(nodes.contrast.tau.all$label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 

nodes.contrast.exp.teleosts<-nodes.contrast.exp.all[which(nodes.contrast.exp.all$label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,]

##Testing on tau for teleosts
speciation.picTau.teleosts <-speciation.contrast(nodes.contrast.tau.teleosts,"Tau.abs")
duplication.picTau.teleosts <-duplication.contrast(nodes.contrast.tau.teleosts,"Tau.abs")
Pval.Tau.teleosts <- two.tailed.wilcox(duplication.picTau.teleosts,speciation.picTau.teleosts)

## Median data of tau for teleosts
levels(nodes.contrast.tau.teleosts$Event)[levels(nodes.contrast.tau.teleosts$Event)=="speciation"] <- "Speciation"
levels(nodes.contrast.tau.teleosts$Event)[levels(nodes.contrast.tau.teleosts$Event)=="duplication"] <- "Duplication"
median.Tau.teleosts<-NULL
median.Tau.teleosts<-nodes.contrast.tau.teleosts %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.Tau.teleosts$pic.round<-paste0("Median = ",round(median.Tau.teleosts$Tau.abs,4), sep="")
#levels(median.Tau.teleosts$Event)<-c("Speciation", "Duplication")
median.Tau.teleosts$count<- paste0(median.Tau.teleosts$Event,"\n","(n = ",median.Tau.teleosts$freq,")")

##Testing on tau for vertebrates
speciation.picTau <-speciation.contrast(nodes.contrast.tau.all,"Tau.abs")
duplication.picTau <-duplication.contrast(nodes.contrast.tau.all,"Tau.abs")
Pval.Tau <- two.tailed.wilcox(duplication.picTau,speciation.picTau)

## Median data of tau for vertebrates
levels(nodes.contrast.tau.all$Event)[levels(nodes.contrast.tau.all$Event)=="speciation"] <- "Speciation"
levels(nodes.contrast.tau.all$Event)[levels(nodes.contrast.tau.all$Event)=="duplication"] <- "Duplication"
median.Tau<-NULL
median.Tau<-nodes.contrast.tau.all %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.Tau$pic.round<-paste0("Median = ",round(median.Tau$Tau.abs,4), sep="")
#levels(median.Tau$Event)<-c("Speciation", "Duplication")
median.Tau$count<- paste0(median.Tau$Event,"\n","(n = ",median.Tau$freq,")")

##Testing for expression levels of teleosts
speciation.picExp.teleosts <-speciation.contrast(nodes.contrast.exp.teleosts,"Exp.abs")
duplication.picExp.teleosts <-duplication.contrast(nodes.contrast.exp.teleosts,"Exp.abs")
Pval.Exp.teleosts <- two.tailed.wilcox(duplication.picExp.teleosts,speciation.picExp.teleosts)

## Median data for expression levels of teleosts
levels(nodes.contrast.exp.teleosts$Event)<-c("Speciation", "Duplication")
median.Exp.teleosts<-NULL
median.Exp.teleosts<-nodes.contrast.exp.teleosts %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.Exp.teleosts$pic.round<-paste0("Median = ",round(median.Exp.teleosts$Exp.abs,4), sep="")
#levels(median.Exp.teleosts$Event)<-list(speciation= "Speciation", duplication= "Duplication")
median.Exp.teleosts$count<- paste0(median.Exp.teleosts$Event,"\n","(n = ",median.Exp.teleosts$freq,")")

##Testing for expression levels of vertebrates
speciation.picExp <-speciation.contrast(nodes.contrast.exp.all,"Exp.abs")
duplication.picExp <-duplication.contrast(nodes.contrast.exp.all,"Exp.abs")
Pval.Exp <- two.tailed.wilcox(duplication.picExp,speciation.picExp)

## Median data of expression levels for vertebrates
levels(nodes.contrast.exp.all$Event)<-c("Speciation", "Duplication")
median.Exp<-NULL
median.Exp<-nodes.contrast.exp.all %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.Exp$pic.round<-paste0("Median = ",round(median.Exp$Exp.abs,4), sep="")
#levels(median.Exp$Event)<-c("Speciation", "Duplication")
median.Exp$count<- paste0(median.Exp$Event,"\n","(n = ",median.Exp$freq,")")

## Plotting 
dodge <- position_dodge(width = 0.05)

plot2A<- boxplot.new(nodes.contrast.tau.teleosts,Pval.Tau.teleosts,"Tau",median.Tau.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  #geom_text(data = median.Tau, aes(label = pic.round),fontface = 2,hjust=0.5,vjust =-1)+
  ggtitle(expression(bold(paste("A. The OC test on ", tau, " for teleosts"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.Tau.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.043, ysize = 0.003)
  #annotate("text", x=2, y=0.045, label = "italic(P < 2.2%*%10^{-16})", parse =T,size=4)

plot2B<- boxplot.new(nodes.contrast.exp.teleosts,Pval.Exp.teleosts,"Exp",median.Exp.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. The OC test on exp levels for teleosts"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.Exp.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.019)

plot2C<- boxplot.new(nodes.contrast.tau.all,Pval.Tau,"Tau",median.Tau) +
  coord_cartesian(ylim=c(0, 0.05)) +
  #geom_text(data = median.Tau, aes(label = pic.round),fontface = 2,hjust=0.5,vjust =-1)+
  ggtitle(expression(bold(paste("C. The OC test on ", tau, " for 15 species"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.Tau$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.043, ysize = 0.002)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.003)
 
plot2D<- boxplot.new(nodes.contrast.exp.all,Pval.Exp,"Exp",median.Exp) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. The OC test on exp levels for 15 species"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.Exp$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.013)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.019)
 
##Plotting results together
  gridExtra::grid.arrange(plot2A,plot2B,plot2C,plot2D, nrow=2,ncol=2) 

```


```{r Analyses_on_sure_3RWGD_trees, echo=F, message=FALSE, warning=FALSE, cache=T}

##Identifying teleost specific clades
nodes.contrast.tau.3RWGDs.teleosts<-nodes.contrast.tau.3RWGDs[which(nodes.contrast.tau.3RWGDs$label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
nodes.contrast.exp.3RWGDs.teleosts<-nodes.contrast.exp.3RWGDs[which(nodes.contrast.exp.3RWGDs$label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,]

##Testing on tau for 3RWGDs
speciation.picTau.3RWGDs <-speciation.contrast(nodes.contrast.tau.3RWGDs.teleosts ,"Tau.abs")
duplication.picTau.3RWGDs1 <-duplication.contrast.3R(nodes.contrast.tau.3RWGDs.teleosts,"Tau.abs")
duplication.picTau.3RWGDs2 <-duplication.contrast(nodes.contrast.tau.3RWGDs.teleosts,"Tau.abs")
Pval.Tau.3RWGDs.teleosts1 <- two.tailed.wilcox(duplication.picTau.3RWGDs1,speciation.picTau.3RWGDs)
Pval.Tau.3RWGDs.teleosts2 <- two.tailed.wilcox(duplication.picTau.3RWGDs2,speciation.picTau.3RWGDs)

## Median data of tau for teleosts specific 3RWGDs
levels(nodes.contrast.tau.3RWGDs.teleosts$Event)<-c("Speciation", "SSD", "FishWGD")
median.Tau.3RWGDs.teleosts<-NULL
median.Tau.3RWGDs.teleosts<-nodes.contrast.tau.3RWGDs.teleosts %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.Tau.3RWGDs.teleosts$pic.round<-paste0("Median = ",round(median.Tau.3RWGDs.teleosts$Tau.abs,4), sep="")
median.Tau.3RWGDs.teleosts$count<- paste0(median.Tau.3RWGDs.teleosts$Event,"\n","(n = ",median.Tau.3RWGDs.teleosts$freq,")")

##Testing on average expression levels for 3RWGDs
speciation.picExp.3RWGDs <-speciation.contrast(nodes.contrast.exp.3RWGDs.teleosts ,"Exp.abs")
duplication.picExp.3RWGDs1 <-duplication.contrast.3R(nodes.contrast.exp.3RWGDs.teleosts,"Exp.abs")
duplication.picExp.3RWGDs2 <-duplication.contrast(nodes.contrast.exp.3RWGDs.teleosts,"Exp.abs")
Pval.Exp.3RWGDs.teleosts1 <- two.tailed.wilcox(duplication.picExp.3RWGDs1,speciation.picExp.3RWGDs)
Pval.Exp.3RWGDs.teleosts2 <- two.tailed.wilcox(duplication.picExp.3RWGDs2,speciation.picExp.3RWGDs)

## Median data of exp levels for teleosts specific 3RWGDs
levels(nodes.contrast.exp.3RWGDs.teleosts$Event)<-c("Speciation", "SSD", "FishWGD")
median.Exp.3RWGDs.teleosts<-NULL
median.Exp.3RWGDs.teleosts<-nodes.contrast.exp.3RWGDs.teleosts %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.Exp.3RWGDs.teleosts$pic.round<-paste0("Median = ",round(median.Exp.3RWGDs.teleosts$Exp.abs,4), sep="")
median.Exp.3RWGDs.teleosts$count<- paste0(median.Exp.3RWGDs.teleosts$Event,"\n","(n = ",median.Exp.3RWGDs.teleosts$freq,")")

```


```{r Analyses_on_sure_SSD_trees, echo=F, message=FALSE, warning=FALSE, cache=T}

##Identifying teleost specific clades
nodes.contrast.tau.SSDs.teleosts<-nodes.contrast.tau.SSDs[which(nodes.contrast.tau.SSDs$label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Characoidei","Oryzias","Atherinomorphae","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
nodes.contrast.exp.SSDs.teleosts<-nodes.contrast.exp.SSDs[which(nodes.contrast.exp.SSDs$label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Characoidei","Oryzias","Atherinomorphae","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,]

##Testing on tau for SSDs
speciation.picTau.SSDs <-speciation.contrast(nodes.contrast.tau.SSDs.teleosts ,"Tau.abs")
duplication.picTau.SSDs <-duplication.contrast(nodes.contrast.tau.SSDs.teleosts,"Tau.abs")
Pval.Tau.SSDs.teleosts <- two.tailed.wilcox(duplication.picTau.SSDs,speciation.picTau.SSDs)

## Median data of tau for teleosts specific SSDs
levels(nodes.contrast.tau.SSDs.teleosts$Event)<-c("Speciation", "SSD")
median.Tau.SSDs.teleosts<-NULL
median.Tau.SSDs.teleosts<-nodes.contrast.tau.SSDs.teleosts %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.Tau.SSDs.teleosts$pic.round<-paste0("Median = ",round(median.Tau.SSDs.teleosts$Tau.abs,4), sep="")
#levels(median.Tau.SSDs.teleosts$Event)<-c("Speciation", "SSD")
median.Tau.SSDs.teleosts$count<- paste0(median.Tau.SSDs.teleosts$Event,"\n","(n = ",median.Tau.SSDs.teleosts$freq,")")

##Testing on average expression levels for SSDs
speciation.picExp.SSDs <-speciation.contrast(nodes.contrast.exp.SSDs.teleosts ,"Exp.abs")
duplication.picExp.SSDs <-duplication.contrast(nodes.contrast.exp.SSDs.teleosts,"Exp.abs")
Pval.Exp.SSDs.teleosts <- two.tailed.wilcox(duplication.picExp.SSDs,speciation.picExp.SSDs)

## Median data of exp levels for teleosts specific SSDs
levels(nodes.contrast.exp.SSDs.teleosts$Event)<-c("Speciation", "SSD")
median.Exp.SSDs.teleosts<-NULL
median.Exp.SSDs.teleosts<-nodes.contrast.exp.SSDs.teleosts %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.Exp.SSDs.teleosts$pic.round<-paste0("Median = ",round(median.Exp.SSDs.teleosts$Exp.abs,4), sep="")
median.Exp.SSDs.teleosts$count<- paste0(median.Exp.SSDs.teleosts$Event,"\n","(n = ",median.Exp.SSDs.teleosts$freq,")")

```


```{r Fig_3, eval=T, dpi=400,fig.width=8, fig.height=8, fig.cap= paste0("**Fig 3**: Tests of the ortholog conjecture by phylogenetic independent contrasts according to type of duplication in teleosts. ", length(Calibrated.3R.tau), " and ", length(Calibrated.3R.exp), " trees with whole genome duplications were used in (A)-(B), while ",  length(Calibrated.SSD.tau), " and ", length(Calibrated.SSD.exp), " trees with small-scale gene duplications were used in (C)-(D). PICs: Phylogenetic independent contrasts, FishWGD(s): Fish specific 3R whole genome duplicate(s) (or ohnologs), SSD(s): small-scale duplicate(s). PICs of teleosts specific nodes were used for these analyses.  *P* values are from Wilcoxon two-tailed tests. Since trees with whole genome duplicates may also contain SSDs apart from trusted FishWGDs, both types of duplicates are used for analyses in (A)-(B).")}

## Plotting 
dodge <- position_dodge(width = 0.05)

plot3A<- boxplot.new.3RWGD(nodes.contrast.tau.3RWGDs.teleosts,Pval.Tau.3RWGDs.teleosts1,Pval.Tau.3RWGDs.teleosts2,"Tau.3RWGD",median.Tau.3RWGDs.teleosts) + 
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("A. Test on ", tau, " for trusted FishWGDs"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.Tau.3RWGDs.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.043, ysize = 0.003)


plot3B<- boxplot.new.3RWGD(nodes.contrast.exp.3RWGDs.teleosts,Pval.Exp.3RWGDs.teleosts1,Pval.Exp.3RWGDs.teleosts2,"Exp.3RWGD",median.Exp.3RWGDs.teleosts) + 
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. Test on exp levels for trusted FishWGDs"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.Exp.3RWGDs.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.25, ysize = 0.019)

plot3C<- boxplot.new(nodes.contrast.tau.SSDs.teleosts,Pval.Tau.SSDs.teleosts,"Tau",median.Tau.SSDs.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  #geom_text(data = median.Tau, aes(label = pic.round),fontface = 2,hjust=0.5,vjust =-1)+
  ggtitle(expression(bold(paste("C. Test on ", tau, " for trusted SSDs"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.Tau.SSDs.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.4, y = 0.043, ysize = 0.003)

plot3D<- boxplot.new(nodes.contrast.exp.SSDs.teleosts,Pval.Exp.SSDs.teleosts,"Exp",median.Exp.SSDs.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Test on exp levels for trusted SSDs"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.Exp.SSDs.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.4, y = 0.25, ysize = 0.019)

##Plotting results together
  gridExtra::grid.arrange(plot3A,plot3B,plot3C,plot3D, nrow=2,ncol=2) 

```


Phylogenetic Independent Contrasts is a phylogenetic comparative method which is well suited to test whether gene duplication promotes more functional divergence than evolution of genes in the absence of duplication, if care is taken in its application [@Dunn2018; @Begum2021a]. We used gene expression as a proxy for function in vertebrates (Fig. 1), through two quantitative traits: i) tissue specificity, $\tau$ [@Yanai2005], and ii) average gene expression levels. To compute average gene expression levels, we normalized gene expression levels such that the cross-species expression data clearly separates into clusters per organ (Supplementary figs. S1A and S1B)). This way gene expression is comparable across species and tissues or organs.


```{r plot_Fig4A, eval=F,echo=F, message=FALSE, warning=FALSE, cache=T}

##Test tree (considering tree with minimal tip labels for demonstration)
test.tree.painted<-paint.tree(Duplicates.our.interest[[1200]])## tree with 9 tips
test.tree.painted$tip.label<-c("Species A","Species B","Species C","Species D", "Species A","Species C","Species D","Species E","Species F")
colSimmap <- c("#F8766D", "#00BFC4")
names(colSimmap) <- c("S", "D")

plot4A<-plotSimmap(test.tree.painted,col=colSimmap, lwd=1.5,ftype = "b",offset = 1,mar=c(1,3,2,3),type = "phylogram")+
title(main = "A. Schematic representation",line = 1,adj=0.1)+
tiplabels(pch=19,col="#F8766D",cex=1.5)+
nodelabels(node=c(10,11,12,15),frame="none",col=c("#F8766D","#00BFC4","#F8766D","#F8766D","#F8766D"),bg=c("#F8766D","#00BFC4","#F8766D","#F8766D","#F8766D"),pch = 21)+ 
#nodelabels("Post_predup_speciation",17,frame="none",adj=c(-0.1,-0.1),font=2,cex=0.7)+
nodelabels("Predup_speciation",10,frame="none",adj=c(-0.1,-0.1),font=2,cex=0.8)+
nodelabels("FishWGD or SSD",11,frame="none",adj=c(-0.1,-0.1),font=2,cex=0.8)+
nodelabels("Postdup_speciation",12,frame="none",adj=c(-0.1,-0.1),font=2,cex=0.8)+ 
nodelabels("Postdup_speciation",15,frame="none",adj=c(-0.1,-0.1),font=2,cex=0.8)+ 
legend("bottomleft",legend=c("Speciation","Duplication"), ncol=1, pch=17, col=c("#F8766D","#00BFC4"), cex=0.8,bty="y",text.font = 2)

## Saving the plot
#png("/Users/admin/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/Figures/Figure4A.png",width=550,height=300)
#plot4A
#dev.off()

```


```{r Analyses_on_sure_3RWGD_trees_our_method_teleosts, eval=T, echo=F, message=FALSE, warning=FALSE, cache=T}

duplicate.effect.teleosts.our.method<-duplicate.effect3R
duplicate.effect.exp.teleosts.our.method<-duplicate.effect3R.exp

###Testing result for tau and plotting data
paired.pval1.tau.3RWGD.teleost.our.method<-paired.wilcox(duplicate.effect.teleosts.our.method$predup.spe.pic,duplicate.effect.teleosts.our.method$duplication.pic)
paired.pval2.tau.3RWGD.teleost.our.method<-paired.wilcox(duplicate.effect.teleosts.our.method$predup.spe.pic,duplicate.effect.teleosts.our.method$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.3RWGD.teleost.our.method<-data.frame(pic=NA,label=NA)
Frame.tau.3RWGD.teleost.our.method<-rbind(Frame.tau.3RWGD.teleost.our.method,data.frame(pic=duplicate.effect.teleosts.our.method$predup.spe.pic,label="Predup_speciation"))
Frame.tau.3RWGD.teleost.our.method<-rbind(Frame.tau.3RWGD.teleost.our.method,data.frame(pic=duplicate.effect.teleosts.our.method$duplication.pic,label="FishWGD"))
Frame.tau.3RWGD.teleost.our.method<-rbind(Frame.tau.3RWGD.teleost.our.method,data.frame(pic=duplicate.effect.teleosts.our.method$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.3RWGD.teleost.our.method<-Frame.tau.3RWGD.teleost.our.method[-1,]
Frame.tau.3RWGD.teleost.our.method$Event <- factor(Frame.tau.3RWGD.teleost.our.method$label, levels=c("Predup_speciation", "FishWGD","Postdup_speciation"))

## For median data

median.data.3RWGD.teleost.our.method<-NULL
median.data.3RWGD.teleost.our.method<-Frame.tau.3RWGD.teleost.our.method %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.3RWGD.teleost.our.method$pic.round<-round(median.data.3RWGD.teleost.our.method$Median,4)
median.data.3RWGD.teleost.our.method$count<- paste0(median.data.3RWGD.teleost.our.method$Event," \n(n = ",median.data.3RWGD.teleost.our.method$freq,")")

###Testing result for exp levels and plotting data
paired.pval1.exp.3RWGD.teleost.our.method<-paired.wilcox(duplicate.effect.exp.teleosts.our.method$predup.spe.pic,duplicate.effect.exp.teleosts.our.method$duplication.pic)
paired.pval2.exp.3RWGD.teleost.our.method<-paired.wilcox(duplicate.effect.exp.teleosts.our.method$predup.spe.pic,duplicate.effect.exp.teleosts.our.method$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.3RWGD.teleost.our.method<-data.frame(pic=NA,label=NA)
Frame.exp.3RWGD.teleost.our.method<-rbind(Frame.exp.3RWGD.teleost.our.method,data.frame(pic=duplicate.effect.exp.teleosts.our.method$predup.spe.pic,label="Predup_speciation"))
Frame.exp.3RWGD.teleost.our.method<-rbind(Frame.exp.3RWGD.teleost.our.method,data.frame(pic=duplicate.effect.exp.teleosts.our.method$duplication.pic,label="FishWGD"))
Frame.exp.3RWGD.teleost.our.method<-rbind(Frame.exp.3RWGD.teleost.our.method,data.frame(pic=duplicate.effect.exp.teleosts.our.method$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.3RWGD.teleost.our.method<-Frame.exp.3RWGD.teleost.our.method[-1,]
Frame.exp.3RWGD.teleost.our.method$Event <- factor(Frame.exp.3RWGD.teleost.our.method$label, levels=c("Predup_speciation", "FishWGD","Postdup_speciation"))

## For median data
median.data.3RWGD.exp.teleost.our.method<-NULL
median.data.3RWGD.exp.teleost.our.method<-Frame.exp.3RWGD.teleost.our.method %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.3RWGD.exp.teleost.our.method$pic.round<-round(median.data.3RWGD.exp.teleost.our.method$Median,4)
median.data.3RWGD.exp.teleost.our.method$count<- paste0(median.data.3RWGD.exp.teleost.our.method$Event,"\n(n = ",median.data.3RWGD.exp.teleost.our.method$freq,")")

```



```{r Analyses_on_sure_SSD_trees_our_method_teleosts, eval=T,echo=F, message=FALSE, warning=FALSE, cache=T}

duplicate.effect.SSD.teleosts.our.method<-duplicate.effect.SSD[which(duplicate.effect.SSD$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Characoidei","Oryzias","Atherinomorphae","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,]

duplicate.effect.exp.SSD.teleosts.our.method<-duplicate.effect.SSD.exp[which(duplicate.effect.SSD.exp$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Characoidei","Oryzias","Atherinomorphae","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 

###Testing result for tau and plotting data
paired.pval1.tau.SSD.teleost.our.method<-paired.wilcox(duplicate.effect.SSD.teleosts.our.method$predup.spe.pic,duplicate.effect.SSD.teleosts.our.method$duplication.pic)
paired.pval2.tau.SSD.teleost.our.method<-paired.wilcox(duplicate.effect.SSD.teleosts.our.method$predup.spe.pic,duplicate.effect.SSD.teleosts.our.method$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.SSD.teleost.our.method<-data.frame(pic=NA,label=NA)
Frame.tau.SSD.teleost.our.method<-rbind(Frame.tau.SSD.teleost.our.method,data.frame(pic=duplicate.effect.SSD.teleosts.our.method$predup.spe.pic,label="Predup_speciation"))
Frame.tau.SSD.teleost.our.method<-rbind(Frame.tau.SSD.teleost.our.method,data.frame(pic=duplicate.effect.SSD.teleosts.our.method$duplication.pic,label="Duplication"))
Frame.tau.SSD.teleost.our.method<-rbind(Frame.tau.SSD.teleost.our.method,data.frame(pic=duplicate.effect.SSD.teleosts.our.method$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.SSD.teleost.our.method<-Frame.tau.SSD.teleost.our.method[-1,]
Frame.tau.SSD.teleost.our.method$Event <- factor(Frame.tau.SSD.teleost.our.method$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation"))
levels(Frame.tau.SSD.teleost.our.method$Event)<-c("Predup_speciation", "SSD","Postdup_speciation")

## For median data
library(tidyverse)
median.data.SSD.teleost.our.method<-NULL
median.data.SSD.teleost.our.method<-Frame.tau.SSD.teleost.our.method %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.SSD.teleost.our.method$pic.round<-round(median.data.SSD.teleost.our.method$Median,4)
#median.data.SSD.teleost.our.method$pic.round<-paste0("Median = ",round(median.data.SSD.teleost.our.method$Median,4), sep="")
median.data.SSD.teleost.our.method$count<- paste0(median.data.SSD.teleost.our.method$Event," \n(n = ",median.data.SSD.teleost.our.method$freq,")")

###Testing result for exp levels and plotting data
paired.pval1.exp.SSD.teleost.our.method<-paired.wilcox(duplicate.effect.exp.SSD.teleosts.our.method$predup.spe.pic,duplicate.effect.exp.SSD.teleosts.our.method$duplication.pic)
paired.pval2.exp.SSD.teleost.our.method<-paired.wilcox(duplicate.effect.exp.SSD.teleosts.our.method$predup.spe.pic,duplicate.effect.exp.SSD.teleosts.our.method$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.SSD.teleost.our.method<-data.frame(pic=NA,label=NA)
Frame.exp.SSD.teleost.our.method<-rbind(Frame.exp.SSD.teleost.our.method,data.frame(pic=duplicate.effect.exp.SSD.teleosts.our.method$predup.spe.pic,label="Predup_speciation"))
Frame.exp.SSD.teleost.our.method<-rbind(Frame.exp.SSD.teleost.our.method,data.frame(pic=duplicate.effect.exp.SSD.teleosts.our.method$duplication.pic,label="Duplication"))
Frame.exp.SSD.teleost.our.method<-rbind(Frame.exp.SSD.teleost.our.method,data.frame(pic=duplicate.effect.exp.SSD.teleosts.our.method$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.SSD.teleost.our.method<-Frame.exp.SSD.teleost.our.method[-1,]
Frame.exp.SSD.teleost.our.method$Event <- factor(Frame.exp.SSD.teleost.our.method$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ))
levels(Frame.exp.SSD.teleost.our.method$Event)<-c("Predup_speciation", "SSD","Postdup_speciation")

## For median data
median.data.SSD.exp.teleost.our.method<-NULL
median.data.SSD.exp.teleost.our.method<-Frame.exp.SSD.teleost.our.method %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.SSD.exp.teleost.our.method$pic.round<-round(median.data.SSD.exp.teleost.our.method$Median,4)
median.data.SSD.exp.teleost.our.method$count<- paste0(median.data.SSD.exp.teleost.our.method$Event," \n(n = ",median.data.SSD.exp.teleost.our.method$freq,")")

```


```{r Fig_4, eval=T,dpi=400,fig.width=8, fig.height=8, fig.cap= paste0("**Fig 4**: Tests of the ortholog conjecture by phylogenetic independent contrasts including speciations before and after duplications in teleosts. (A) Illustration of how we can test the impact of a duplication by comparing duplication nodes with speciation nodes before and after the duplication. (B)-(E) ", length(Duplicates.our.interest.3R),", ",length(Duplicates.our.interest.3R.exp), ", ", length(All.duplicates.interest.SSD), " and ", length(Duplicates.our.interest.SSD.exp), " trees passing our criteria of Fig. 4A, respectively were used for the analyses. We used the oldest duplication node along a path of a phylogeny for identification of pre-duplication and post-duplication speciations. We considered the contrasts of teleosts specific gene or genome duplication and descendant speciation nodes in (B)-(E) in absence of teleost-specific pre-duplication speciation nodes contrasts. PICs: Phylogenetic independent contrasts, exp: expression, Predup_speciation: pre-duplication speciation, Postdup_speciation: post-duplication speciation, FishWGD: fish specific 3R whole genome duplication, and SSD: small-scale duplication. Since contrasts are estimated for each tree, paired Wilcoxon tests were used to compare them.")}


## Loading previously generated plot
plot4A1 <- png::readPNG("~/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/Figures/Figure4A.png")
plot4A <-grid::rasterGrob(plot4A1)

plot4B<-boxplot.our.method.3RWGD(Frame.tau.3RWGD.teleost.our.method,paired.pval1.tau.3RWGD.teleost.our.method,paired.pval2.tau.3RWGD.teleost.our.method,"Tau")+   
  ggtitle(expression(bold(paste("B. Tissue specificity, ", tau, ", for ohnologs"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.data.3RWGD.teleost.our.method$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.033, ysize = 0.004, color = "black")+
  theme(axis.text.y = element_text(size = 12, face="bold"))
  

## Plot on 3RWGD expression levels considering the oldest duplication event for our method

plot4C<-boxplot.our.method.3RWGD(Frame.exp.3RWGD.teleost.our.method,paired.pval1.exp.3RWGD.teleost.our.method,paired.pval2.exp.3RWGD.teleost.our.method,"Exp") + 
  ggtitle(expression(bold(paste("C. Expression levels for ohnologs"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.data.3RWGD.exp.teleost.our.method$count)+       add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.23, ysize = 0.03, color = "black")+
  theme(axis.text.y = element_text(size = 12, face="bold"))
  

## Plot on sure SSD tau considering the oldest duplication event for our method

plot4D<-boxplot.our.method(Frame.tau.SSD.teleost.our.method,paired.pval1.tau.SSD.teleost.our.method,paired.pval2.tau.SSD.teleost.our.method,"Tau")+
  ggtitle(expression(bold(paste("D. Tissue specificity, ", tau, ", for other duplicates"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.data.SSD.teleost.our.method$count)+        
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.033, ysize = 0.004, color = "black")+
  theme(axis.text.y = element_text(size = 12, face="bold"))
  

## Plot on sure SSD expression levels considering the oldest duplication event for our method

plot4E<-boxplot.our.method(Frame.exp.SSD.teleost.our.method,paired.pval1.exp.SSD.teleost.our.method,paired.pval2.exp.SSD.teleost.our.method,"Exp") +   
  ggtitle(expression(bold(paste("E. Expression levels for other duplicates"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.data.SSD.exp.teleost.our.method$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.23, ysize = 0.03, color = "black")+
  theme(axis.text.y = element_text(size = 12, face="bold"))
  

##Plotting results together
  gridExtra::grid.arrange(plot4A,plot4B,plot4C,plot4D,plot4E,ncol=2,layout_matrix=rbind(c(1,1),c(2,3),c(4,5))) 

```

If duplication favors trait diversification, we expect higher Phylogenetic Independent Contrast values (PICs) for duplications than for speciations. This is indeed the case for both traits, and both for standardized phylogenies restricted to teleosts or including all 15 vertebrates of this study (Fig. 2). These results support the ortholog conjecture for vertebrates. Further analyses on synteny validated teleost third-round whole genome duplicates (FishWGDs) or ohnologs, and on trusted small-scale duplicates (SSDs) in teleosts (Figs. 3A-3D) also support the ortholog conjecture for both traits. So do similar analyses on all 15 vertebrates (Supplementary figs. S2A-S2D). We observed a much weaker impact of genome duplications than of small-scale duplications on trait divergence in terms of effect size. This could be because the ohnologs are much older than the small-scale paralogs (median age of ohnologs: `r round(median(nodes.contrast.tau.3RWGDs.teleosts$node.age[which(nodes.contrast.tau.3RWGDs.teleosts$events=="FishWGD")]),0)` Mya, (median age of SSDs: `r round(median(nodes.contrast.tau.SSDs.teleosts$node.age[which(nodes.contrast.tau.SSDs.teleosts$events=="duplication")]),0)` Mya, Wilcoxon two-tailed tests *P* < 2.2 $\times$ 10^-16^), or because of differences in evolutionary dynamics.  

\vspace{0.5cm}
If the effect of duplications is time-dependent [@Gu2005; @Begum2021b]. ), evolution along speciation branches following a duplication might be impacted by the duplication. This could impact testing of the ortholog conjecture if we simply compare all duplication branches to all speciation branches. To account for this effect, we compared contrasts between speciation nodes prior to duplications (pre-duplication speciations or "ancestral" speciations), duplication nodes, and speciation nodes following duplications (post-duplication speciations or "descendant" speciations) (Fig. 4A). We first restricted this comparison to the oldest duplication node along a path in each tree.  

\vspace{0.5cm}
The contrasts of duplications were always significantly higher than those of ancestral or descendant speciations (Figs. 4B-4E, Supplementary fig. S3). In addition, there was a trend for descendant speciations to have higher contrasts than ancestral speciations, indicating a longer-term impact of the duplication on evolutionary rates. Depending on duplication type and species sampling, this difference was sometimes but not always significant. This might be due in part to small effect sizes which are difficult to test on small subsets of the data, e.g. only ~`r round((length(All.duplicates.interest.SSD)/length(Calibrated.SSD.tau)*100),0)`% SSD trees fit the criteria of Fig. 4A; and in part to large divergence times between the FishWGD event and the next speciations. Randomized traits showed a pattern distinct from that of real data in all cases, indicating a biological pattern rather than a bias (Supplementary figs. S4 and S5)[@Begum2021a].Indeed after randomization the contrasts no longer follow a Brownian model, and become dependent on phylogenetic parameters such as node depth or node age. Hence, the contrasts increase as the node depths decrease towards present time along the randomized phylogenies (Supplementary figs. S4 and S5).    

\vspace{0.5cm}
Our result was robust when we considered the youngest duplication node along a path whenever available (Supplementary figs. S6 and S7). This indicates that the pattern we observed was not strongly affected by the timing of duplication nor by the risk of a few wrongly annotated genome duplication nodes. Overall, we find that phylogenetic contrasts support the ortholog conjecture in vertebrate gene expression evolution.


```{r plot_Fig5A1, eval=F,echo=F, message=FALSE, warning=FALSE, cache=T}

## Demonstration of trait jump model
##Test tree (considering tree with minimal tip labels for demonstration)
test.tree<-Calibrated.3R.exp[[414]]
test.tree@data$events[which(test.tree@data$node==11)]<-"speciation"
test.tree@data$events[which(test.tree@data$node==16)]<-"duplication"
test.tree@data$events[which(test.tree@data$node==17)]<-"duplication"

test.WGD.tree.painted<-paint.tree.mod(test.tree)## tree with 9 tips 
test.WGD.tree.painted$tip.label<-c("Species A","Species B","Species A","Species B", "Species C","Species D","Species E","Species E","Species E")
colSimmap <- c("#F8766D", "#008000","#00BFC4")
names(colSimmap) <- c("Speciation", "FishWGD","Duplication")

plot5A1<-plotSimmap(test.WGD.tree.painted,col=colSimmap, lwd=1.5,ftype = "b",offset = 0.7,mar=c(1,3,2,3),type = "phylogram")+
title("A. An example of trait jump model",line = 1,adj=0.1,cex=0.4)+
#tiplabels(pch=19,col="#F8766D",cex=0.5)+
tiplabels(pch=19,col="black",cex=0.4)+
  
# nodelabels("FishWGD",12,frame="none",adj=c(-0.1,-0.1),font=2,cex=0.6)+
edgelabels(edge=10,pie=c(0.6937,0.3063),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=1,pie=c(0.8733,0.1267),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=9,pie=c(0.5045,0.4955),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=2,pie=c(0.8018,0.1982),piecol=c("orange","purple"),cex=0.7)+ edgelabels(edge=3,pie=c(0.8333,0.1667),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=6,pie=c(0.2354,0.7646),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=4,pie=c(0.836,0.1640),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=5,pie=c(0.8575,0.1425),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=8,pie=c(0.5106,0.4894),piecol=c("orange","purple"),cex=0.7)+ edgelabels(edge=7,pie=c(0.7721,0.2279),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=11,pie=c(0.9599,0.0401),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=12,pie=c(0.8769,0.1231),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=13,pie=c(0.9866,0.0134),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=16,pie=c(0.15,0.85),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=15,pie=c(0.93,0.07),piecol=c("orange","purple"),cex=0.7)+
edgelabels(edge=14,pie=c(0.6965,0.3035),piecol=c("orange","purple"),cex=0.7)+
legend("topleft",legend=c("Low branch jump posterior","High branch jump posterior"), ncol=1, pch=16, col=c("orange", "purple"), cex=0.85,bty="n",text.font = 2)
#add.simmap.legend(colors=colSimmap,fsize=0.8)
legend(2,7,legend=c("Speciation","FishWGD","SSD"), ncol=1, pch=15, pt.cex=2, col=c("#F8766D", "#008000","#00BFC4"),cex= 0.9, bty="n",text.font = 3)

# Draw the color legend
# colfunc <- colorRampPalette(c("blue","orange"))
# legend_image <- as.raster(matrix(colfunc(20), ncol=1))
# plot(c(0,2),c(0,1), pch = 19, cex=1, col = colfunc(20),type = 'n',axes = F,xlab = '', ylab = '', main = "Branch jump posterior",cex.main=0.65)
# #plot(c(0,1),c(0,1),type = 'n', axes = F,xlab = '', ylab = '', main = "Branch jump posterior")
# text(x=1.6, y = seq(0,1,l=3), labels = seq(0,1,l=3), cex=0.8)
# rasterImage(legend_image, 0, 0, 1,1,cex=0.5)
# dev.off()


```


```{r plot_Fig5A2, eval=F,echo=F, message=FALSE, warning=FALSE, cache=T}

## Pablo supplied code
eff_seed <- 29925 #For different simulations comment out this line and uncomment the next line.
#eff_seed <- sample(1:2^15, 1)
print(sprintf("Seed for session: %s", eff_seed))##29925
set.seed(eff_seed)

#tree <- read.tree("testtree.tre")
tree <- Calibrated.3R.exp[[414]]@phylo
tree$tip.label<-c("Species A","Species B","Species A","Species B", "Species C","Species D","Species E","Species E","Species E")

#Adding colors to branches. Branches with lower jump probabilities have lighter colors.
col1 <- "#ec494b" # "red"
col1s <- "#fac7c8" # lighter red
col2 <- "#125f03" # "green"
col2s <- "#b8cfb3" # lighter green
col3 <- "#1fa4a7" # "ligth blue"
col3s <- "#bbe3e4" # lighter blue
    
colores <- rep(col1s,length(tree$edge[,1]))
colores[c(3,6)] <- col2s
colores[c(14,15,16)] <- col3s
colores[6] <- col2
colores[16] <- col3
colores[c(8,9)] <- col1

#par(mfrow=c(1,2))
#plot(tree)
#edgelabels()

sig2 <- 10 #BM variance.
F <- 100
tree$edge.length <- tree$edge.length*F

#Modified from Luke Harmon's and Liam Revell's 'Comparative Methods in R"
## simulate evolution along each edge
X<-lapply(tree$edge.length,function(x) c(0,cumsum(rnorm(n=x,sd=sqrt(sig2)))))

#Add jumps to edges 6, 16, 8, and 9
jumpSize <- 500
midpoint <- round(length(X[6][[1]])/2)
X[6][[1]][midpoint:length(X[6][[1]])] <- X[6][[1]][midpoint:length(X[6][[1]])]-jumpSize
midpoint <- round(length(X[16][[1]])/2)
X[16][[1]][midpoint:length(X[16][[1]])] <- X[16][[1]][midpoint:length(X[16][[1]])]+jumpSize

jumpSize <- 250
midpoint <- round(length(X[8][[1]])/2)
X[8][[1]][midpoint:length(X[8][[1]])] <- X[8][[1]][midpoint:length(X[8][[1]])]+jumpSize
midpoint <- round(length(X[9][[1]])/2)
X[9][[1]][midpoint:length(X[9][[1]])] <- X[9][[1]][midpoint:length(X[9][[1]])]-jumpSize

## Simulate on the tree
ll<-tree$edge.length+1
for(i in 1:nrow(tree$edge)){
    pp<-which(tree$edge[,2]==tree$edge[i,1])
    if(length(pp)>0) X[[i]]<-X[[i]]+X[[pp]][ll[pp]]
    else X[[i]]<-X[[i]]+X[[1]][1]
}
## Get the starting and ending points of each edge for plotting
H<-nodeHeights(tree)

dev.off()
## Plot
par(mar=c(2, 3, 2, 2), xpd=TRUE)
plot(H[1,1],X[[1]][1],ylim=range(X),xlim=c(0,max(H)+15000),xaxt='n', yaxt='n',xlab=expression(bold("Evolutionary time")),ylab=expression(bold("Trait")), axes= FALSE,frame.plot = FALSE, type="l",line=1, cex.lab=1.3)+title(main= "B. Phylogenetic simulation of the example",frame.plot = FALSE, type="l",line=1,font.main=2,cex.main=1.5,adj=0)

#plot(H[1,1],X[[1]][1],ylim=range(X),xlim=c(0,max(H)+15000),xaxt='n', yaxt='n',xlab=expression(bold("Evolutionary time")),ylab=expression(bold("Trait")), axes= FALSE,main= "B. Phylogenetic simulation of the example",frame.plot = FALSE, type="l",line=1, cex.main=1.5)

#
for(i in 1:length(X)) lines(H[i,1]:H[i,2],X[[i]],col=colores[i])

## For axes with line
box(bty="l")
## To add arrow at the axes
usr <- par("usr")
shape::Arrowhead(
  x0 = usr[1:2], 
  y0 = usr[4:3], 
  angle = c(90, 0), 
  xpd = T
)

## Tip labels
yy<-sapply(1:length(tree$tip.label),function(x,y) which(x==y),y=tree$edge[,2])
yy<-sapply(yy,function(x,y) y[[x]][length(y[[x]])],y=X)
text(x=max(H)+7500,y=yy,tree$tip.label, cex = 0.9, font=2)

```


### **Lévy’s trait jump: sudden shifts in gene expression after small-scale duplication**  

Continuous traits can change in evolution in sudden shifts or "jumps" in values, without a change of the evolutionary rate outside of this jump. If such jumps were common in gene evolution, this would impact characterization of the ortholog conjecture. Hence, we applied the method "levolution" software [@Duchen2017; @Duchen2021] that couples a Brownian model with occasional jumps occurring as a Poisson process, to infer the locations and strength of evolutionary jumps for gene expression traits (Fig. 5A and 5B). Under the ortholog conjecture, our expectation is that there should be more jumps after duplications than after speciations.    


```{r jump_analyses_on_sure_3RWGD_trees_teleosts, echo=F, message=FALSE, warning=FALSE, cache=T}

# Creating summary statistics of tau for sure 3RWGD trees 
tree.jump.3RWGD.tau <- data.frame(
  group=c("Supporting trait jump model","No support for trait jump model"),
  value=c(435,311)
)

#Processing data for tau of 3RWGD trees
tree.jump.3RWGD.tau <- tree.jump.3RWGD.tau %>% 
  arrange(desc(group)) %>%
  mutate(prop = value / sum(tree.jump.3RWGD.tau$value) *100) %>%
  mutate(percent=paste0(prop = round((value / sum(tree.jump.3RWGD.tau$value) *100),0),"%","\n(",value," trees)", sep="")) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )

## Statistics of trait jumps in tau for sure teleost 3RWGDs
pval.jump.3RWGD.tau.fish.spe2dup <- paired.wilcox(info.jump.trees.final2.3RWGD.tau.fish$spe.jump.prob,info.jump.trees.final2.3RWGD.tau.fish$dup.jump.prob) 
pval.jump.3RWGD.tau.fish.spe2WGD <- paired.wilcox(info.jump.trees.final2.3RWGD.tau.fish$spe.jump.prob,info.jump.trees.final2.3RWGD.tau.fish$WGD.jump.prob) 

## Data frame of jumps in tau for sure teleost 3RWGDs
dtfr.3RWGD.tau.fish<-NULL
dtfr.3RWGD.tau.fish<-data.frame(prop.jump=info.jump.trees.final2.3RWGD.tau.fish$spe.jump.prob,events="speciation")
dtfr.3RWGD.tau.fish<-rbind(dtfr.3RWGD.tau.fish,data.frame(prop.jump=info.jump.trees.final2.3RWGD.tau.fish$dup.jump.prob,events="duplication"))
dtfr.3RWGD.tau.fish<-rbind(dtfr.3RWGD.tau.fish,data.frame(prop.jump=info.jump.trees.final2.3RWGD.tau.fish$WGD.jump.prob,events="FishWGD"))
dtfr.3RWGD.tau.fish$Tau.abs <-abs(dtfr.3RWGD.tau.fish$prop.jump)
dtfr.3RWGD.tau.fish$Event<-factor(dtfr.3RWGD.tau.fish$events, levels=c("speciation","duplication","FishWGD"))
levels(dtfr.3RWGD.tau.fish$Event)<-c("Speciation","SSD","FishWGD")

##Median data of jumps in tau for sure teleost 3RWGDs
median.data.jump.3RWGD.tau.fish<-NULL
median.data.jump.3RWGD.tau.fish<- dtfr.3RWGD.tau.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Tau.abs=median(prop.jump),
            freq = n())
median.data.jump.3RWGD.tau.fish$pic.round<-paste0("Median = ",round(median.data.jump.3RWGD.tau.fish$Tau.abs,4), sep="")
median.data.jump.3RWGD.tau.fish$count<- paste0(median.data.jump.3RWGD.tau.fish$Event,"\n","(n = ",median.data.jump.3RWGD.tau.fish$freq,")")

# Creating summary statistics of average expression levels for sure teleost 3RWGD trees 
tree.jump.3RWGD.exp <- data.frame(
  group=c("Supporting trait jump model","No support for trait jump model"),
  value=c(398,376)
)

##Processing data for average expression of teleost sure 3RWGD trees
tree.jump.3RWGD.exp <- tree.jump.3RWGD.exp %>% 
  arrange(desc(group)) %>%
  mutate(prop = value / sum(tree.jump.3RWGD.exp$value) *100) %>%
  mutate(percent=paste0(prop = round((value / sum(tree.jump.3RWGD.exp$value) *100),0),"%","\n(",value," trees)", sep="")) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )

## Statistics of trait jumps in average expression levels for sure teleost 3RWGDs
pval.jump.3RWGD.exp.fish.spe2dup <- paired.wilcox(info.jump.trees.final2.3RWGD.fish$spe.jump.prob,info.jump.trees.final2.3RWGD.fish$dup.jump.prob) 
pval.jump.3RWGD.exp.fish.spe2WGD <- paired.wilcox(info.jump.trees.final2.3RWGD.fish$spe.jump.prob,info.jump.trees.final2.3RWGD.fish$WGD.jump.prob) 

## Data frame of jumps in average expression levels for sure teleost 3RWGDs
dtfr.3RWGD.exp.fish<-NULL
dtfr.3RWGD.exp.fish<-data.frame(prop.jump=info.jump.trees.final2.3RWGD.fish$spe.jump.prob,events="speciation")
dtfr.3RWGD.exp.fish<-rbind(dtfr.3RWGD.exp.fish,data.frame(prop.jump=info.jump.trees.final2.3RWGD.fish$dup.jump.prob,events="duplication"))
dtfr.3RWGD.exp.fish<-rbind(dtfr.3RWGD.exp.fish,data.frame(prop.jump=info.jump.trees.final2.3RWGD.fish$WGD.jump.prob,events="FishWGD"))
dtfr.3RWGD.exp.fish$Exp.abs <-abs(dtfr.3RWGD.exp.fish$prop.jump)
dtfr.3RWGD.exp.fish$Event<-factor(dtfr.3RWGD.exp.fish$events, levels=c("speciation","duplication","FishWGD"))
levels(dtfr.3RWGD.exp.fish$Event)<-c("Speciation","SSD","FishWGD")

##Median data of jumps in average expression levels for sure teleost 3RWGDs
median.data.jump.3RWGD.exp.fish<-NULL
median.data.jump.3RWGD.exp.fish<- dtfr.3RWGD.exp.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.3RWGD.exp.fish$pic.round<-paste0("Median = ",round(median.data.jump.3RWGD.exp.fish$Exp.abs,4), sep="")
median.data.jump.3RWGD.exp.fish$count<- paste0(median.data.jump.3RWGD.exp.fish$Event,"\n","(n = ",median.data.jump.3RWGD.exp.fish$freq,")")

```



```{r jump_analyses_on_sure_SSD_trees_teleosts, echo=F, message=FALSE, warning=FALSE, cache=T}

## Statistics of trait jumps in tau for sure teleost SSDs
pval.jump.SSD.tau.fish <- paired.wilcox(info.jump.trees.SSD.tau.final1.fish$spe.jump.prob,info.jump.trees.SSD.tau.final1.fish$dup.jump.prob) 

## Data frame of jumps in tau for sure teleost SSDs
dtfr.SSD.tau.fish<-NULL
dtfr.SSD.tau.fish<-data.frame(prop.jump=info.jump.trees.SSD.tau.final1.fish$spe.jump.prob,events="speciation")
dtfr.SSD.tau.fish<-rbind(dtfr.SSD.tau.fish,data.frame(prop.jump=info.jump.trees.SSD.tau.final1.fish$dup.jump.prob,events="duplication"))
dtfr.SSD.tau.fish$Tau.abs <-abs(dtfr.SSD.tau.fish$prop.jump)
dtfr.SSD.tau.fish$Event<-factor(dtfr.SSD.tau.fish$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.tau.fish$Event)<-c("Speciation","SSD")

##Median data of jumps in tau for sure teleost SSDs
median.data.jump.SSD.tau.fish<-NULL
median.data.jump.SSD.tau.fish<- dtfr.SSD.tau.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Tau.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.tau.fish$pic.round<-paste0("Median = ",round(median.data.jump.SSD.tau.fish$Tau.abs,4), sep="")
median.data.jump.SSD.tau.fish$count<- paste0(median.data.jump.SSD.tau.fish$Event,"\n","(n = ",median.data.jump.SSD.tau.fish$freq,")")


## Statistics of trait jumps in average expression levels for sure teleost SSDs
pval.jump.SSD.exp.fish <- paired.wilcox(info.jump.trees.SSD.exp.final1.fish$spe.jump.prob,info.jump.trees.SSD.exp.final1.fish$dup.jump.prob) 


## Data frame of jumps in average expression levels for sure teleost SSDs
dtfr.SSD.exp.fish<-NULL
dtfr.SSD.exp.fish<-data.frame(prop.jump=info.jump.trees.SSD.exp.final1.fish$spe.jump.prob,events="speciation")
dtfr.SSD.exp.fish<-rbind(dtfr.SSD.exp.fish,data.frame(prop.jump=info.jump.trees.SSD.exp.final1.fish$dup.jump.prob,events="duplication"))
dtfr.SSD.exp.fish$Exp.abs <-abs(dtfr.SSD.exp.fish$prop.jump)
dtfr.SSD.exp.fish$Event<-factor(dtfr.SSD.exp.fish$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.exp.fish$Event)<-c("Speciation","SSD")

##Median data of jumps in average expression levels for sure teleost SSDs
median.data.jump.SSD.exp.fish<-NULL
median.data.jump.SSD.exp.fish<- dtfr.SSD.exp.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.exp.fish$pic.round<-paste0("Median = ",round(median.data.jump.SSD.exp.fish$Exp.abs,4), sep="")
median.data.jump.SSD.exp.fish$count<- paste0(median.data.jump.SSD.exp.fish$Event,"\n","(n = ",median.data.jump.SSD.exp.fish$freq,")")

```


```{r Fig_5, eval=T,dpi=400,fig.width=8, fig.height=8, fig.cap= paste0("**Fig 5**: Trait jump models for tissues-specificity and for average gene expression levels contrasting types of duplication in teleosts. FishWGD: fish specific 3R whole genome duplication, SSD: small-scale duplication, and exp: average gene expression levels. (A) An example of a FishWGD tree to demonstrate the branch-specific posterior probability of trait jumps as inferred by levolution (Duchen et al. 2017; Duchen et al. 2021) in a phylogenetic tree. (B) Phylogenetic simulation following the example tree of Fig. 5A. Colors show branches belonging to different events as shown in Fig. 5A. Branches following jumps in their traits are shown by darker shades. (C)-(E) Proportion of trait jumps according to type of duplications and to expression traits. We used a strict branch jump posterior cutoff of $\\ge$ 0.7 to consider a rapid shift in trait values. Only teleosts-specific clades were used for the analyses. We used ", length(Calibrated.SSD.tau.jump), " and ", length(Calibrated.SSD.exp.jump), " trusted SSD trees in (B) and (C), while ", length(Calibrated.3RWGD.tau.jump), " and ", length(Calibrated.3RWGD.exp.jump), " trusted FishWGD trees in (D) and (E), respectively showing at least one branch jump with the above-mentioned threshold along a phylogeny for our analyses. Since proportion of shifts in traits per branch of events is estimated for each tree, paired Wilcoxon test is used to compare the difference.")}


## Loading previously generated plot
#plot5A1 <- png::readPNG("~/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/Plot3A_demo.png")
plot5A <- png::readPNG("~/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/Figures/Figure5A.png")
plot5A1 <-grid::rasterGrob(plot5A)

plot5AA <- png::readPNG("~/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/Figures/Figure5A2.png")
plot5A2 <-grid::rasterGrob(plot5AA)

# Piechart of tau for sure 3RWGDs
plot5BI<-ggplot(tree.jump.3RWGD.tau, aes(x="", y=prop, fill=group)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0) +
  theme_void() + 
  theme(legend.title=element_blank(),legend.position=c(1.1, 0.8)) +
  geom_text(aes(y = ypos, label = percent), color = "black", size=4,fontface = "bold") +
  scale_fill_brewer(palette = "Greens")

plot5DI<-ggplot(tree.jump.3RWGD.exp, aes(x="", y=prop, fill=group)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0) +
  theme_void() + 
  theme(legend.title=element_blank(),legend.position=c(1.2, 0.8)) +
  geom_text(aes(y = ypos, label = percent), color = "black", size=4,fontface = "bold") +
  scale_fill_brewer(palette = "Greens")

plot5BM <- vioplot.new2(dtfr.SSD.tau.fish,pval.jump.SSD.tau.fish,"Tau",median.data.jump.SSD.tau.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("C. Jumps in ", tau, " for trusted SSD trees"))))+
  ylab(expression(bold(paste("Proportions of trait jumps")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.8, ymin = 0.75, ymax = 0.75, alpha = 0.8,colour = "blue")+
  annotate("text", x = 1.5, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.tau.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.tau.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.tau.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.tau.fish$count)+  
  theme(axis.text.y = element_text(size=12,face="bold"))+
  add_phylopic(fish_pic, alpha = 1,x = 2.35, y = 0.97, ysize = 0.1, color = "black")  

plot5CM <- vioplot.new2(dtfr.SSD.exp.fish,pval.jump.SSD.exp.fish,"Exp",median.data.jump.SSD.exp.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("D. Jumps in exp levels for trusted SSD trees"))))+
  #ylab(expression(bold(paste("Proportions of jump events for teleosts ")))) +
  ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.8, ymin = 0.75, ymax = 0.75, alpha = 0.8,colour = "blue")+
  annotate("text", x = 1.5, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.exp.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.exp.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.exp.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.exp.fish$count)+   
  theme(axis.text.y = element_text(size=12,face="bold"))+
  add_phylopic(fish_pic, alpha = 1,x = 2.35, y = 0.97, ysize = 0.1, color = "black")  

plot5DM <- vioplot.new.3RWGD2(dtfr.3RWGD.tau.fish,pval.jump.3RWGD.tau.fish.spe2dup,pval.jump.3RWGD.tau.fish.spe2WGD,"Tau.3RWGD",median.data.jump.3RWGD.tau.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("E. Jumps in ", tau, " for trusted FishWGD trees"))))+
  ylab(expression(bold(paste("Proportions of trait jumps")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.8, ymin = 0.65, ymax = 0.65, alpha = 0.8,colour = "blue")+
  annotate("text", x = 1.5, y = 0.75, label= as.character(paste0("italic(",pval.jump.3RWGD.tau.fish.spe2dup,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.75, label= as.character(paste0("italic(",pval.jump.3RWGD.tau.fish.spe2dup,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.75, label= as.character(paste0("italic(",pval.jump.3RWGD.tau.fish.spe2dup,")")), parse=T, size=4.04)+
  annotate("rect", xmin = 1.2, xmax = 2.8, ymin = 0.88, ymax = 0.88, alpha = 0.8,colour = "blue")+
  annotate("text", x = 2, y = 0.95,label= as.character(paste0("italic(",pval.jump.3RWGD.tau.fish.spe2WGD,")")), parse=T,size=4)+
  annotate("text", x = 2, y = 0.95,label= as.character(paste0("italic(",pval.jump.3RWGD.tau.fish.spe2WGD,")")), parse=T,size=4.02)+
  annotate("text", x = 2, y = 0.95,label= as.character(paste0("italic(",pval.jump.3RWGD.tau.fish.spe2WGD,")")), parse=T,size=4.04)+
  scale_x_discrete(labels=median.data.jump.3RWGD.tau.fish$count)+ 
  theme(axis.text.y = element_text(size=12,face="bold"))+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.97, ysize = 0.1,color = "black") 


plot5EM <- vioplot.new.3RWGD2(dtfr.3RWGD.exp.fish,pval.jump.3RWGD.exp.fish.spe2dup,pval.jump.3RWGD.exp.fish.spe2WGD,"Exp.3RWGD",median.data.jump.3RWGD.exp.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("F. Jumps in exp levels for trusted FishWGD trees"))))+
  #ylab(expression(bold(paste("Proportions of jump events")))) +
  ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.8, ymin = 0.65, ymax = 0.65, alpha = 0.8,colour = "blue")+
  annotate("text", x = 1.5, y = 0.75, label= as.character(paste0("italic(",pval.jump.3RWGD.exp.fish.spe2dup,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.75, label= as.character(paste0("italic(",pval.jump.3RWGD.exp.fish.spe2dup,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.75, label= as.character(paste0("italic(",pval.jump.3RWGD.exp.fish.spe2dup,")")), parse=T, size=4.04)+
  annotate("rect", xmin = 1.2, xmax = 2.8, ymin = 0.88, ymax = 0.88, alpha = 0.8,colour = "blue")+
  annotate("text", x = 2, y = 0.95, label= as.character(paste0("italic(",pval.jump.3RWGD.exp.fish.spe2WGD,")")), parse=T, size=4)+
  annotate("text", x = 2, y = 0.95, label= as.character(paste0("italic(",pval.jump.3RWGD.exp.fish.spe2WGD,")")), parse=T, size=4.02)+
  annotate("text", x = 2, y = 0.95, label= as.character(paste0("italic(",pval.jump.3RWGD.exp.fish.spe2WGD,")")), parse=T, size=4.04)+
 scale_x_discrete(labels=median.data.jump.3RWGD.exp.fish$count)+ 
  theme(axis.text.y = element_text(size=12,face="bold"))+
  add_phylopic(fish_pic, alpha = 1,x = 3.33, y = 0.97, ysize = 0.1, color = "black") 

 ##Drawing subplots
  #plot3B <- ggdraw() +
  #  draw_plot(plot3BM,scale=1.0) +
  #  draw_plot(plot3BI, x = .4, y = .4, width = .5, height = .9, scale = 0.5)
  #plot3B <-plot3BM + patchwork::inset_element(plot3BI, 0.5, 0.5, 1, 1)
  
  #plot3D <- ggdraw() +
  #  draw_plot(plot3DM,scale=1.0) +
  #  draw_plot(plot3DI, x = .4, y = .4, width = .5, height = .4, scale=0.7)
  

##Plotting results together
 # grid.arrange(plot5A1,plot5A2,plot5BM,plot5CM,plot5DM,plot5EM,ncol=2, layout_matrix=rbind(c(1,2),c(3,4),c(5,6)))
  grid.arrange(plot5A1,plot5A2,plot5BM,plot5CM,plot5DM,plot5EM,ncol=2,nrow=3)

```


```{r stats_on_props_of_jumps_in_calibrated_trees, echo=F, message=FALSE, warning=FALSE, cache=T}

## Setting seed number for reproducibility again
set.seed(786)

## Proportionality test on jumps in tau for 4247 calibrated trees
prop.tau.pval<- proportion.test(2636,1611,4247,4247)

## Proportionality test on jumps in average gene expression levels for 4603 calibrated trees
prop.exp.pval<- proportion.test(2604,1999,4603,4603)

##Asymmetric vs. symmetric branches for average expression levels
## Matching indices of trees with dataframe of jumps in expression levels
info.jump.trees$Gene<-info.jump.trees$index.old
df.exp1<-merge(info.jump.trees,test.output.exp.processed,by=c("Gene"))
df.exp1<-df.exp1[c(7:8,12)]
jump.symmetric<-df.exp1[duplicated(df.exp1),]
jump.symmetric<-jump.symmetric[order(jump.symmetric$tree.num,jump.symmetric$From),] #4398
jump.asymmetric<-anti_join(df.exp1,jump.symmetric,by=c("tree.num","From")) #8310
sym.exp.jump.branch<-round((((nrow(jump.symmetric)*2)/(nrow(df.exp1)))*100),2)
asym.exp.jump.branch<-round(((nrow(jump.asymmetric)/nrow(df.exp1))*100),2)

##Asymmetric vs. symmetric branches for tau
## Matching indices of trees with dataframe of jumps in tau
info.jump.trees.tau$Gene<-info.jump.trees.tau$index.old
df.tau1<-merge(info.jump.trees.tau,test.output.tau.processed,by=c("Gene"))
df.tau1<-df.tau1[c(8,10,14)]
df.tau1<-df.tau1[order(df.tau1$tree.num,df.tau1$From),]
jump.symmetric.tau<-df.tau1[duplicated(df.tau1),]
jump.symmetric.tau<-jump.symmetric.tau[order(jump.symmetric.tau$tree.num,jump.symmetric.tau$From),]
jump.asymmetric.tau<-anti_join(df.tau1,jump.symmetric.tau,by=c("tree.num","From"))
sym.tau.jump.branch<-round(((nrow(df.tau1[duplicated(df.tau1),])*2/nrow(df.tau1))*100),2)
asym.tau.jump.branch<-round((((nrow(jump.asymmetric.tau))/(nrow(df.tau1)))*100),2)


## Proportionality test on jumps in tau for 746 3RWGD calibrated trees
prop.tau.pval.3RWGDs<- proportion.test(435,311,746,746)

## Proportionality test on jumps in exp levels for 774 3RWGD calibrated trees
prop.exp.pval.3RWGDs<- proportion.test(398,376,774,774)

## Proportionality test on jumps in tau for 2482 SSD calibrated trees
prop.tau.pval.SSDs<- proportion.test(1557,925,2482,2482)

## Proportionality test on jumps in exp levels for 2717 SSD calibrated trees
prop.exp.pval.SSDs<- proportion.test(1593,1124,2717,2717)

## Proportionality test on jumps in average gene expression levels for 4603 calibrated trees
prop.exp.pval<- proportion.test(2604,1999,4603,4603)


```


First, we do detect jumps in the evolution of gene expression, affecting more than half of all gene trees (Figs. S8A and S8B), using a branch jump posterior threshold of 0.7. Second, there are indeed more than twice as many jumps on duplication branches than on speciation branches for small scale duplications (Figs. 5C and 5D). Third, there are almost no jumps after whole genome duplication (Figs. 5E and 5F). There are jumps on the speciation and SSD branches of WGD trees, thus the lack of jumps on WGD branches is not due to a bias in WGD genes. If different types of duplications are not distinguished, in a naive testing of the ortholog conjecture, there is overall a significantly higher proportion of jump per branch for both the traits in teleosts and in vertebrates following duplications than speciations (Supplementary figs. S8C-S8F). Such rapid trait shifts are slightly more asymmetric for $\tau$ (Asymmetric *vs.* symmetric jump branches: `r asym.tau.jump.branch`% *vs.* `r sym.tau.jump.branch`%, two-sample test for equality of proportions, $\chi^2$ test, *P* = 4.05 $\times$ 10^-3^) and are slightly more symmetric for average gene expression levels (Asymmetric *vs.* symmetric jump branches: `r asym.exp.jump.branch`% *vs.* `r sym.exp.jump.branch`%, two-sample test for equality of proportions, $\chi^2$ test, *P* = 1.57 $\times$ 10^-7^).  

\vspace{0.5cm}
When we applied PICs on the phylogenies supporting trait jumps in all 15 species, we observed a clear support for the ortholog conjecture with more than 21% higher trait contrasts for paralogs than orthologs (Supplementary figs. S9A and S9B). This suggests that higher PICs could sometimes be due to these rapid shifts in trait values. For a better understanding, we removed the contrasts of the nodes for which the daughter branch(es) experience rapid shifts in their traits. In consequence, the ortholog conjecture is no longer supported for the traits (Supplementary figs. S9C and S9D). This confirms that trait jumps play an important role in duplicate evolution. Interestingly, there is PIC support for the ortholog conjecture for the phylogenies not supporting a trait jump model (Supplementary figs. S9E and S9F), implying that these are genes which evolve functional change after duplication in a more continuous manner. Our results were robust when we replicated our analyses on teleosts only (Supplementary fig. S10).  

          
```{r OC_analyses_on_SSDs_jump_trees_teleosts, eval=T,echo=F, message=FALSE, warning=FALSE, cache=T}

##Testing on tau for SSD trees supporting jumps in teleosts
speciation.tau.jump.SSD.fish <-speciation.contrast(nodes.contrast.SSD.tau.jumps.fish ,"Tau.abs")
duplication.tau.jump.SSD.fish <-duplication.contrast(nodes.contrast.SSD.tau.jumps.fish,"Tau.abs")
#duplication.tau.jump.SSD.fish <-nodes.contrast.SSD.tau.jumps.fish$Tau.abs[which(nodes.contrast.SSD.tau.jumps.fish$Event=="SSD")]

##Stats on tau for SSD trees supporting jumps in teleosts
Pval.tau.jump.SSD.fish <- two.tailed.wilcox(duplication.tau.jump.SSD.fish,speciation.tau.jump.SSD.fish)

nodes.contrast.SSD.tau.jumps.fish$Event<-ifelse(nodes.contrast.SSD.tau.jumps.fish$Event %in% "speciation", "Speciation","SSD")
nodes.contrast.SSD.tau.jumps.fish$Event<-factor( nodes.contrast.SSD.tau.jumps.fish$Event, levels=c( "Speciation", "SSD" ))
levels(nodes.contrast.SSD.tau.jumps.fish$Event)<-c("Speciation","SSD")

## Median data for SSD trees supporting jump in tau for teleosts
median.tau.jump.SSD.teleosts<-NULL
median.tau.jump.SSD.teleosts<-nodes.contrast.SSD.tau.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.jump.SSD.teleosts$pic.round<-paste0("Median = ",round(median.tau.jump.SSD.teleosts$Tau.abs,4), sep="")
median.tau.jump.SSD.teleosts$count<- paste0(median.tau.jump.SSD.teleosts$Event,"\n","(n = ",median.tau.jump.SSD.teleosts$freq,")")

##Testing on tau after removal of nodes for which daughter branches support jumps in SSD trees in teleosts
speciation.tau.jump.SSD.fish.removing.jumpnode <-speciation.contrast(nodes.contrast.SSD.tau.jumps.fish.removing.jump.node ,"Tau.abs")
duplication.tau.jump.SSD.fish.removing.jumpnode <-duplication.contrast(nodes.contrast.SSD.tau.jumps.fish.removing.jump.node,"Tau.abs")

##Stats for trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in SSD trees in teleosts
Pval.tau.jump.SSD.fish.removing.jumpnode <- two.tailed.wilcox(duplication.tau.jump.SSD.fish.removing.jumpnode,speciation.tau.jump.SSD.fish.removing.jumpnode)

nodes.contrast.SSD.tau.jumps.fish.removing.jump.node$Event<-ifelse(nodes.contrast.SSD.tau.jumps.fish.removing.jump.node$Event %in% "speciation", "Speciation","SSD")
nodes.contrast.SSD.tau.jumps.fish.removing.jump.node$Event<-factor( nodes.contrast.SSD.tau.jumps.fish.removing.jump.node$Event, levels=c( "Speciation", "SSD" ))
levels(nodes.contrast.SSD.tau.jumps.fish.removing.jump.node$Event)<-c("Speciation","SSD")

## Median data of trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in SSD trees in teleosts
median.tau.removing.SSD.jumpnode.teleosts<-NULL
median.tau.removing.SSD.jumpnode.teleosts<-nodes.contrast.SSD.tau.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.removing.SSD.jumpnode.teleosts$pic.round<-paste0("Median = ",round(median.tau.removing.SSD.jumpnode.teleosts$Tau.abs,4), sep="")
median.tau.removing.SSD.jumpnode.teleosts$count<- paste0(median.tau.removing.SSD.jumpnode.teleosts$Event,"\n","(n = ",median.tau.removing.SSD.jumpnode.teleosts$freq,")")

##Testing on tau for trees not supporting jump in SSD trees in teleosts
speciation.tau.nojump.SSD.fish <-speciation.contrast(nodes.contrast.SSD.tau.nojumps.fish ,"Tau.abs")
duplication.tau.nojump.SSD.fish <-duplication.contrast(nodes.contrast.SSD.tau.nojumps.fish,"Tau.abs")

##Stats on tau for for trees not supporting jump in SSD trees in teleosts
Pval.tau.nojump.SSD.fish <- two.tailed.wilcox(duplication.tau.nojump.SSD.fish,speciation.tau.nojump.SSD.fish)

nodes.contrast.SSD.tau.nojumps.fish$Event<-ifelse(nodes.contrast.SSD.tau.nojumps.fish$Event %in% "speciation", "Speciation","SSD")
nodes.contrast.SSD.tau.nojumps.fish$Event<-factor( nodes.contrast.SSD.tau.nojumps.fish$Event, levels=c("Speciation", "SSD"))
levels(nodes.contrast.SSD.tau.nojumps.fish$Event)<-c("Speciation","SSD")

## Median data for trees not supporting jump in SSD trees in teleosts
median.tau.nojump.SSD.teleosts<-NULL
median.tau.nojump.SSD.teleosts<-nodes.contrast.SSD.tau.nojumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.nojump.SSD.teleosts$pic.round<-paste0("Median = ",round(median.tau.nojump.SSD.teleosts$Tau.abs,4), sep="")

median.tau.nojump.SSD.teleosts$count<- paste0(median.tau.nojump.SSD.teleosts$Event,"\n","(n = ",median.tau.nojump.SSD.teleosts$freq,")")


##Testing on average expression levels for trees supporting jumps in SSD trees in teleosts
speciation.exp.jump.SSD.fish <-speciation.contrast(nodes.contrast.SSD.exp.jumps.fish ,"Exp.abs")
duplication.exp.jump.SSD.fish <-duplication.contrast(nodes.contrast.SSD.exp.jumps.fish,"Exp.abs")

##Stats on average expression levels for trees supporting jumps in SSD trees in teleosts
Pval.exp.jump.SSD.fish <- two.tailed.wilcox(duplication.exp.jump.SSD.fish,speciation.exp.jump.SSD.fish)

nodes.contrast.SSD.exp.jumps.fish$Event<-ifelse(nodes.contrast.SSD.exp.jumps.fish$Event %in% "speciation", "Speciation","SSD")
nodes.contrast.SSD.exp.jumps.fish$Event<-factor( nodes.contrast.SSD.exp.jumps.fish$Event, levels=c("Speciation", "SSD"))
levels(nodes.contrast.SSD.exp.jumps.fish$Event)<-c("Speciation","SSD")

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
median.exp.jump.SSD.teleosts<-NULL
median.exp.jump.SSD.teleosts<-nodes.contrast.SSD.exp.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.teleosts$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.teleosts$Exp.abs,4), sep="")
median.exp.jump.SSD.teleosts$count<- paste0(median.exp.jump.SSD.teleosts$Event,"\n","(n = ",median.exp.jump.SSD.teleosts$freq,")")

##Testing on average expression levels after removal of nodes for which daughter branches support jumps in SSD trees in teleosts
speciation.exp.SSD.jump.fish.removing.jumpnode <-speciation.contrast(nodes.contrast.SSD.exp.jumps.fish.removing.jump.node,"Exp.abs")
duplication.exp.SSD.jump.fish.removing.jumpnode <-duplication.contrast(nodes.contrast.SSD.exp.jumps.fish.removing.jump.node,"Exp.abs")

##Stats for trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in SSD trees in teleosts
Pval.exp.SSD.jump.fish.removing.jumpnode <- two.tailed.wilcox(duplication.exp.SSD.jump.fish.removing.jumpnode,speciation.exp.SSD.jump.fish.removing.jumpnode)

nodes.contrast.SSD.exp.jumps.fish.removing.jump.node$Event<-ifelse(nodes.contrast.SSD.exp.jumps.fish.removing.jump.node$Event %in% "speciation", "Speciation","SSD")
nodes.contrast.SSD.exp.jumps.fish.removing.jump.node$Event<-factor( nodes.contrast.SSD.exp.jumps.fish.removing.jump.node$Event, levels=c("Speciation", "SSD"))
levels(nodes.contrast.SSD.exp.jumps.fish.removing.jump.node$Event)<-c("Speciation","SSD")

## Median data of trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in SSD trees in teleosts
median.exp.SSD.removing.jumpnode.teleosts<-NULL
median.exp.SSD.removing.jumpnode.teleosts<-nodes.contrast.SSD.exp.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.SSD.removing.jumpnode.teleosts$pic.round<-paste0("Median = ",round(median.exp.SSD.removing.jumpnode.teleosts$Exp.abs,4), sep="")
median.exp.SSD.removing.jumpnode.teleosts$count<- paste0(median.exp.SSD.removing.jumpnode.teleosts$Event,"\n","(n = ",median.exp.SSD.removing.jumpnode.teleosts$freq,")")

##Testing on average expression levels for trees not supporting jump in SSD trees in teleosts
speciation.exp.SSD.nojump.fish <-speciation.contrast(nodes.contrast.SSD.exp.nojumps.fish ,"Exp.abs")
duplication.exp.SSD.nojump.fish <-duplication.contrast(nodes.contrast.SSD.exp.nojumps.fish,"Exp.abs")

##Stats on average expression levels for for trees not supporting jump in SSD trees in teleosts
Pval.exp.SSD.nojump.fish <- two.tailed.wilcox(duplication.exp.SSD.nojump.fish,speciation.exp.SSD.nojump.fish)

nodes.contrast.SSD.exp.nojumps.fish$Event<-ifelse(nodes.contrast.SSD.exp.nojumps.fish$Event %in% "speciation", "Speciation","SSD")
nodes.contrast.SSD.exp.nojumps.fish$Event<-factor( nodes.contrast.SSD.exp.nojumps.fish$Event, levels=c("Speciation", "SSD"))
levels(nodes.contrast.SSD.exp.nojumps.fish$Event)<-c("Speciation","SSD")

## Median data for trees not supporting jump in average expression levels in SSD trees in teleosts
median.exp.SSD.nojump.teleosts<-NULL
median.exp.SSD.nojump.teleosts<-nodes.contrast.SSD.exp.nojumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.SSD.nojump.teleosts$pic.round<-paste0("Median = ",round(median.exp.SSD.nojump.teleosts$Exp.abs,4), sep="")
median.exp.SSD.nojump.teleosts$count<- paste0(median.exp.SSD.nojump.teleosts$Event,"\n","(n = ",median.exp.SSD.nojump.teleosts$freq,")")

```



```{r Fig_6, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig 6**: Trait jump test of the ortholog conjecture on small-scale duplicate trees in teleosts. We used a strict branch jump posterior cutoff of $\\ge$ 0.7 as an evidence of trait jump. Only teleosts-specific clades were used for the analyses. The ortholog conjecture tests on (A) ", length(Calibrated.SSD.tau.jump), " calibrated contrasts standardized trusted SSD trees supporting rapid jumps in $\\tau$, and on (B) ", length(Calibrated.SSD.exp.jump)," trusted SSD trees supporting jumps in average gene expression levels with the above-mentioned threshold. (C)-(D) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for ", length(Calibrated.SSD.tau.jump), " and ", length(Calibrated.SSD.exp.jump), " trusted SSD trees, respectively. (E)-(F) The ortholog conjecture tests on ",length(Calibrated.SSD.tau.nojump)," and ", length(Calibrated.SSD.exp.nojump)," trusted SSD trees for which the trait jump model is not supported for $\\tau$ and for average gene expression levels by any of their branches, respectively. *P* values are from Wilcoxon two-tailed tests.")}

dodge <- position_dodge(width = 0.31)

# The ortholog conjecture test on tau for trees supporting jump in SSD teleosts 
plot6A<- boxplot.new(nodes.contrast.SSD.tau.jumps.fish,Pval.tau.jump.SSD.fish,"Tau",median.tau.jump.SSD.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("A. OC test on trees supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.jump.SSD.teleosts$count)+       
  add_phylopic(fish_pic, alpha = 1,x =2.3, y = 0.044, ysize = 0.005) 

# The ortholog conjecture test on average expression levels for trees supporting jump in SSD teleosts 

plot6B<- boxplot.new(nodes.contrast.SSD.exp.jumps.fish,Pval.exp.jump.SSD.fish,"Exp",median.exp.jump.SSD.teleosts)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. OC test on trees supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.teleosts$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.27, ysize = 0.03) 

# The ortholog conjecture test on tau after removal of node contrasts for which the daughter branches experience jumps in tau in SSD teleosts

plot6C<- boxplot.new(nodes.contrast.SSD.tau.jumps.fish.removing.jump.node,Pval.tau.jump.SSD.fish.removing.jumpnode,"Tau",median.tau.removing.SSD.jumpnode.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("C.After removing nodes supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.removing.SSD.jumpnode.teleosts$count)+      add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.044, ysize = 0.005) 


# The ortholog conjecture test on average expression levels after removal of node contrasts for which the daughter branches experience jumps in average expression levels in SSD teleosts 

plot6D<- boxplot.new(nodes.contrast.SSD.exp.jumps.fish.removing.jump.node,Pval.exp.SSD.jump.fish.removing.jumpnode,"Exp",median.exp.SSD.removing.jumpnode.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. After removing nodes supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.SSD.removing.jumpnode.teleosts$count)+      add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.27, ysize = 0.03) 

# The ortholog conjecture test on tau for trees do not provide support for jumps in SSD teleosts

plot6E<- boxplot.new(nodes.contrast.SSD.tau.nojumps.fish,Pval.tau.nojump.SSD.fish ,"Tau",median.tau.nojump.SSD.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("E. OC test on trees not supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.nojump.SSD.teleosts$count)+        
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.044, ysize = 0.005) 

# The ortholog conjecture test on average expression levels for trees do not provide support for jumps in SSD teleosts

plot6F<- boxplot.new(nodes.contrast.SSD.exp.nojumps.fish,Pval.exp.SSD.nojump.fish ,"Exp",median.exp.SSD.nojump.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. OC test on trees not supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.SSD.nojump.teleosts$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.27, ysize = 0.03)


gridExtra::grid.arrange(plot6A,plot6B,plot6C,plot6D,plot6E,plot6F,ncol=2, nrow=3)

```


### **Tissue-specific trait jumps** 


```{r Fig_7, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig 7**: Statistics of contrast standardized small-scale duplicate trees supporting jumps in each tissue. Only trusted SSD trees were used. For each tissue, there are more trees supporting trait jumps than not (two-sample test for equality of proportions,$\\chi^2$ test, *P* < 2.2 × 10^-16")}


# Bar graph of brain
dodge <- position_dodge(width = 0.04)
# Create Data
tree.jump.SSD.brain <- data.frame(
  group=c("Supporting trait jumps","Not supporting jump"),
  value=c(1752,969)
)

# Compute the position of labels
tree.jump.SSD.brain <- tree.jump.SSD.brain %>% 
  arrange(group) %>%
  mutate(cumulative = cumsum(value),
         midpoint = cumulative - value / 2,
         label = paste0(round(value / sum(value) * 100, 1), "%")) %>%
         mutate(percent=paste0(prop = round((value /  
        sum(tree.jump.SSD.brain$value) *100),1),"%","\n(",value," trees)", sep="")) 

tree.jump.SSD.brain$group <- factor(tree.jump.SSD.brain$group, levels = c("Supporting trait jumps","Not supporting jump"))

plot7A<-ggplot(tree.jump.SSD.brain, aes(x=group, y=value, fill=group)) +
  geom_bar(stat="identity",position ="dodge")+
  xlab(NULL) +
  ylab(expression(bold("Number of gene trees"))) +
  coord_cartesian(ylim=c(0, 2700)) +
  theme_classic()+
  theme(legend.title=element_blank(),legend.position=c(4.9,4.9)) +
  theme(axis.text.y = element_text(size=10)) +
  theme(axis.text.x = element_text(size=10,face="bold")) +
  ggtitle(expression(bold(paste("A. Statistics for brain"))))+
  theme(plot.title = element_text(face = "bold"))+
  geom_text(aes(label = percent), position = position_stack(vjust = 0.5), size=4,fontface=2)+ 
  scale_fill_manual(values=c("#FFAE6D","#FFD384"))

 # add_phylopic(tetrapods_pic, alpha = 1,x = 0.3, y = 0.044, ysize = 0.1)
  #add_phylopic(fish_pic, alpha = 1,x = 0.4, y = 0.044, ysize = 0.1)
  #gganatogram(data=organPlot$organ[1], fill="black")

## Bar graph of heart

# Create Data
tree.jump.SSD.heart <- data.frame(
  group=c("Supporting trait jumps","Not supporting jump"),
  value=c(2135,841)
)

# Compute the position of labels
tree.jump.SSD.heart <- tree.jump.SSD.heart %>% 
  arrange((factor(group))) %>%
  mutate(cumulative = cumsum(value),
         midpoint = cumulative - value / 2,
         label = paste0(round(value / sum(value) * 100, 1), "%")) %>%
         mutate(percent=paste0(prop = round((value /  
        sum(tree.jump.SSD.heart$value) *100),1),"%","\n(",value," trees)", sep="")) 

tree.jump.SSD.heart$group <- factor(tree.jump.SSD.heart$group, levels = c("Supporting trait jumps","Not supporting jump"))

plot7B<-ggplot(tree.jump.SSD.heart, aes(x=group, y=value, fill=group)) +
  geom_bar(stat="identity", position ="dodge")+
  xlab(NULL) +
  ylab(NULL) +
  coord_cartesian(ylim=c(0, 2700)) +
  theme_classic()+
  theme(legend.title=element_blank(),legend.position=c(4.9,4.9)) +
  theme(axis.text.y = element_text(size=10)) +
  theme(axis.text.x = element_text(size=10,face="bold")) +
  ggtitle(expression(bold(paste("B. Statistics for heart"))))+
  theme(plot.title = element_text(face = "bold"))+
  geom_text(aes(label = percent), position = position_stack(vjust = 0.5), size=4,fontface=2)+ 
  scale_fill_manual(values=c("#FFAE6D","#FFD384"))

## Bar graph of kidney

# Create Data
tree.jump.SSD.kidney <- data.frame(
  group=c("Supporting trait jumps","Not supporting jump"),
  value=c(1788,913)
)

# Compute the position of labels
tree.jump.SSD.kidney <- tree.jump.SSD.kidney %>% 
  arrange((factor(group))) %>%
  mutate(cumulative = cumsum(value),
         midpoint = cumulative - value / 2,
         label = paste0(round(value / sum(value) * 100, 1), "%")) %>%
         mutate(percent=paste0(prop = round((value /  
        sum(tree.jump.SSD.kidney$value) *100),1),"%","\n(",value," trees)", sep="")) 

tree.jump.SSD.kidney$group <- factor(tree.jump.SSD.kidney$group, levels = c("Supporting trait jumps","Not supporting jump"))

plot7C<-ggplot(tree.jump.SSD.kidney, aes(x=group, y=value, fill=group)) +
  geom_bar(stat="identity",position ="dodge")+
  xlab(NULL) +
  ylab(expression(bold("Number of gene trees"))) +
  coord_cartesian(ylim=c(0, 2700)) +
  theme_classic()+
  theme(legend.title=element_blank(),legend.position=c(4.9,4.9)) +
  theme(axis.text.y = element_text(size=10)) +
  theme(axis.text.x = element_text(size=10,face="bold")) +
  ggtitle(expression(bold(paste("C. Statistics for kidney"))))+
  theme(plot.title = element_text(face = "bold"))+
  geom_text(aes(label = percent), position = position_stack(vjust = 0.5), size=4,fontface=2)+ 
  scale_fill_manual(values=c("#FFAE6D","#FFD384"))

## Bar graph of muscle

# Create Data
tree.jump.SSD.muscle <- data.frame(
  group=c("Supporting trait jumps","Not supporting jump"),
  value=c(2319,473)
)

# Compute the position of labels
tree.jump.SSD.muscle <- tree.jump.SSD.muscle %>% 
  arrange((factor(group))) %>%
  mutate(cumulative = cumsum(value),
         midpoint = cumulative - value / 2,
         label = paste0(round(value / sum(value) * 100, 1), "%")) %>%
         mutate(percent=paste0(prop = round((value /  
        sum(tree.jump.SSD.muscle$value) *100),1),"%","\n(",value," trees)", sep="")) 

tree.jump.SSD.muscle$group <- factor(tree.jump.SSD.muscle$group, levels = c("Supporting trait jumps","Not supporting jump"))

plot7D<-ggplot(tree.jump.SSD.muscle, aes(x=group, y=value, fill=group)) +
  geom_bar(stat="identity",position ="dodge")+
  xlab(NULL) +
  ylab(NULL) +
  coord_cartesian(ylim=c(0, 2700)) +
  theme_classic()+
  theme(legend.title=element_blank(),legend.position=c(4.9,4.9)) +
  theme(axis.text.y = element_text(size=10)) +
  theme(axis.text.x = element_text(size=10,face="bold")) +
  ggtitle(expression(bold(paste("D. Statistics for muscle"))))+
  theme(plot.title = element_text(face = "bold"))+
  geom_text(aes(label = percent), position = position_stack(vjust = 0.5), size=4,fontface=2)+ 
  scale_fill_manual(values=c("#FFAE6D","#FFD384"))

## Bar graph of liver
tree.jump.SSD.liver <- data.frame(
  group=c("Supporting trait jumps","Not supporting jump"),
  value=c(2641,841)
)

# Compute the position of labels
tree.jump.SSD.liver <- tree.jump.SSD.liver %>% 
  arrange((factor(group))) %>%
  mutate(cumulative = cumsum(value),
         midpoint = cumulative - value / 2,
         label = paste0(round(value / sum(value) * 100, 1), "%")) %>%
         mutate(percent=paste0(prop = round((value /  
        sum(tree.jump.SSD.liver$value) *100),1),"%","\n(",value," trees)", sep="")) 

tree.jump.SSD.liver$group <- factor(tree.jump.SSD.liver$group, levels = c("Supporting trait jumps","Not supporting jump"))

plot7E<-ggplot(tree.jump.SSD.liver, aes(x=group, y=value, fill=group)) +
  geom_bar(stat="identity",position ="dodge")+
  xlab(NULL) +
  ylab(expression(bold("Number of gene trees"))) +
  coord_cartesian(ylim=c(0, 2700)) +
  theme_classic()+
  theme(legend.title=element_blank(),legend.position=c(4.9,4.9)) +
  theme(axis.text.y = element_text(size=10)) +
  theme(axis.text.x = element_text(size=10,face="bold")) +
  ggtitle(expression(bold(paste("E. Statistics for liver"))))+
  theme(plot.title = element_text(face = "bold"))+
  geom_text(aes(label = percent), position = position_stack(vjust = 0.5), size=4,fontface=2)+ 
  scale_fill_manual(values=c("#FFAE6D","#FFD384"))

## Bar graph of testis
# Create Data
tree.jump.SSD.testis <- data.frame(
  group=c("Supporting trait jumps","Not supporting jump"),
  value=c(1441,831)
)

# Compute the position of labels
tree.jump.SSD.testis <- tree.jump.SSD.testis %>% 
  arrange((factor(group))) %>%
  mutate(cumulative = cumsum(value),
         midpoint = cumulative - value / 2,
         label = paste0(round(value / sum(value) * 100, 1), "%")) %>%
         mutate(percent=paste0(prop = round((value /  
        sum(tree.jump.SSD.testis$value) *100),1),"%","\n(",value," trees)", sep="")) 

tree.jump.SSD.testis$group <- factor(tree.jump.SSD.testis$group, levels = c("Supporting trait jumps","Not supporting jump"))

plot7F<-ggplot(tree.jump.SSD.testis, aes(x=group, y=value, fill=group)) +
  geom_bar(stat="identity", position ="dodge")+
  xlab(NULL) +
  ylab(NULL) +
  coord_cartesian(ylim=c(0, 2700)) +
  theme_classic()+
  theme(legend.title=element_blank(),legend.position=c(4.9,4.9)) +
  theme(axis.text.y = element_text(size=10)) +
  theme(axis.text.x = element_text(size=10,face="bold")) +
  ggtitle(expression(bold(paste("F. Statistics for testis"))))+
  theme(plot.title = element_text(face = "bold"))+
  geom_text(aes(label = percent), position = position_stack(vjust = 0.5), size=4,fontface=2)+ 
  scale_fill_manual(values=c("#FFAE6D","#FFD384"))
  #scale_fill_manual(values=c("#7978FF", "#AFB4FF"))

 
gridExtra::grid.arrange(plot7A,plot7B,plot7C,plot7D,plot7E,plot7F,ncol=2, nrow=3)

```


We next performed our analyses separately on each tissue to investigate whether jumps in expression following duplication were tissue-specific or shared between tissues. Since the trait jump model was supported for SSD and speciation branches, but not genome duplication branches, in the global analyses (Figs. 5 and 6; Supplementary figs. S9-S13), we restricted the tissue-level analysis to trusted SSD trees. Most of these trees supported the trait jump model for all 6 tissues (Fig. 7).    

\vspace{0.5cm}
When we applied our PICs on the trusted SSD phylogenies supporting jumps in their traits for individual tissues, we observed a strong support for the ortholog conjecture in brain, kidney and testis (Supplementary figs. S14-S19). There is also support for the ortholog conjecture in heart, but only in the teleost-only analysis (Supplementary fig. S15). Removal of nodes for which one or of both the daughter branches experiences rapid shifts in their traits removes the signal of the ortholog conjecture for these four tissues, including teleost-only heart (Supplementary figs. S14C-S14F, S15C, S15F, S16C-S16F, S19C-S19F).    


### **Relation between coding sequence positive selection and evolutionary shifts in expression**

To investigate a potential role of positive selection on genes which shift expression after duplication, we turned to the detection of positively selected sites in the protein coding sequence. We aim to test whether positive selection on the protein is associated with rapid shifts in expression of small-scale duplicates.  

\vspace{0.5cm}
Parsing the positive selection data for vertebrates from the Selectome database [@Moretti2014], and applying a false-discovery-rate correction  [@Storey2003; @Anisimova2007; @Studer2009] cutoff of q < 0.10, we found less than 100 cases with both protein-level positive selection and expression jumps. To gain power, we used the sum of likelihood ratios (Daub et al. 2017), where higher likelihood ratio ($\Delta$lnL)score indicates more evidence for positive selection. Branches supporting rapid shifts in expression traits have indeed higher $\Delta$lnL scores for paralogs than for orthologs, thus more evidence for positive selection (Supplementary figs. S20A and S20B). While the highest $\Delta$lnL scores were for paralogs with expression shifts, removal of branches supporting rapid jumps in traits still showed higher $\Delta$lnL   
scores for paralogs (Supplementary figs. S20C and S20D). Our trends remain mostly unchanged when we repeated our analyses for teleosts only (Supplementary fig. S21), with smaller sample sizes. Thus there is overall more positive selection on protein-coding sequences after duplication than speciation, and even more for genes which shift in expression. This supports the ortholog conjecture regarding protein sequence evolutionary rates, as observed by David et al. [@David2020] and Fukushima and Pollock [@Fukushima2020].  


\pagebreak

### **Discussion**  

Understanding the dynamics of gene function change after duplication is important both to our fundamental knowledge of gene evolution, and to the practical use of the ortholog conjecture [@Koonin2005; @Studer2009; @Nehrt2011; @Altenhoff2012; @Chen2012; @Gabaldon2013; @Rogozin2014; @Kryuchkova-Mostacci2016; @Dunn2018; @Fukushima2020; @Stamboulian2020; @Begum2021a]. Indeed, majority of proteins functions are not experimentally determined, but rather are extrapolated from homologs, often specifically from orthologs. Although the ortholog conjecture remains debated [@Nehrt2011; @Dunn2018; @Stamboulian2020; @Begum2021a]), we previously found some support for the conjecture for a single trait, tissue specificity $\tau$,
from eight vertebrate species using a properly controlled phylogenetic independent contrasts method, as well as using phylogenetic modeling [@Begum2021a].In that study we did not distinguish types of duplication, and only tested for branch-specific changes in gradual rates.  

\vspace{0.5cm}
Here we provide strong confirmation for the ortholog conjecture on gene expression in vertebrates: it holds both for average expression and for tissue-specificity; it holds with an improved phylogenetic independent contrast which accounts for the long-term impact of duplication on the next speciation branch; and it holds for both small-scale and whole genome duplicates. Our results are in agreement with the results of pairwise-based analyses [@Altenhoff2012; @Chen2012; @Gabaldon2013; @Rogozin2014; @Kryuchkova-Mostacci2016]. While the usual interpretation of the orthologue conjecture has considered a change in sequence or trait evolutionary rates over a phylogenetic branch, another model could be a sudden change of trait values without altering the evolutionary rates [@Begum2021b]. And indeed, we find evidence for such a rapid trait jump model for evolution after duplication.    

\vspace{0.5cm}
Evolutionary jumps have been shown to affect the evolution of traits in adaptive landscapes [@Simpson1944; @Bokma2008; @Landis2013; @Duchen2017; @Duchen2021; @Gao2022]), but have been rarely applied to the evolution of gene function. Yet small-scale gene duplication can be a rapid and potentially asymmetric phenomenon [@Kaessmann2009]. Indeed, we find evidence for the existence of a trait jump mechanism for small-scale duplicates. This can correspond to a sudden change of function, e.g., from incomplete duplication of the coding portion of the progenitor locus or of its regulatory regions, fusion of incomplete segmental duplications, or a change in chromatin environment [@Lynch2004; @Dennis2012; @Dougherty2017; @Guschanski2017]. Paralogs can differ in structure from their progenitor copy [@Dougherty2017], and can acquire novel promoters (e.g., gene fusion), or partially or completely lose regulatory elements (e.g., RNA-mediated duplication, [@Kaessmann2009]). These processes can lead to rapid functional changes, notably through altered expression patterns. These mechanisms rapidly change the trait values rather than changing their evolutionary rates over long evolutionary time. Tissue-specific support for trait jumps after duplication appears to differ between organs, with strongest support in brain, testis and kidney (Figs S14-S19). On the other hand, genome duplication does not seem to lead to sudden or very asymmetric changes in gene expression, which is consistent with the short-term conservation of regulatory environment (Figs. 3 and 4, Supplementary figs. S2 and S7). Interestingly, even for small-scale duplicates, those which do not fit a jump model still support the ortholog conjecture, i.e. with a change in gradual evolutionary rate. Thus the two models of divergence after duplication co-exist: rapid jumps in function after some small-scale duplications, and gradual changes after whole genome and other small-scale duplications.  

\vspace{0.5cm}
Positive selection on the coding sequence can contribute to the evolution of gene function following duplication [Han et al. 2009; @Pegueroles2013]. We found indeed in general higher evidence for positive selection after duplication. We also find slightly more evidence for positive selection on the branches supporting rapid jumps in their traits following duplication than speciation, but tests aren't significant. This might be in part due to a lack of power, as sample sizes for genes with evidence of both processes are small. Thus we cannot exclude a relation, but if there is one it is weak and would need larger sample sizes to detect reliably.  


\vspace{0.5cm}
In conclusion, we find strong support for rapid evolution of gene expression after duplication, both by sudden jumps and by long-term acceleration of evolutionary rates. Processes differ between small-scale and genome-scale duplications, but in all cases support the ortholog conjecture, i.e. stronger divergence of paralogs than of orthologs. Thus duplication is critical to the evolution of genome function, and is best studied with a careful use of phylogenetic methods.   


\pagebreak
### Materials and Methods

__*Processing gene trees*__

`r length(exp.annotated.trees)` empirical gene trees were parsed from Genomicus v.95.01 [@Nguyen2022]. This version was chosen to avoid the division of large trees into subtrees from Ensembl in later versions (Martin et al. 2023). Trees were processed to annotate "speciation", and "duplication" to the internal node events using available information from the tree data object ("D" = 0 as "speciation" and "D" = "Y" as "duplication"). We estimated tissue specificity data, $\tau$, and average gene expression levels from ten tetrapods (human, mouse, rat, macaque, cow, dog, ferret, opossum, rabbit, and chicken) and from five teleost fishes (zebrafish, medaka, cavefish, northern pike, and tilapia) across six common tissues (brain, heart, kidney, liver, muscle, and testis), and added those estimates to the tree data objects according to their corresponding Ensembl gene ID. After pruning the gene trees to remove tips with missing trait estimates we obtained `r length(pruned.annotated.trees)` trees. We excluded `r length(trees.all.duplication)` gene trees with only duplication nodes and `r length(trees.all.speciation.new)` gene trees with no duplication nodes, and retained `r length(trees.all.pruned)` trees with at least one speciation and one duplication node for further processing. Cross-species normalized gene expression levels (see below) were then added to the tree data object, and all `r length(norm.exp.annotated.pruned.trees)` trees were retained after filtering for the eventuality of trees with genes that were not expressed in any of the six tissues or which had less than four tips.  

\vspace{0.5cm}
The gene trees were then time calibrated using speciation clade ages from Timetree [@Hedges2006]. For time calibration, we used the "correlated" model of chronos() function from the R "ape" package [@Paradis2004; @Dunn2018; @Begum2021a]. Since the focus of the study was teleost species, we considered tetrapods as the outgroups. We, therefore, excluded trees with no teleost tip data for our analyses. We thus selected `r length(trees.all.interest)` trees trees for which at least three tips belonged to teleosts, and at least one tip contained trait data for an outgroup tetrapod species.       

\vspace{0.5cm}
__*Estimation of phylogenetic independent contrasts (PICs)*__ 

For each individual trait, phylogenetic independent contrast or PIC was estimated for each internal node of a gene tree using the pic() function of the R "ape" package [@Paradis2004]. PIC of a parental node is an absolute ratio of changes in trait values for descendant nodes to the lengths of its two descendant branches, *i.e.*, the expected variance [@Felsenstein1985a; @Dunn2018; @Begum2021a], in million years (My).

\vspace{0.5cm}
__*Classification of gene trees based on types of duplications*__  

We identified synteny validated teleost-specific third round (relative to previous two rounds at the base of vertebrates) whole genome duplicates (FishWGDs), or ohnologs, from the database OHNOLOGS v.2 [@Singh2020]. Out of five teleosts, we obtained 3R ohnolog pairs only for two (*e.g.*, zebrafish and medaka) from this database. We applied relaxed criteria (q-score(outgroups) < 0.05 and q-score(self comparison) < 0.3) as available in OHNOLOGS database [@Singh2020] to obtain a large dataset of ohnologs. Only "protein_coding" genes with duplication time annotated as "FishWGD" were used. We annotated the most recent common ancestral node of the identified ohnolog pairs in each gene tree as "FishWGD". Other duplication nodes in the same trees were annotated as small-scale gene duplication or "SSD".  

\vspace{0.5cm}
The teleost fish underwent a third round of whole genome duplication at the base of the Osteoglossocephalai or the Clupeocephala clade, which has been dated 225 $-$ 333 My ago [@Berthelot2014; @Parey2020; @Parey2022a]. We thus considered "FishWGD" annotated to Osteoglossocephalai or Clupeocephala clade descendants as "confident" FishWGD ohnologs. We called the `r length(Sure.WGD.trees)` phylogenies with such confident ohnologs "trusted WGD" trees. Other ohnolog-annotated genes of the OHNOLOGS database were classified as "dubious 3RWGDs" when their duplication time did not match with "FishWGD" or when the "FishWGD" annotation matched with taxonomic levels other than Osteoglossocephalai or Clupeocephala.  We thus obtained `r length(Unsure.WGD.trees)` "dubious WGD" trees.  

To identify small scale duplication (SSD) trees, we first excluded the (`r length(Sure.WGD.trees)` + `r length(Unsure.WGD.trees)`) trusted or dubious WGD gene trees. From the remaining 5136 gene trees, we then eliminated phylogenies where duplication was present at the Clupeocephala or Osteoglossocephalai taxonomic level. We called those `r length(sure.SSD.trees)` trees "trusted SSD" trees. The other  `r length(unsure.SSD.trees)` phylogenies were called "dubious SSD" trees, where duplication occurred at the same Clupeocephala or Osteoglossocephalai clade as the 3R WGD. To perform our phylogenetic analysis on dubious SSD trees, we removed contrasts (PICs) of those Clupeocephala and Osteoglossocephalai clades.  

\vspace{0.5cm}
__*Selection of contrasts standardized trees*__  

Since the performance of the PIC method primarily depends on the assumption of Brownian model (BM) of trait evolution, we need to identify phylogenies for which contrasts (PICs) are adequately standardized according to BM  [@Felsenstein1985a; @Garland1992b; @Diaz-Uriarte1996; @Diaz-Uriarte1998; @Freckleton2006; @Cooper2016a; @Begum2021a]. Node contrasts for those trees should be independent of internal phylogenetic parameters, *i.e.*, logarithm of node age, expected standard deviations and node depth in this case. We used diagnostic tests on phylogenies with duplications to draw unbiased inference by PIC method, as recommended earlier [@Garland1992b; @Diaz-Uriarte1996; @Diaz-Uriarte1998; @Freckleton2006; @Cooper2016a; @Begum2021a]. In these tests, we perform correlation between the absolute values of PICs and each of the three internal phylogenetic parameters individually for each trait per phylogeny. Any significant correlation (*P* < 0.05) indicates departure from BM, *i.e.*, phylogenetic dependence for that tree. By eliminating trees showing a significant trend (positive or negative), we considered sets of trees passing all three diagnostic tests for further analyses for each trait of our interest. We estimated node age and node depth using branching.times() and node.depth() functions from the R "ape" package, respectively, for each tree [@Paradis2004]. We applied crunch() function of the caper package to perform diagnostic tests [@Purvis1995; @Cooper2016a; @Orme2018; @Begum2021a].   

\vspace{0.5cm}
Following this, `r length(Calibrated.3R.tau)` trusted FishWGD trees and `r length(Calibrated.SSD.tau)` trusted SSD trees passed diagnostic tests for  $\tau$, and `r length(Calibrated.3R.exp)` trusted FishWGD trees and `r length(Calibrated.SSD.exp)` trusted SSD trees passed diagnostic tests for average gene expression levels. Overall, we obtained `r length(Calibrated.all.standardized.tau)` and `r length(Calibrated.all.standardized.exp)` trees passing diagnostic tests for $\tau$ and and average gene expression levels, respectively, including “dubious” FishWGD or SSD trees.    

\vspace{0.5cm}
__*Randomization test*__  

To confirm biological effects vs. biases in data structure, we used randomization test of trait values on different sets of contrasts standardized trees, following Begum and Robinson-Rechavi (2021). To randomize, we permuted actual trait values of the tips without changing internal node events for each tree. After randomization, we recalculated the node contrasts (PICs) for each tree using permuted trait values of tips following above stated procedure.  

\vspace{0.5cm}
__*Expression data*__

We retrieved processed adult stage paired-end RNA-seq data for seven tetrapods (human: GSE30611, mouse: GSE41367, rat: GSE41367, macaque: GSE41367, cow: GSE41367, opossum: GSE30352, and chicken: GSE41367) and one teleost fish (zebrafish: SRP044781) from Bgee database v.14 [@Bastian2021]. Pre-processed expression data for three other tetrapods (dog, ferret, and rabbit) submitted to the NCBI Gene Expression Omnibus (GEO; http://www.ncbi.nlm.nih.gov/geo/) by Chen et al. [@Chen2019] under accession number GSE106077 were directly downloaded from https://portals.broadinstitute.org/evee/. Paired-end Illumina RNA-seq reads of three teleosts (medaka, cavefish, and northern pike) were retrieved from the PhyloFish database [@Pasquier2016]. All the tissue data of PhyloFish came from mature stage, except for testis, which came either from adult, or from unknown stage. We fetched paired-end RNA-seq raw reads for tilapia fish directly from NCBI using accession number SRR391690. Developmental stage information was unavailable for dog, ferret, rabbit, and tilapia.    

\vspace{0.5cm}
Quality assessment of raw and trimmed reads was conducted using FastQC v.0.11.5 [@Andrew2019]. Quality filtration was performed using Trimmomatic v.0.36 [@Bolger2014] to discard adaptor sequence or low-quality reads (settings: LEADING:4 TRAILING:4 SLIDINGWINDOW:4:20 HEADCROP:12 MINLEN:25). Transcript information was downloaded from Ensemble v.95 (Martin et al. 2023) and was used as reference for fast and accurate mapping against the cleaned reads with the aid of kallisto v.0.43.1 [@Bray2016] pseudoaligner. For rapid RNA-seq quantification, we applied quant argument of kallisto.  

\vspace{0.5cm}
We used transcript abundances in transcripts per million as an estimate of gene expression levels, with log transformation ($log{_2}{TPM}$). In case of tilapia, we had three biological replicates for each tissue sample. We took average TPM for all the replicates against each tissue to obtain per tissue gene expression level for each gene. TPM values from several sub parts of the tissues for tetrapods from Bgee database [@Bastian2021] are averaged to obtain gene expression levels of the main organ or tissue. Teleosts specifically have a organ called "head kidney", which comprises about 20% of the anterior part of the kidney. Considering it as "kidney", we obtained six common tissues (brain, heart, kidney, liver, muscle, and testis) for all the vertebrates of this study.

\vspace{0.5cm}
__*Normalization of gene expression data*__

For a few species of this study, the developmental stages of tissues were unknown (see above). Most PhyloFish samples, except testis, were extracted from females [@Pasquier2016]. On the other hand, data from Bgee [@Bastian2021] sometimes came from male, sometimes from female, or also from mixed sex in some cases (e.g., heart and kidney expression of zebrafish). In addition, these data come from different studies and different species. Thus we needed to normalize gene expression levels to compare them across species and tissues.  

For cross-species expression data normalization, we adopted a scaling procedure following Brawand *et al.* [@Brawand2011]. Among the `r nrow(pca.ortholog)` 1:1 orthologs of all vertebrate genes with their expression values in the interquartile range, we identified the top 500 conserved genes that have the maximum conserved ranks among tissues. Median expression levels for those 500 genes were assessed in each tissue for each species. We then derived scaling factors for each tissue of a species using median of human tissues as references. Finally, we applied those scaling factors on the log transformed corresponding tissue dataset of 15 species to obtain normalized gene expression level for each gene of a species in that tissue.  

\vspace{0.5cm}
__*Estimation of trait data*__

Tissue specificity, $\tau$, of a gene was computed across all the six tissues of each species using the following equation [@Yanai2005]:

   $\tau$ = $\dfrac{\sum_{i=1}^{N}(1-x_i)}{N-1}$, 
  
where N is the total number of tissues or organs and $x_i$ is the $log{_2}{TPM}$ value of gene x in i^th^ tissue. $\tau$ close to 1 indicates higher tissue specific function of a gene, while more ubiquitous expression is denoted by values close to 0 [@Yanai2005; @Kryuchkova-Mostacci2016].  

\vspace{0.5cm}
Average gene expression level was estimated by taking the mean of cross species scaling factor normalized $log{_2}{TPM}$ values across all the six tissues for each gene per species.  

\vspace{0.5cm}
__*Estimation of rapid trait jump*__

Evolutionary jumps were estimated with the software "levolution" [@Duchen2017]. The input consists of an ultrametric tree in Newick format, and a two-column file with trait data for each tip of the tree. In our case the trees used were the gene trees described above, and the trait data were the normalized gene expression for the different tissues, as well as for the $\tau$ and average gene expression levels. “levolution” was then run separately for individual trait on contrast standardized gene trees.    

\vspace{0.5cm}
Briefly, "levolution" works by calculating the variance-covariance matrix of the traits, and assigning a starting vector of jump configurations at every branch of the tree. Through a Marcov Chain Monte Carlo (MCMC) method, jump configurations are then updated one branch at the time with fixed hierarchical parameters, such as the jump rate, Brownian rate, jump strength, and root state. Two steps take place during the MCMC run. First, an Expectation-Maximization approach is used to infer maximum likelihood (ML) estimates of these hierarchical parameters. Then, the position of jumps along the tree is inferred by using an empirical Bayes approach, where the hierarchical parameters are fixed to their ML estimate, and the MCMC is then used to calculate posterior probabilities of jumps at a particular branch [@Duchen2017].    

\vspace{0.5cm}
Mathematically, this jump model is depicted as follows. Given ***x*** as the set of traits at the tips of the tree, the likelihood of x given the hierarchical parameters is f(x $\mid$ hierarchical parameters) $=\sum_{n=1}^{\infty} P(\nu=n)\phi(x\mid n)$, where ***n*** is the number of jumps, $P(\nu=n)$ is the Poisson probability of observing ***n*** jumps, and $\phi(x\mid n)$ is the multivariate normal probability of observing traits ***x*** given that ***n*** jumps are present (Duchen et al. 2017, Eq. 4). The sum over all possible jump configurations is taken over by the MCMC approach described above, and the entries in the variance covariance matrix in $\phi(x\mid n)$ are given by the sum of the variance present along each branch and the variance due to the jumps.
 
\vspace{0.5cm}
__*Analyses of positive selection *__
We retrieved precomputed statistics on positive selection from Selectome [@Moretti2014], a database that applies the branch-site likelihood test for selection to Ensembl gene trees (release 98, Martin et al. 2023). Each processed Genomicus tree (See ‘Processing gene trees’) was mapped, if possible, to the corresponding Selectome tree by Ensembl gene IDs. Of 6923 processed Genomicus trees, 6649 trees were successfully mapped to their corresponding tree in Selectome. Because the Ensembl-based Selectome trees sample more taxa than we have included in our study, we retrieved the likelihood-ratio test statistic only for nodes that have matching taxonomic labels between each mapped Selectome and Genomicus tree, and excluded taxa that are not contained in the former.  

\vspace{0.5cm}
There were 116 (respectively 402) trusted SSD trees supporting both positive selection as well as jumps in $\tau$ (respectively in average expression levels), in their branches for the 15 vertebrates. For teleost-specific analyses, there were only 31 and 78 such trees.


\vspace{0.5cm}
__*Script and data availability*__  
All the scripts and gene expression data files are available on GitHub: https://github.com/tbegum/Phylogenetic-modeling-of-evolutionary-trait-jumps. For reproducibility, “norm_exp_15spe_4tips_new.RData” file is archived at Zenodo (https://doi.org/10.5281/zenodo.10418895).  

\vspace{0.5cm}
__*Details of other R packages used*__

We performed other analyses and plotting in R version 4.0.2 [@Rcore2018] using rphylopic [@Chamberlain2018], phytools [@Revell2012], treeio [@Guangchuang2018], ggtree [@Guangchuang2017], digest [@digest2018], dplyr [@dplyr2017], ggplot2 [@ggplot2016]; gtools [@gtools2018], gridExtra [@gridExtra2017], tidyverse [@tidyverse2017], stringr [@Stringr2019], and png [@png2013] libraries.
 

## Acknowledgements

We sincerely thank all the members of the Robinson-Rechavi and Salamin groups for their help and useful discussions. Parts of the computations were performed at the Vital-IT Center for high-performance computing of the SIB Swiss Institute of Bioinformatics, as well as using the Wally cluster of the University of Lausanne. Work supported by Swiss National Science Foundation grant 31003A_173048. 


\pagebreak  

### **Supporting Information**   

```{r Fig_S1P, eval=F, dpi=400,fig.width=7, fig.height=8, fig.cap= paste0("**Fig S1 backup**: Principal component analysis (PCA) plots. (A) This plot uses normalized expression levels of ",nrow(pca.ortholog), " 1:1 orthologous gene set from 15 species. (B) We consider trusted ohnologs or 3R whole genome duplicates (FishWGDs) from 5 teleosts in this plot. Normalized gene expression levels of 300 ohnolog set are used here. Duplicates in each teleost fish are classified based on their expression levels as 'highly expressed' and 'lowly expressed' ohnologs. Tissue colors for the highly expressed ohnolog copy and the lowly expressed ohnolog copy for a species are labeled as 'High exp' and 'Low exp'.  The plot shows that highly expressed ohnologs cluster separately from their lowly expressed copy for each tissue. Percentage of variations are indicated in each axis. PC1: Principle component 1, PC2: Principle component 2.")}

source("/Users/admin/Desktop/nwork/Empirical/Manuscript_3/functions_TPCM.R")

PCA.plot(exp.normalized.ortholog,type="ortholog",group.by="tissue",method="normal.2D")+ title(main= "A. PCA plot for single copy orthologs",font.main=2,cex.main=1,outer = "F", adj=0) ## orthologs after normalization

## Saved plot as "FigS1A.png"

PCA.plot(pca.WGD1,type="common.WGD",group.by="tissue",method="normal.2D") + title(main= "B. PCA plot for FishWGDs",font.main=2,cex.main=1,outer = "F",adj=0)

## Saved plot as "FigS1B.png"

```


```{r Fig_S1, eval=T, dpi=400,fig.width=7, fig.height=8, fig.cap= paste0("**Fig S1**: Principal component analysis (PCA) plots. (A) This plot uses normalized expression levels of ",nrow(pca.ortholog), " 1:1 orthologous gene set from 15 species. (B) We consider trusted ohnologs or 3R whole genome duplicates (FishWGDs) from 5 teleosts in this plot. Normalized gene expression levels of 300 ohnolog set are used here. Duplicates in each teleost fish are classified based on their expression levels as 'highly expressed' and 'lowly expressed' ohnologs. Tissue colors for the highly expressed ohnolog copy and the lowly expressed ohnolog copy for a species are labeled as 'High exp' and 'Low exp'.  The plot shows that highly expressed ohnologs cluster separately from their lowly expressed copy for each tissue. Percentage of variations are indicated in each axis. PC1: Principle component 1, PC2: Principle component 2.")}

source("/Users/admin/Desktop/nwork/Empirical/Manuscript_3/functions_TPCM.R")

plotS1AA <- png::readPNG("~/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/Figures/FigS1A.png")
plotS1A <-grid::rasterGrob(plotS1AA)

plotS1BB <- png::readPNG("~/Desktop/nwork/Empirical/Manuscript_3/Plot_Manuscipt3/Figures/FigS1B.png")
plotS1B <-grid::rasterGrob(plotS1BB)

grid.arrange(plotS1A,plotS1B,ncol=1,nrow=2)

```



```{r Analyses_on_sure_3RWGD_trees_vertebrates, echo=F, message=FALSE, warning=FALSE, cache=T}

##Testing on tau for 3RWGDs
speciation.picTau.3RWGDs.all <- speciation.contrast(nodes.contrast.tau.3RWGDs ,"Tau.abs")
duplication.picTau.3RWGDs.all1 <- duplication.contrast.3R(nodes.contrast.tau.3RWGDs,"Tau.abs")
duplication.picTau.3RWGDs.all2 <- duplication.contrast(nodes.contrast.tau.3RWGDs,"Tau.abs")
Pval.Tau.3RWGDs.vertebrates1<- two.tailed.wilcox(duplication.picTau.3RWGDs.all1,speciation.picTau.3RWGDs.all)
Pval.Tau.3RWGDs.vertebrates2 <- two.tailed.wilcox(duplication.picTau.3RWGDs.all2,speciation.picTau.3RWGDs.all)


## Median data of tau for vertebrates specific 3RWGDs
levels(nodes.contrast.tau.3RWGDs$Event)<-c("Speciation", "SSD", "FishWGD")
median.Tau.3RWGDs.vertebrates<-NULL
median.Tau.3RWGDs.vertebrates<-nodes.contrast.tau.3RWGDs %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.Tau.3RWGDs.vertebrates$pic.round<-paste0("Median = ",round(median.Tau.3RWGDs.vertebrates$Tau.abs,4), sep="")
median.Tau.3RWGDs.vertebrates$count<- paste0(median.Tau.3RWGDs.vertebrates$Event,"\n","(n = ",median.Tau.3RWGDs.vertebrates$freq,")")

##Testing on average expression levels for 3RWGDs
speciation.picExp.3RWGDs.all <-speciation.contrast(nodes.contrast.exp.3RWGDs ,"Exp.abs")
duplication.picExp.3RWGDs.all1 <-duplication.contrast.3R(nodes.contrast.exp.3RWGDs,"Exp.abs")
duplication.picExp.3RWGDs.all2 <-duplication.contrast(nodes.contrast.exp.3RWGDs,"Exp.abs")
Pval.Exp.3RWGDs.vertebrates1 <- two.tailed.wilcox(duplication.picExp.3RWGDs.all1,speciation.picExp.3RWGDs.all)
Pval.Exp.3RWGDs.vertebrates2 <- two.tailed.wilcox(duplication.picExp.3RWGDs.all2,speciation.picExp.3RWGDs.all)

## Median data of exp levels for teleosts specific 3RWGDs
levels(nodes.contrast.exp.3RWGDs$Event)<-c("Speciation", "SSD", "FishWGD")
median.Exp.3RWGDs.vertebrates<-NULL
median.Exp.3RWGDs.vertebrates<-nodes.contrast.exp.3RWGDs %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.Exp.3RWGDs.vertebrates$pic.round<-paste0("Median = ",round(median.Exp.3RWGDs.vertebrates$Exp.abs,4), sep="")
median.Exp.3RWGDs.vertebrates$count<- paste0(median.Exp.3RWGDs.vertebrates$Event,"\n","(n = ",median.Exp.3RWGDs.vertebrates$freq,")")

```


```{r Analyses_on_sure_SSD_trees_vertebrates, echo=F, message=FALSE, warning=FALSE, cache=T}

##Testing on tau for SSDs
speciation.picTau.SSDs.all <-speciation.contrast(nodes.contrast.tau.SSDs ,"Tau.abs")
duplication.picTau.SSDs.all <-duplication.contrast(nodes.contrast.tau.SSDs,"Tau.abs")
Pval.Tau.SSDs.vertebrates <- two.tailed.wilcox(duplication.picTau.SSDs.all,speciation.picTau.SSDs.all)

## Median data of tau for teleosts specific SSDs
levels(nodes.contrast.tau.SSDs$Event)<-c("Speciation", "SSD")
median.Tau.SSDs.vertebrates<-NULL
median.Tau.SSDs.vertebrates<-nodes.contrast.tau.SSDs %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.Tau.SSDs.vertebrates$pic.round<-paste0("Median = ",round(median.Tau.SSDs.vertebrates$Tau.abs,4), sep="")
median.Tau.SSDs.vertebrates$count<- paste0(median.Tau.SSDs.vertebrates$Event,"\n","(n = ",median.Tau.SSDs.vertebrates$freq,")")

##Testing on average expression levels for SSDs
speciation.picExp.SSDs.all <-speciation.contrast(nodes.contrast.exp.SSDs ,"Exp.abs")
duplication.picExp.SSDs.all <-duplication.contrast(nodes.contrast.exp.SSDs,"Exp.abs")
Pval.Exp.SSDs.vertebrates <- two.tailed.wilcox(duplication.picExp.SSDs.all,speciation.picExp.SSDs.all)

## Median data of exp levels for teleosts specific SSDs
levels(nodes.contrast.exp.SSDs$Event)<-c("Speciation", "SSD")
median.Exp.SSDs.vertebrates<-NULL
median.Exp.SSDs.vertebrates<-nodes.contrast.exp.SSDs %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.Exp.SSDs.vertebrates$pic.round<-paste0("Median = ",round(median.Exp.SSDs.vertebrates$Exp.abs,4), sep="")
median.Exp.SSDs.vertebrates$count<- paste0(median.Exp.SSDs.vertebrates$Event,"\n","(n = ",median.Exp.SSDs.vertebrates$freq,")")

```


```{r Fig_S2, eval=T,dpi=400,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S2**: The ortholog conjecture (OC) tests using calibrated contrasts standardized trees in 15 vertebrates. ", length(Calibrated.3R.tau), " and ", length(Calibrated.3R.exp), " trees with whole genome duplications were used in (A)-(B), while ",  length(Calibrated.SSD.tau), " and ", length(Calibrated.SSD.exp), " trees with small-scale gene duplications were used in (C)-(D). PICs: Phylogenetic independent contrasts, 3RWGDs or FishWGD: Fish specific 3R whole genome duplication, 3RWGDs: ohnologs, SSDs: small-scale duplicates. Duplications in (A)-(B) refers to SSDs like (C)-(D). *P* values are from Wilcoxon two-tailed tests.")}

## Plotting 
dodge <- position_dodge(width = 0.05)

plotS2A<- boxplot.new.3RWGD(nodes.contrast.tau.3RWGDs,Pval.Tau.3RWGDs.vertebrates1,Pval.Tau.3RWGDs.vertebrates2,"Tau.3RWGD",median.Tau.3RWGDs.vertebrates) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("A. Test on ", tau, ", for FishWGDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.Tau.3RWGDs.vertebrates$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.043, ysize = 0.002)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.037, ysize = 0.003)
  #add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.044, ysize = 0.1)+
  #add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.038, ysize = 0.15)

plotS2B<- boxplot.new.3RWGD(nodes.contrast.exp.3RWGDs,Pval.Exp.3RWGDs.vertebrates1,Pval.Exp.3RWGDs.vertebrates2,"Exp.3RWGD",median.Exp.3RWGDs.vertebrates) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. Test on exp levels for FishWGDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.Exp.3RWGDs.vertebrates$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.26, ysize = 0.013)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.019)
  #add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.26, ysize = 0.1)+
  #add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.15)

plotS2C<- boxplot.new(nodes.contrast.tau.SSDs,Pval.Tau.SSDs.vertebrates,"Tau",median.Tau.SSDs.vertebrates) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("C. Test on ", tau, ", for SSDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.Tau.SSDs.vertebrates$count)+
  #add_phylopic(tetrapods_pic, alpha = 1,x = 2.4, y = 0.044, ysize = 0.07)+
  #add_phylopic(fish_pic, alpha = 1,x = 2.4, y = 0.038, ysize = 0.11)
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.4, y = 0.043, ysize = 0.002)+
  add_phylopic(fish_pic, alpha = 1,x = 2.4, y = 0.037, ysize = 0.003)

plotS2D<- boxplot.new(nodes.contrast.exp.SSDs,Pval.Exp.SSDs.vertebrates,"Exp",median.Exp.SSDs.vertebrates) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Test on exp levels for SSDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.Exp.SSDs.vertebrates$count)+
  #add_phylopic(tetrapods_pic, alpha = 1,x = 2.4, y = 0.26, ysize = 0.07)+
  #add_phylopic(fish_pic, alpha = 1,x = 2.4, y = 0.22, ysize = 0.11)
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.4, y = 0.26, ysize = 0.013)+
  add_phylopic(fish_pic, alpha = 1,x = 2.4, y = 0.22, ysize = 0.019)

##Plotting results together
  grid.arrange(plotS2A,plotS2B,plotS2C,plotS2D, nrow=2,ncol=2) 

```


```{r Fig_S3, eval=T,dpi=400,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S3**: Modified phylogenetic approach to test the ortholog conjecture for teleosts and for vertebrates, respectively. We used the oldest duplication node along a path of a phylogeny for identification of pre-duplication and post-duplication speciations. For (A)-(D) plots are based on ", length(Duplicates.our.interest), " and ", length(Duplicates.our.interest.exp), " trees for $\\tau$ and for average gene expression levels, respecively. For (A) and (B), we consider the contrasts of teleosts specific duplication and descendant speciation nodes in absence of teleosts specific pre-duplication speciation nodes contrasts. Results of (C) and (D) are based on all 15 vertebrate species. PICs: Phylogenetic Independent Contrasts, exp: expression, Predup_speciation: Pre-duplication speciation, Postdup_speciation: Post-duplication speciation. Since contrasts are estimated for each tree, paired Wilcoxon test is used to compare the difference.")}

duplicate.effect.teleosts<-duplicate.effect[which(duplicate.effect$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
duplicate.effect.exp.teleosts<-duplicate.effect.exp[which(duplicate.effect.exp$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
###Testing result for tau in teleosts and plotting data
paired.pval1.tau.all.teleost<-paired.wilcox(duplicate.effect.teleosts$predup.spe.pic,duplicate.effect.teleosts$duplication.pic)
paired.pval2.tau.all.teleost<-paired.wilcox(duplicate.effect.teleosts$predup.spe.pic,duplicate.effect.teleosts$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.all.teleost<-data.frame(pic=NA,label=NA)
Frame.tau.all.teleost<-rbind(Frame.tau.all.teleost,data.frame(pic=duplicate.effect.teleosts$predup.spe.pic,label="Predup_speciation"))
Frame.tau.all.teleost<-rbind(Frame.tau.all.teleost,data.frame(pic=duplicate.effect.teleosts$duplication.pic,label="Duplication"))
Frame.tau.all.teleost<-rbind(Frame.tau.all.teleost,data.frame(pic=duplicate.effect.teleosts$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.all.teleost<-Frame.tau.all.teleost[-1,]
Frame.tau.all.teleost$Event <- factor(Frame.tau.all.teleost$label, levels=c( "Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
library(tidyverse)
median.data.all.tau.teleost<-NULL
median.data.all.tau.teleost<-Frame.tau.all.teleost %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.all.tau.teleost$pic.round<-round(median.data.all.tau.teleost$Median,4)
median.data.all.tau.teleost$count<- paste0(median.data.all.tau.teleost$Event," \n(n = ",median.data.all.tau.teleost$freq,")")

plotS3A<-boxplot.our.method(Frame.tau.all.teleost,paired.pval1.tau.all.teleost,paired.pval2.tau.all.teleost,"Tau")+
  ggtitle(expression(bold(paste("A. tissue specificity, ", tau, ", for teleosts"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.all.tau.teleost$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.033, ysize = 0.003)

###Testing result for exp levels in teleosts and plotting data
paired.pval1.exp.all.teleost<-paired.wilcox(duplicate.effect.exp.teleosts$predup.spe.pic,duplicate.effect.exp.teleosts$duplication.pic)
paired.pval2.exp.all.teleost<-paired.wilcox(duplicate.effect.exp.teleosts$predup.spe.pic,duplicate.effect.exp.teleosts$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.all.teleost<-data.frame(pic=NA,label=NA)
Frame.exp.all.teleost<-rbind(Frame.exp.all.teleost,data.frame(pic=duplicate.effect.exp.teleosts$predup.spe.pic,label="Predup_speciation"))
Frame.exp.all.teleost<-rbind(Frame.exp.all.teleost,data.frame(pic=duplicate.effect.exp.teleosts$duplication.pic,label="Duplication"))
Frame.exp.all.teleost<-rbind(Frame.exp.all.teleost,data.frame(pic=duplicate.effect.exp.teleosts$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.all.teleost<-Frame.exp.all.teleost[-1,]
Frame.exp.all.teleost$Event <- factor(Frame.exp.all.teleost$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
median.data.all.exp.teleost<-NULL
median.data.all.exp.teleost<-Frame.exp.all.teleost %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.all.exp.teleost$pic.round<-round(median.data.all.exp.teleost$Median,4)
median.data.all.exp.teleost$count<- paste0(median.data.all.exp.teleost$Event," \n(n = ",median.data.all.exp.teleost$freq,")")

plotS3B<-boxplot.our.method(Frame.exp.all.teleost,paired.pval1.exp.all.teleost,paired.pval2.exp.all.teleost,"Exp") + 
  ggtitle(expression(bold(paste("B. Gene exp levels for teleosts"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.all.exp.teleost$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.23, ysize = 0.019)

###Testing result for tau and plotting data
paired.pval1.tau.all<-paired.wilcox(duplicate.effect$predup.spe.pic,duplicate.effect$duplication.pic)
paired.pval2.tau.all<-paired.wilcox(duplicate.effect$predup.spe.pic,duplicate.effect$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.all<-data.frame(pic=NA,label=NA)
Frame.tau.all<-rbind(Frame.tau.all,data.frame(pic=duplicate.effect$predup.spe.pic,label="Predup_speciation"))
Frame.tau.all<-rbind(Frame.tau.all,data.frame(pic=duplicate.effect$duplication.pic,label="Duplication"))
Frame.tau.all<-rbind(Frame.tau.all,data.frame(pic=duplicate.effect$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.all<-Frame.tau.all[-1,]
Frame.tau.all$Event <- factor(Frame.tau.all$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
library(tidyverse)
median.data.all.tau<-NULL
median.data.all.tau<-Frame.tau.all %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.all.tau$pic.round<-round(median.data.all.tau$Median,4)
median.data.all.tau$count<- paste0(median.data.all.tau$Event," \n(n = ",median.data.all.tau$freq,")")

plotS3C<-boxplot.our.method(Frame.tau.all,paired.pval1.tau.all,paired.pval2.tau.all,"Tau") + 
  ggtitle(expression(bold(paste("C. Tissue specificity, ", tau, ", for 15 species"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.all.tau$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.33, y = 0.032, ysize = 0.002)+
  add_phylopic(fish_pic, alpha = 1,x = 3.33, y = 0.027, ysize = 0.003)

###Testing result for exp levels and plotting data
paired.pval1.exp.all<-paired.wilcox(duplicate.effect.exp$predup.spe.pic,duplicate.effect.exp$duplication.pic)
paired.pval2.exp.all<-paired.wilcox(duplicate.effect.exp$predup.spe.pic,duplicate.effect.exp$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.all<-data.frame(pic=NA,label=NA)
Frame.exp.all<-rbind(Frame.exp.all,data.frame(pic=duplicate.effect.exp$predup.spe.pic,label="Predup_speciation"))
Frame.exp.all<-rbind(Frame.exp.all,data.frame(pic=duplicate.effect.exp$duplication.pic,label="Duplication"))
Frame.exp.all<-rbind(Frame.exp.all,data.frame(pic=duplicate.effect.exp$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.all<-Frame.exp.all[-1,]
Frame.exp.all$Event <- factor(Frame.exp.all$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
median.data.all.exp<-NULL
median.data.all.exp<-Frame.exp.all %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic))
median.data.all.exp$pic.round<-round(median.data.all.exp$Median,4)
median.data.all.exp$count<- paste0(median.data.all.exp$Event," \n(n = ",median.data.all.exp$freq,")")

plotS3D<-boxplot.our.method(Frame.exp.all,paired.pval1.exp.all,paired.pval2.exp.all, "Exp")+
  ggtitle(expression(bold(paste("D. Gene exp levels for 15 species"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.all.tau$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.33, y = 0.22, ysize = 0.013)+
  add_phylopic(fish_pic, alpha = 1,x = 3.33, y = 0.19, ysize = 0.019)

grid.arrange(plotS3A,plotS3B,plotS3C,plotS3D, nrow=2,ncol=2) 

```


```{r Fig_S4, eval=T,dpi=400,fig.width=9, fig.height=8, fig.cap= paste0("**Fig S4**: Randomization tests on our modified phylogenetic approach for teleosts and for vertebrates, respectively. Randomization of traits through tip shuffling allow us to use the same nodes used for the ortholog conjecture tests in Fig. S3. For (A)-(D) plots are based on randomization tests on ", length(Randomized.tau.tree), " and ", length(Randomized.exp.tree), " trees for $\\tau$ and for average gene expression levels, respecively. We considered the contrasts of teleosts specific gene or genome duplication and descendant speciation nodes in (A)-(D) in absence of teleost-specific pre-duplication speciation nodes contrasts. PICs: Phylogenetic Independent Contrasts, exp: expression, Predup_speciation: Pre-duplication speciation, Postdup_speciation: Post-duplication speciation. Since contrasts are estimated for each tree, paired Wilcoxon test is used to compare the difference.")}

duplicate.effect.tau.rand.teleosts<-duplicate.effect.tau.random[which(duplicate.effect.tau.random$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
duplicate.effect.exp.rand.teleosts<-duplicate.effect.exp.random[which(duplicate.effect.exp.random$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
###Testing result for randomized tau in teleosts and plotting data
paired.pval1.tau.rand.teleost<-paired.wilcox(duplicate.effect.tau.rand.teleosts$predup.spe.pic,duplicate.effect.tau.rand.teleosts$duplication.pic)
paired.pval2.tau.rand.teleost<-paired.wilcox(duplicate.effect.tau.rand.teleosts$predup.spe.pic,duplicate.effect.tau.rand.teleosts$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.rand.teleost<-data.frame(pic=NA,label=NA)
Frame.tau.rand.teleost<-rbind(Frame.tau.rand.teleost,data.frame(pic=duplicate.effect.tau.rand.teleosts$predup.spe.pic,label="Predup_speciation"))
Frame.tau.rand.teleost<-rbind(Frame.tau.rand.teleost,data.frame(pic=duplicate.effect.tau.rand.teleosts$duplication.pic,label="Duplication"))
Frame.tau.rand.teleost<-rbind(Frame.tau.rand.teleost,data.frame(pic=duplicate.effect.tau.rand.teleosts$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.rand.teleost<-Frame.tau.rand.teleost[-1,]
Frame.tau.rand.teleost$Event <- factor(Frame.tau.rand.teleost$label, levels=c( "Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
library(tidyverse)
median.data.rand.tau.teleost<-NULL
median.data.rand.tau.teleost<-Frame.tau.rand.teleost %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.rand.tau.teleost$pic.round<-round(median.data.rand.tau.teleost$Median,4)
median.data.rand.tau.teleost$count<- paste0(median.data.rand.tau.teleost$Event," \n(n = ",median.data.rand.tau.teleost$freq,")")

plotS4A<-boxplot.our.method(Frame.tau.rand.teleost,paired.pval1.tau.rand.teleost,paired.pval2.tau.rand.teleost,"Tau")+
  ggtitle(expression(bold(paste("A. Randomized tissue specificity, ", tau, ", for teleosts"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.rand.tau.teleost$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.031, ysize = 0.003)

###Testing result for randomized exp levels in teleosts and plotting data
paired.pval1.exp.rand.teleost<-paired.wilcox(duplicate.effect.exp.rand.teleosts$predup.spe.pic,duplicate.effect.exp.rand.teleosts$duplication.pic)
paired.pval2.exp.rand.teleost<-paired.wilcox(duplicate.effect.exp.rand.teleosts$predup.spe.pic,duplicate.effect.exp.rand.teleosts$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.rand.teleost<-data.frame(pic=NA,label=NA)
Frame.exp.rand.teleost<-rbind(Frame.exp.rand.teleost,data.frame(pic=duplicate.effect.exp.rand.teleosts$predup.spe.pic,label="Predup_speciation"))
Frame.exp.rand.teleost<-rbind(Frame.exp.rand.teleost,data.frame(pic=duplicate.effect.exp.rand.teleosts$duplication.pic,label="Duplication"))
Frame.exp.rand.teleost<-rbind(Frame.exp.rand.teleost,data.frame(pic=duplicate.effect.exp.rand.teleosts$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.rand.teleost<-Frame.exp.rand.teleost[-1,]
Frame.exp.rand.teleost$Event <- factor(Frame.exp.rand.teleost$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
median.data.rand.exp.teleost<-NULL
median.data.rand.exp.teleost<-Frame.exp.rand.teleost %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.rand.exp.teleost$pic.round<-round(median.data.rand.exp.teleost$Median,4)
median.data.rand.exp.teleost$count<- paste0(median.data.rand.exp.teleost$Event," \n(n = ",median.data.rand.exp.teleost$freq,")")

plotS4B<-boxplot.our.method(Frame.exp.rand.teleost,paired.pval1.exp.rand.teleost,paired.pval2.exp.rand.teleost,"Exp") + 
  ggtitle(expression(bold(paste("B. Randomized gene exp levels for teleosts"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.rand.exp.teleost$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.019)

###Testing result for randomized tau for vertebrates and plotting data
paired.paired.pval1.tau.rand<-paired.wilcox(duplicate.effect.tau.random$predup.spe.pic,duplicate.effect.tau.random$duplication.pic)
paired.paired.pval2.tau.rand<-paired.wilcox(duplicate.effect.tau.random$predup.spe.pic,duplicate.effect.tau.random$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.rand<-data.frame(pic=NA,label=NA)
Frame.tau.rand<-rbind(Frame.tau.rand,data.frame(pic=duplicate.effect.tau.random$predup.spe.pic,label="Predup_speciation"))
Frame.tau.rand<-rbind(Frame.tau.rand,data.frame(pic=duplicate.effect.tau.random$duplication.pic,label="Duplication"))
Frame.tau.rand<-rbind(Frame.tau.rand,data.frame(pic=duplicate.effect.tau.random$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.rand<-Frame.tau.rand[-1,]
Frame.tau.rand$Event <- factor(Frame.tau.rand$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
library(tidyverse)
median.data.rand.tau<-NULL
median.data.rand.tau<-Frame.tau.rand %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.rand.tau$pic.round<-round(median.data.rand.tau$Median,4)
median.data.rand.tau$count<- paste0(median.data.rand.tau$Event," \n(n = ",median.data.rand.tau$freq,")")

plotS4C<-boxplot.our.method(Frame.tau.rand,paired.paired.pval1.tau.rand,paired.paired.pval2.tau.rand,"Tau") + 
  ggtitle(expression(bold(paste("C. Randomized tissue specificity, ", tau, ", for 15 species"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.rand.tau$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.33, y = 0.031, ysize = 0.002)+
  add_phylopic(fish_pic, alpha = 1,x = 3.33, y = 0.027, ysize = 0.003)

###Testing result for randomized exp levels for vertebrates and plotting data
paired.pval1.exp.rand<-paired.wilcox(duplicate.effect.exp.random$predup.spe.pic,duplicate.effect.exp.random$duplication.pic)
paired.pval2.exp.rand<-paired.wilcox(duplicate.effect.exp.random$predup.spe.pic,duplicate.effect.exp.random$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.rand<-data.frame(pic=NA,label=NA)
Frame.exp.rand<-rbind(Frame.exp.rand,data.frame(pic=duplicate.effect.exp.random$predup.spe.pic,label="Predup_speciation"))
Frame.exp.rand<-rbind(Frame.exp.rand,data.frame(pic=duplicate.effect.exp.random$duplication.pic,label="Duplication"))
Frame.exp.rand<-rbind(Frame.exp.rand,data.frame(pic=duplicate.effect.exp.random$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.rand<-Frame.exp.rand[-1,]
Frame.exp.rand$Event <- factor(Frame.exp.rand$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
median.data.rand.exp<-NULL
median.data.rand.exp<-Frame.exp.rand %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.rand.exp$pic.round<-round(median.data.rand.exp$Median,4)
median.data.rand.exp$count<- paste0(median.data.rand.exp$Event," \n(n = ",median.data.rand.exp$freq,")")

plotS4D<-boxplot.our.method(Frame.exp.rand,paired.pval1.exp.rand,paired.pval2.exp.rand, "Exp")+
  ggtitle(expression(bold(paste("D. Randomized gene exp levels for 15 species"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.rand.exp$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.33, y = 0.22, ysize = 0.013)+
  add_phylopic(fish_pic, alpha = 1,x = 3.33, y = 0.19, ysize = 0.019)

grid.arrange(plotS4A,plotS4B,plotS4C,plotS4D, nrow=2,ncol=2) 

```



```{r Fig_S5, eval=T,dpi=400,fig.width=9, fig.height=8, fig.cap= paste0("**Fig S5**: Randomization test on our modified phylogenetic approach for trusted 3RWGD and for trusted SSD trees in teleosts, respectively. Randomization of traits through tip shuffling allow us to use the same nodes used for the ortholog conjecture tests in Fig. 4. (A)-(D): Plots are based on randomization tests for ", length(Randomized.tau.tree.3R), " and ", length(Randomized.exp.tree.3R), " trusted 3R whole genome duplicate trees and for ",  length(Randomized.tau.tree.SSD), " and ", length(Randomized.exp.tree.SSD), " trusted small-scale duplicate trees for $\\tau$ and average gene expression levels, respecively. We considered the contrasts of teleosts specific gene or genome duplication and descendant speciation nodes in (A)-(D) in absence of teleost-specific pre-duplication speciation nodes contrasts. PICs: Phylogenetic Independent Contrasts, exp: expression, Predup_speciation: Pre-duplication speciation, Postdup_speciation: Post-duplication speciation, FishWGD: fish specific 3R whole genome duplication, 3RWGDs: ohnologs, SSDs: small-scale duplicates. Since contrasts are estimated for each tree, paired Wilcoxon test is used to compare the difference. ")}

duplicate.effect.3R.teleosts.rand<-duplicate.effect.3R.random
duplicate.effect.3R.exp.teleosts.rand<- duplicate.effect.3R.exp.random

## For randomized tau on trusted 3RWGDs
###Testing result for tau in teleosts and plotting data
paired.pval1.tau.3RWGD.teleost.rand<-paired.wilcox(duplicate.effect.3R.teleosts.rand$predup.spe.pic,duplicate.effect.3R.teleosts.rand$duplication.pic)
paired.pval2.tau.3RWGD.teleost.rand<-paired.wilcox(duplicate.effect.3R.teleosts.rand$predup.spe.pic,duplicate.effect.3R.teleosts.rand$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.3RWGD.teleost.rand<-data.frame(pic=NA,label=NA)
Frame.tau.3RWGD.teleost.rand<-rbind(Frame.tau.3RWGD.teleost.rand,data.frame(pic=duplicate.effect.3R.teleosts.rand$predup.spe.pic,label="Predup_speciation"))
Frame.tau.3RWGD.teleost.rand<-rbind(Frame.tau.3RWGD.teleost.rand,data.frame(pic=duplicate.effect.3R.teleosts.rand$duplication.pic,label="FishWGD"))
Frame.tau.3RWGD.teleost.rand<-rbind(Frame.tau.3RWGD.teleost.rand,data.frame(pic=duplicate.effect.3R.teleosts.rand$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.3RWGD.teleost.rand<-Frame.tau.3RWGD.teleost.rand[-1,]
Frame.tau.3RWGD.teleost.rand$Event <- factor(Frame.tau.3RWGD.teleost.rand$label, levels=c("Predup_speciation","FishWGD","Postdup_speciation"))

## For median data
library(tidyverse)
median.data.3RWGD.tau.teleost.rand<-NULL
median.data.3RWGD.tau.teleost.rand<-Frame.tau.3RWGD.teleost.rand %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.3RWGD.tau.teleost.rand$pic.round<-round(median.data.3RWGD.tau.teleost.rand$Median,4)
median.data.3RWGD.tau.teleost.rand$count<- paste0(median.data.3RWGD.tau.teleost.rand$Event," \n(n = ",median.data.3RWGD.tau.teleost.rand$freq,")")

plotS5A<-boxplot.our.method.3RWGD(Frame.tau.3RWGD.teleost.rand,paired.pval1.tau.3RWGD.teleost.rand,paired.pval2.tau.3RWGD.teleost.rand,"Tau")+
  ggtitle(expression(bold(paste("A. Randomized tissue specificity, ", tau, ", for FishWGDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.3RWGD.tau.teleost.rand$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.031, ysize = 0.0025)

## For randomized exp on trusted 3RWGDs
###Testing result for exp levels in teleosts and plotting data
paired.pval1.exp.3RWGD.teleost.rand<-paired.wilcox(duplicate.effect.3R.exp.teleosts.rand$predup.spe.pic,duplicate.effect.3R.exp.teleosts.rand$duplication.pic)
paired.pval1.exp.3RWGD.teleost.rand<-paired.wilcox(duplicate.effect.3R.exp.teleosts.rand$predup.spe.pic,duplicate.effect.3R.exp.teleosts.rand$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.3RWGD.teleost.rand<-data.frame(pic=NA,label=NA)
Frame.exp.3RWGD.teleost.rand<-rbind(Frame.exp.3RWGD.teleost.rand,data.frame(pic=duplicate.effect.3R.exp.teleosts.rand$predup.spe.pic,label="Predup_speciation"))
Frame.exp.3RWGD.teleost.rand<-rbind(Frame.exp.3RWGD.teleost.rand,data.frame(pic=duplicate.effect.3R.exp.teleosts.rand$duplication.pic,label="FishWGD"))
Frame.exp.3RWGD.teleost.rand<-rbind(Frame.exp.3RWGD.teleost.rand,data.frame(pic=duplicate.effect.3R.exp.teleosts.rand$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.3RWGD.teleost.rand<-Frame.exp.3RWGD.teleost.rand[-1,]
Frame.exp.3RWGD.teleost.rand$Event <- factor(Frame.exp.3RWGD.teleost.rand$label, levels=c("Predup_speciation","FishWGD","Postdup_speciation"))

## For median data
library(tidyverse)
median.data.3RWGD.exp.teleost.rand<-NULL
median.data.3RWGD.exp.teleost.rand<-Frame.exp.3RWGD.teleost.rand%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.3RWGD.exp.teleost.rand$pic.round<-round(median.data.3RWGD.exp.teleost.rand$Median,4)
median.data.3RWGD.exp.teleost.rand$count<- paste0(median.data.3RWGD.exp.teleost.rand$Event," \n(n = ",median.data.3RWGD.exp.teleost.rand$freq,")")

plotS5B<-boxplot.our.method.3RWGD(Frame.exp.3RWGD.teleost.rand,paired.pval1.exp.3RWGD.teleost.rand,paired.pval1.exp.3RWGD.teleost.rand,"Exp") + 
  ggtitle(expression(bold(paste("B. Randomized gene exp levels for FishWGDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.3RWGD.exp.teleost.rand$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.019)

duplicate.effect.SSD.teleosts.rand<-duplicate.effect.SSD.random[which(duplicate.effect.SSD.random$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Characoidei","Oryzias","Atherinomorphae","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")),] 
duplicate.effect.SSD.exp.teleosts.rand<- duplicate.effect.SSD.exp.random[which(duplicate.effect.SSD.exp.random$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Characoidei","Oryzias","Atherinomorphae","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")),]

## For randomized tau on trusted SSDs
###Testing result for tau and plotting data
paired.pval1.tau.SSD.teleost.rand<-paired.wilcox(duplicate.effect.SSD.teleosts.rand$predup.spe.pic,duplicate.effect.SSD.teleosts.rand$duplication.pic)
paired.pval2.tau.SSD.teleost.rand<-paired.wilcox(duplicate.effect.SSD.teleosts.rand$predup.spe.pic,duplicate.effect.SSD.teleosts.rand$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.SSD.teleost.rand<-data.frame(pic=NA,label=NA)
Frame.tau.SSD.teleost.rand<-rbind(Frame.tau.SSD.teleost.rand,data.frame(pic=duplicate.effect.SSD.teleosts.rand$predup.spe.pic,label="Predup_speciation"))
Frame.tau.SSD.teleost.rand<-rbind(Frame.tau.SSD.teleost.rand,data.frame(pic=duplicate.effect.SSD.teleosts.rand$duplication.pic,label="Duplication"))
Frame.tau.SSD.teleost.rand<-rbind(Frame.tau.SSD.teleost.rand,data.frame(pic=duplicate.effect.SSD.teleosts.rand$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.SSD.teleost.rand<-Frame.tau.SSD.teleost.rand[-1,]
Frame.tau.SSD.teleost.rand$Event <- factor(Frame.tau.SSD.teleost.rand$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ))
levels(Frame.tau.SSD.teleost.rand$Event)<-c("Predup_speciation", "SSD","Postdup_speciation")

## For median data
library(tidyverse)
median.data.SSD.tau.teleost.rand<-NULL
median.data.SSD.tau.teleost.rand<-Frame.tau.SSD.teleost.rand %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.SSD.tau.teleost.rand$pic.round<-round(median.data.SSD.tau.teleost.rand$Median,4)
median.data.SSD.tau.teleost.rand$count<- paste0(median.data.SSD.tau.teleost.rand$Event," \n(n = ",median.data.SSD.tau.teleost.rand$freq,")")

plotS5C<-boxplot.our.method(Frame.tau.SSD.teleost.rand,paired.pval1.tau.SSD.teleost.rand,paired.pval2.tau.SSD.teleost.rand,"Tau") + 
  ggtitle(expression(bold(paste("C. Randomized tissue specificity, ", tau, ", for SSDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.SSD.tau.teleost.rand$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.031, ysize = 0.0025)

## For randomized exp on trusted SSDs
###Testing result for exp levels and plotting data
paired.pval1.exp.SSD.teleost.rand<-paired.wilcox(duplicate.effect.SSD.exp.teleosts.rand$predup.spe.pic,duplicate.effect.SSD.exp.teleosts.rand$duplication.pic)
paired.pval2.exp.SSD.teleost.rand<-paired.wilcox(duplicate.effect.SSD.exp.teleosts.rand$predup.spe.pic,duplicate.effect.SSD.exp.teleosts.rand$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.SSD.teleost.rand<-data.frame(pic=NA,label=NA)
Frame.exp.SSD.teleost.rand<-rbind(Frame.exp.SSD.teleost.rand,data.frame(pic=duplicate.effect.SSD.exp.teleosts.rand$predup.spe.pic,label="Predup_speciation"))
Frame.exp.SSD.teleost.rand<-rbind(Frame.exp.SSD.teleost.rand,data.frame(pic=duplicate.effect.SSD.exp.teleosts.rand$duplication.pic,label="Duplication"))
Frame.exp.SSD.teleost.rand<-rbind(Frame.exp.SSD.teleost.rand,data.frame(pic=duplicate.effect.SSD.exp.teleosts.rand$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.SSD.teleost.rand<-Frame.exp.SSD.teleost.rand[-1,]
Frame.exp.SSD.teleost.rand$Event <- factor(Frame.exp.SSD.teleost.rand$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )
levels(Frame.exp.SSD.teleost.rand$Event)<-c("Predup_speciation", "SSD","Postdup_speciation")

## For median data
median.data.SSD.exp.teleost.rand<-NULL
median.data.SSD.exp.teleost.rand<-Frame.exp.SSD.teleost.rand %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.SSD.exp.teleost.rand$pic.round<-round(median.data.SSD.exp.teleost.rand$Median,4)
median.data.SSD.exp.teleost.rand$count<- paste0(median.data.SSD.exp.teleost.rand$Event," \n(n = ",median.data.SSD.exp.teleost.rand$freq,")")

plotS5D<-boxplot.our.method(Frame.exp.SSD.teleost.rand,paired.pval1.exp.SSD.teleost.rand,paired.pval2.exp.SSD.teleost.rand, "Exp")+
  ggtitle(expression(bold(paste("D. Randomized gene exp levels for SSDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.SSD.exp.teleost.rand$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.019)

grid.arrange(plotS5A,plotS5B,plotS5C,plotS5D, nrow=2,ncol=2) 

```


```{r Fig_S6, eval=T,dpi=600,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S6**: Modified phylogenetic approach to test the ortholog conjecture for teleosts and for vertebrates, respectively. We used the youngest duplication node along a path of a phylogeny for identification of pre-duplication and post-duplication speciations. For (A) and (B), we considered the contrasts of teleosts specific duplication and post-duplication speciation nodes in absence of teleost-specific pre-duplication speciation nodes contrasts. (C) and (D) are based on all 15 vertebrate species. PICs: Phylogenetic Independent Contrasts, exp: expression, Predup_speciation: Pre-duplication speciation, Postdup_speciation: Post-duplication speciation. Since contrasts are estimated for each tree, paired Wilcoxon test is used to compare the difference..")}

duplicate.effect.teleosts.young<-duplicate.effect.young[which(duplicate.effect.young$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
duplicate.effect.exp.teleosts.young<-duplicate.effect.young.exp[which(duplicate.effect.young.exp$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Osteoglossocephalai","Characoidei","Oryzias","Atherinomorphae","Clupeocephala","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
###Testing result for tau and plotting data
paired.pval1.tau.all.teleost.young<-paired.wilcox(duplicate.effect.teleosts.young$predup.spe.pic,duplicate.effect.teleosts.young$duplication.pic)
paired.pval2.tau.all.teleost.young<-paired.wilcox(duplicate.effect.teleosts.young$predup.spe.pic,duplicate.effect.teleosts.young$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.all.teleost.young<-data.frame(pic=NA,label=NA)
Frame.tau.all.teleost.young<-rbind(Frame.tau.all.teleost.young,data.frame(pic=duplicate.effect.teleosts.young$predup.spe.pic,label="Predup_speciation"))
Frame.tau.all.teleost.young<-rbind(Frame.tau.all.teleost.young,data.frame(pic=duplicate.effect.teleosts.young$duplication.pic,label="Duplication"))
Frame.tau.all.teleost.young<-rbind(Frame.tau.all.teleost.young,data.frame(pic=duplicate.effect.teleosts.young$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.all.teleost.young<-Frame.tau.all.teleost.young[-1,]
Frame.tau.all.teleost.young$Event <- factor(Frame.tau.all.teleost.young$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation"))

## For median data
library(tidyverse)
median.data.all.tau.teleost.young<-NULL
median.data.all.tau.teleost.young<-Frame.tau.all.teleost.young %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.all.tau.teleost.young$pic.round<-round(median.data.all.tau.teleost.young$Median,4)
median.data.all.tau.teleost.young$count<- paste0(median.data.all.tau.teleost.young$Event," \n(n = ",median.data.all.tau.teleost.young$freq,")")

plotS6A<-boxplot.our.method(Frame.tau.all.teleost.young,paired.pval1.tau.all.teleost.young,paired.pval2.tau.all.teleost.young, "Tau") + 
  ggtitle(expression(bold(paste("A. Tissue specificity, ", tau, ", for teleosts"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.all.tau.teleost.young$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.031, ysize = 0.003)

###Testing result for exp levels and plotting data
paired.pval1.exp.all.teleost.young<-paired.wilcox(duplicate.effect.exp.teleosts.young$predup.spe.pic,duplicate.effect.exp.teleosts.young$duplication.pic)
paired.pval2.exp.all.teleost.young<-paired.wilcox(duplicate.effect.exp.teleosts.young$predup.spe.pic,duplicate.effect.exp.teleosts.young$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.all.teleost.young<-data.frame(pic=NA,label=NA)
Frame.exp.all.teleost.young<-rbind(Frame.exp.all.teleost.young,data.frame(pic=duplicate.effect.exp.teleosts.young$predup.spe.pic,label="Predup_speciation"))
Frame.exp.all.teleost.young<-rbind(Frame.exp.all.teleost.young,data.frame(pic=duplicate.effect.exp.teleosts.young$duplication.pic,label="Duplication"))
Frame.exp.all.teleost.young<-rbind(Frame.exp.all.teleost.young,data.frame(pic=duplicate.effect.exp.teleosts.young$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.all.teleost.young<-Frame.exp.all.teleost.young[-1,]
Frame.exp.all.teleost.young$Event <- factor(Frame.exp.all.teleost.young$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
median.data.all.exp.teleost.young<-NULL
median.data.all.exp.teleost.young<-Frame.exp.all.teleost.young %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.all.exp.teleost.young$pic.round<-round(median.data.all.exp.teleost.young$Median,4)
median.data.all.exp.teleost.young$count<- paste0(median.data.all.exp.teleost.young$Event," \n(n = ",median.data.all.exp.teleost.young$freq,")")

## Overall_paired test plot considering the oldest duplication event
dodge <- position_dodge(width = 0.11)

plotS6B<-boxplot.our.method(Frame.exp.all.teleost.young,paired.pval1.exp.all.teleost.young,paired.pval2.exp.all.teleost.young,"Exp") + 
  ggtitle(expression(bold(paste("B. Gene exp levels for teleosts"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.all.tau.teleost.young$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.019)

###Testing result for tau and plotting data
paired.pval1.tau.all.young<-paired.wilcox(duplicate.effect.young$predup.spe.pic,duplicate.effect.young$duplication.pic)
paired.pval2.tau.all.young<-paired.wilcox(duplicate.effect.young$predup.spe.pic,duplicate.effect.young$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.all.young<-data.frame(pic=NA,label=NA)
Frame.tau.all.young<-rbind(Frame.tau.all.young,data.frame(pic=duplicate.effect.young$predup.spe.pic,label="Predup_speciation"))
Frame.tau.all.young<-rbind(Frame.tau.all.young,data.frame(pic=duplicate.effect.young$duplication.pic,label="Duplication"))
Frame.tau.all.young<-rbind(Frame.tau.all.young,data.frame(pic=duplicate.effect.young$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.all.young<-Frame.tau.all.young[-1,]
Frame.tau.all.young$Event <- factor(Frame.tau.all.young$label, levels=c( "Predup_speciation", "Duplication","Postdup_speciation"))

## For median data
median.data.all.tau.young<-NULL
median.data.all.tau.young<-Frame.tau.all.young %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.all.tau.young$pic.round<-round(median.data.all.tau.young$Median,4)
median.data.all.tau.young$count<- paste0(median.data.all.tau.young$Event," \n(n = ",median.data.all.tau.young$freq,")")

plotS6C<-boxplot.our.method(Frame.tau.all.young,paired.pval1.tau.all.young,paired.pval2.tau.all.young,"Tau") + 
  ggtitle(expression(bold(paste("C. Tissue specificity, ", tau, ", for 15 species"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.all.tau.young$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.33, y = 0.031, ysize = 0.002)+
  add_phylopic(fish_pic, alpha = 1,x = 3.33, y = 0.027, ysize = 0.003)

###Testing result for exp levels and plotting data
paired.pval1.exp.all.young<-paired.wilcox(duplicate.effect.young.exp$predup.spe.pic,duplicate.effect.young.exp$duplication.pic)
paired.pval2.exp.all.young<-paired.wilcox(duplicate.effect.young.exp$predup.spe.pic,duplicate.effect.young.exp$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.all.young<-data.frame(pic=NA,label=NA)
Frame.exp.all.young<-rbind(Frame.exp.all.young,data.frame(pic=duplicate.effect.young.exp$predup.spe.pic,label="Predup_speciation"))
Frame.exp.all.young<-rbind(Frame.exp.all.young,data.frame(pic=duplicate.effect.young.exp$duplication.pic,label="Duplication"))
Frame.exp.all.young<-rbind(Frame.exp.all.young,data.frame(pic=duplicate.effect.young.exp$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.all.young<-Frame.exp.all.young[-1,]
Frame.exp.all.young$Event <- factor(Frame.exp.all.young$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )

## For median data
median.data.all.exp.young<-NULL
median.data.all.exp.young<-Frame.exp.all.young %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq=n())
median.data.all.exp.young$pic.round<-round(median.data.all.exp.young$Median,4)
median.data.all.exp.young$count<- paste0(median.data.all.exp.young$Event," \n(n = ",median.data.all.exp.young$freq,")")

## Overall_paired test plot considering the oldest duplication event
dodge <- position_dodge(width = 0.11)

plotS6D<-boxplot.our.method(Frame.exp.all.young,paired.pval1.exp.all.young,paired.pval2.exp.all.young,"Exp") + 
  ggtitle(expression(bold(paste("D. Gene exp levels for 15 species"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.all.exp.young$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.33, y = 0.22, ysize = 0.015)+
  add_phylopic(fish_pic, alpha = 1,x = 3.33, y = 0.19, ysize = 0.019)

grid.arrange(plotS6A,plotS6B,plotS6C,plotS6D, nrow=2,ncol=2) 

```



```{r Analyses_on_sure_3RWGD_trees_our_method_teleosts_young_duplication_node, eval=T, echo=F, message=FALSE, warning=FALSE, cache=T}

duplicate.effect.teleosts.young.dup<-duplicate.effect3R.young 
duplicate.effect.exp.teleosts.young.dup<-duplicate.effect3R.young.exp

###Testing result for tau and plotting data
paired.pval1.tau.3RWGD.teleost.young.dup<-paired.wilcox(duplicate.effect.teleosts.young.dup$predup.spe.pic,duplicate.effect.teleosts.young.dup$duplication.pic)
paired.pval2.tau.3RWGD.teleost.young.dup<-paired.wilcox(duplicate.effect.teleosts.young.dup$predup.spe.pic,duplicate.effect.teleosts.young.dup$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.3RWGD.teleost.young.dup<-data.frame(pic=NA,label=NA)
Frame.tau.3RWGD.teleost.young.dup<-rbind(Frame.tau.3RWGD.teleost.young.dup,data.frame(pic=duplicate.effect.teleosts.young.dup$predup.spe.pic,label="Predup_speciation"))
Frame.tau.3RWGD.teleost.young.dup<-rbind(Frame.tau.3RWGD.teleost.young.dup,data.frame(pic=duplicate.effect.teleosts.young.dup$duplication.pic,label="FishWGD"))
Frame.tau.3RWGD.teleost.young.dup<-rbind(Frame.tau.3RWGD.teleost.young.dup,data.frame(pic=duplicate.effect.teleosts.young.dup$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.3RWGD.teleost.young.dup<-Frame.tau.3RWGD.teleost.young.dup[-1,]
Frame.tau.3RWGD.teleost.young.dup$Event <- factor(Frame.tau.3RWGD.teleost.young.dup$label, levels=c("Predup_speciation", "FishWGD","Postdup_speciation"))

## For median data
library(tidyverse)
median.data.3RWGD.teleost.young.dup<-NULL
median.data.3RWGD.teleost.young.dup<-Frame.tau.3RWGD.teleost.young.dup %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.3RWGD.teleost.young.dup$pic.round<-round(median.data.3RWGD.teleost.young.dup$Median,4)
median.data.3RWGD.teleost.young.dup$count<- paste0(median.data.3RWGD.teleost.young.dup$Event,"\n(n = ",median.data.3RWGD.teleost.young.dup$freq,")")

###Testing result for exp levels and plotting data
paired.pval1.exp.3RWGD.teleost.young.dup<-paired.wilcox(duplicate.effect.exp.teleosts.young.dup$predup.spe.pic,duplicate.effect.exp.teleosts.young.dup$duplication.pic)
paired.pval2.exp.3RWGD.teleost.young.dup<-paired.wilcox(duplicate.effect.exp.teleosts.young.dup$predup.spe.pic,duplicate.effect.exp.teleosts.young.dup$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.3RWGD.teleost.young.dup<-data.frame(pic=NA,label=NA)
Frame.exp.3RWGD.teleost.young.dup<-rbind(Frame.exp.3RWGD.teleost.young.dup,data.frame(pic=duplicate.effect.exp.teleosts.young.dup$predup.spe.pic,label="Predup_speciation"))
Frame.exp.3RWGD.teleost.young.dup<-rbind(Frame.exp.3RWGD.teleost.young.dup,data.frame(pic=duplicate.effect.exp.teleosts.young.dup$duplication.pic,label="FishWGD"))
Frame.exp.3RWGD.teleost.young.dup<-rbind(Frame.exp.3RWGD.teleost.young.dup,data.frame(pic=duplicate.effect.exp.teleosts.our.method$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.3RWGD.teleost.young.dup<-Frame.exp.3RWGD.teleost.young.dup[-1,]
Frame.exp.3RWGD.teleost.young.dup$Event <- factor(Frame.exp.3RWGD.teleost.young.dup$label, levels=c("Predup_speciation", "FishWGD","Postdup_speciation"))

## For median data
median.data.3RWGD.exp.teleost.young.dup<-NULL
median.data.3RWGD.exp.teleost.young.dup<-Frame.exp.3RWGD.teleost.young.dup %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.3RWGD.exp.teleost.young.dup$pic.round<-round(median.data.3RWGD.exp.teleost.young.dup$Median,4)
median.data.3RWGD.exp.teleost.young.dup$count<- paste0(median.data.3RWGD.exp.teleost.young.dup$Event," \n(n = ",median.data.3RWGD.exp.teleost.young.dup$freq,")")

```


```{r Analyses_on_sure_SSD_trees_our_method_teleosts_young_duplication_node, eval=T,echo=F, message=FALSE, warning=FALSE, cache=T}

duplicate.effect.SSD.teleosts.young.dup<-duplicate.effect.SSD.young[which(duplicate.effect.SSD.young$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Characoidei","Oryzias","Atherinomorphae","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 
duplicate.effect.exp.SSD.teleosts.young.dup<-duplicate.effect.SSD.young.exp[which(duplicate.effect.SSD.young.exp$duplication.label %in% c("Astyanax.mexicanus","Pseudocrenilabrinae","Danio.rerio","Acanthomorphata","Oryzias.latipes","Characiphysae","Euteleosteomorpha","Characoidei","Oryzias","Atherinomorphae","Otophysi","Oryzias.latipes.ASM223467v1","Oreochromis.niloticus","Cichlidae","Percomorphaceae_A","Percomorphaceae_B" ,"Percomorphaceae","Esox.lucius","Ovalentaria","NAME_3","NAME_9","Neopterygii")) ,] 

###Testing result for tau and plotting data
paired.pval1.tau.SSD.teleost.young.dup<-paired.wilcox(duplicate.effect.SSD.teleosts.young.dup$predup.spe.pic,duplicate.effect.SSD.teleosts.young.dup$duplication.pic)
paired.pval2.tau.SSD.teleost.young.dup<-paired.wilcox(duplicate.effect.SSD.teleosts.young.dup$predup.spe.pic,duplicate.effect.SSD.teleosts.young.dup$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.SSD.teleost.young.dup<-data.frame(pic=NA,label=NA)
Frame.tau.SSD.teleost.young.dup<-rbind(Frame.tau.SSD.teleost.young.dup,data.frame(pic=duplicate.effect.SSD.teleosts.young.dup$predup.spe.pic,label="Predup_speciation"))
Frame.tau.SSD.teleost.young.dup<-rbind(Frame.tau.SSD.teleost.young.dup,data.frame(pic=duplicate.effect.SSD.teleosts.young.dup$duplication.pic,label="Duplication"))
Frame.tau.SSD.teleost.young.dup<-rbind(Frame.tau.SSD.teleost.young.dup,data.frame(pic=duplicate.effect.SSD.teleosts.young.dup$postdup.spe.pic,label="Postdup_speciation"))
Frame.tau.SSD.teleost.young.dup<-Frame.tau.SSD.teleost.young.dup[-1,]
Frame.tau.SSD.teleost.young.dup$Event <- factor(Frame.tau.SSD.teleost.young.dup$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation"))
levels(Frame.tau.SSD.teleost.young.dup$Event)<-c("Predup_speciation", "SSD","Postdup_speciation")

## For median data
library(tidyverse)
median.data.SSD.teleost.young.dup<-NULL
median.data.SSD.teleost.young.dup<-Frame.tau.SSD.teleost.young.dup %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.SSD.teleost.young.dup$pic.round<-round(median.data.SSD.teleost.young.dup$Median,4)
#median.data.SSD.teleost.our.method$pic.round<-paste0("Median = ",round(median.data.SSD.teleost.our.method$Median,4), sep="")
median.data.SSD.teleost.young.dup$count<- paste0(median.data.SSD.teleost.young.dup$Event," \n(n = ",median.data.SSD.teleost.young.dup$freq,")")

###Testing result for exp levels and plotting data
paired.pval1.exp.SSD.teleost.young.dup<-paired.wilcox(duplicate.effect.exp.SSD.teleosts.young.dup$predup.spe.pic,duplicate.effect.exp.SSD.teleosts.young.dup$duplication.pic)
paired.pval2.exp.SSD.teleost.young.dup<-paired.wilcox(duplicate.effect.exp.SSD.teleosts.young.dup$predup.spe.pic,duplicate.effect.exp.SSD.teleosts.young.dup$postdup.spe.pic)

##Getting dataframe to plot the result of tau data
Frame.exp.SSD.teleost.young.dup<-data.frame(pic=NA,label=NA)
Frame.exp.SSD.teleost.young.dup<-rbind(Frame.exp.SSD.teleost.young.dup,data.frame(pic=duplicate.effect.exp.SSD.teleosts.young.dup$predup.spe.pic,label="Predup_speciation"))
Frame.exp.SSD.teleost.young.dup<-rbind(Frame.exp.SSD.teleost.young.dup,data.frame(pic=duplicate.effect.exp.SSD.teleosts.young.dup$duplication.pic,label="Duplication"))
Frame.exp.SSD.teleost.young.dup<-rbind(Frame.exp.SSD.teleost.young.dup,data.frame(pic=duplicate.effect.exp.SSD.teleosts.young.dup$postdup.spe.pic,label="Postdup_speciation"))
Frame.exp.SSD.teleost.young.dup<-Frame.exp.SSD.teleost.young.dup[-1,]
Frame.exp.SSD.teleost.young.dup$Event <- factor(Frame.exp.SSD.teleost.young.dup$label, levels=c("Predup_speciation", "Duplication","Postdup_speciation" ) )
levels(Frame.exp.SSD.teleost.young.dup$Event)<-c("Predup_speciation", "SSD","Postdup_speciation")

## For median data
median.data.SSD.exp.teleost.young.dup<-NULL
median.data.SSD.exp.teleost.young.dup<-Frame.exp.SSD.teleost.young.dup %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.SSD.exp.teleost.young.dup$pic.round<-round(median.data.SSD.exp.teleost.young.dup$Median,4)
median.data.SSD.exp.teleost.young.dup$count<- paste0(median.data.SSD.exp.teleost.young.dup$Event," \n(n = ",median.data.SSD.exp.teleost.young.dup$freq,")")

```


```{r Fig_S7, eval=T,dpi=400,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S7**: The ortholog conjecture tests on phylogenies with genome and gene duplicates in teleosts. ", length(Duplicates.our.interest.3R),", ",length(Duplicates.our.interest.3R.exp), ", ", length(All.duplicates.interest.SSD), " and ", length(Duplicates.our.interest.SSD.exp), " of trees passed our criteria of Fig. 4A in (A)-(D), respectively and were used for the analyses. We here used only teleost-specific duplication and post-duplication speciation nodes for contrasts analyses in absence of teleost-specific pre-duplication speciation nodes contrasts. We considered the youngest duplication node along a path of a phylogeny for identification of pre-duplication and post-duplication speciations. PICs: Phylogenetic independent contrasts, exp: expression, Predup_speciation: pre-duplication speciation, Postdup_speciation: post-duplication speciation, FishWGD: fish specific 3R whole genome duplications, 3RWGDs: ohnologs, SSDs: small-scale duplicates. Since contrasts are estimated for each tree, paired Wilcoxon test is used to compare the difference. ")}

plotS7A<-boxplot.our.method.3RWGD(Frame.tau.3RWGD.teleost.young.dup,paired.pval1.tau.3RWGD.teleost.young.dup,paired.pval2.tau.3RWGD.teleost.young.dup, "Tau") + 
  ggtitle(expression(bold(paste("A. Tissue specificity, ", tau, ", for FishWGDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.3RWGD.teleost.young.dup$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.031, ysize = 0.003)

plotS7B<-boxplot.our.method.3RWGD(Frame.exp.3RWGD.teleost.young.dup,paired.pval1.exp.3RWGD.teleost.young.dup,paired.pval2.exp.3RWGD.teleost.young.dup, "Exp") + 
  ggtitle(expression(bold(paste("B. Gene exp levels for FishWGDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.3RWGD.exp.teleost.young.dup$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.019)

plotS7C<-boxplot.our.method(Frame.tau.SSD.teleost.young.dup,paired.pval1.tau.SSD.teleost.young.dup,paired.pval2.tau.SSD.teleost.young.dup,"Tau") +   
  ggtitle(expression(bold(paste("C. Tissue specificity, ", tau, ", for SSDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.SSD.teleost.young.dup$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.031, ysize = 0.003)

plotS7D<-boxplot.our.method(Frame.exp.SSD.teleost.young.dup,paired.pval1.exp.SSD.teleost.our.method,paired.pval2.exp.SSD.teleost.our.method,"Exp") +  
  ggtitle(expression(bold(paste("D. Gene exp levels for SSDs"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  scale_x_discrete(labels=median.data.SSD.exp.teleost.young.dup$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.019)

##Plotting results together
  grid.arrange(plotS7A,plotS7B,plotS7C,plotS7D, nrow=2,ncol=2) 

```


```{r jump_analyses_on_calibrated_trees, echo=F, message=FALSE, warning=FALSE, cache=T}

# Creating summary statistics of tau for 4247 calibrated trees 
tree.jump.tau <- data.frame(
  group=c("Supporting trait jump model","No support for trait jump model"),
  value=c(2636,1611)
)

#Processing data for tau of 4247 calibrated trees
tree.jump.tau <- tree.jump.tau %>% 
  arrange(desc(group)) %>%
  mutate(prop = value / sum(tree.jump.tau$value) *100) %>%
  mutate(percent=paste0(prop = round((value / sum(tree.jump.tau$value) *100),0),"%","\n(",value," trees)", sep="")) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )


# Creating summary statistics of average expression levels for 4603 calibrated trees 
tree.jump.exp <- data.frame(
  group=c("Supporting trait jump model","No support for trait jump model"),
  value=c(2604,1999)
)

#Processing data for average expression levels of 4603 calibrated trees
tree.jump.exp <- tree.jump.exp %>% 
  arrange(desc(group)) %>%
  mutate(prop = value / sum(tree.jump.exp$value) *100) %>%
  mutate(percent=paste0(prop = round((value / sum(tree.jump.exp$value) *100),0),"%","\n(",value," trees)", sep="")) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )

## Statistics of trait jumps for vertebrates
pval.jump.tau.all <- paired.wilcox(info.jump.trees.tau.final1.all$spe.jump.prob,info.jump.trees.tau.final1.all$dup.jump.prob)
pval.jump.exp.all <- pval.jump.all <- paired.wilcox(info.jump.trees.final1.all$spe.jump.prob,info.jump.trees.final1.all$dup.jump.prob)

## Data frame of jumps in tau for vertebrates
dtfr.tau.all<-NULL
dtfr.tau.all<-data.frame(prop.jump=info.jump.trees.tau.final1.all$spe.jump.prob,events="speciation")
dtfr.tau.all<-rbind(dtfr.tau.all,data.frame(prop.jump=info.jump.trees.tau.final1.all$dup.jump.prob,events="duplication"))
dtfr.tau.all$Tau.abs <-abs(dtfr.tau.all$prop.jump)
dtfr.tau.all$Event<-factor(dtfr.tau.all$events, levels=c("speciation","duplication"))
levels(dtfr.tau.all$Event)<-c("Speciation","Duplication")

##Median data of jumps in tau for vertebrates
median.data.jump.tau<-NULL
median.data.jump.tau<- dtfr.tau.all%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Tau.abs=median(prop.jump),
            freq = n())
median.data.jump.tau$pic.round<-paste0("Median = ",round(median.data.jump.tau$Tau.abs,4), sep="")
median.data.jump.tau$count<- paste0(median.data.jump.tau$Event,"\n","(n = ",median.data.jump.tau$freq,")")

## Data frame of jumps in average expression levels for vertebrates
dtfr.all<-NULL
dtfr.all<-data.frame(prop.jump=info.jump.trees.final1.all$spe.jump.prob,events="speciation")
dtfr.all<-rbind(dtfr.all,data.frame(prop.jump=info.jump.trees.final1.all$dup.jump.prob,events="duplication"))
dtfr.all$Exp.abs <-abs(dtfr.all$prop.jump)
dtfr.all$Event<-factor(dtfr.all$events, levels=c("speciation","duplication"))
levels(dtfr.all$Event)<-c("Speciation","Duplication")

##Median data of jumps in average expression levels for vertebrates
median.data.jump.exp<-NULL
median.data.jump.exp<- dtfr.all%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.exp$pic.round<-paste0("Median = ",round(median.data.jump.exp$Exp.abs,4), sep="")
median.data.jump.exp$count<- paste0(median.data.jump.exp$Event,"\n","(n = ",median.data.jump.exp$freq,")")

## Statistics of trait jumps for fishes
pval.jump.tau.fish <- paired.wilcox(info.jump.trees.tau.final1.fish$spe.jump.prob,info.jump.trees.tau.final1.fish$dup.jump.prob) 
pval.jump.exp.fish <- paired.wilcox(info.jump.trees.final1.fish$spe.jump.prob,info.jump.trees.final1.fish$dup.jump.prob) 

## Data frame of jumps in tau for fishes
dtfr.tau.fish<-NULL
dtfr.tau.fish<-data.frame(prop.jump=info.jump.trees.tau.final1.fish$spe.jump.prob,events="speciation")
dtfr.tau.fish<-rbind(dtfr.tau.fish,data.frame(prop.jump=info.jump.trees.tau.final1.fish$dup.jump.prob,events="duplication"))
dtfr.tau.fish$Tau.abs <-abs(dtfr.tau.fish$prop.jump)
dtfr.tau.fish$Event<-factor(dtfr.tau.fish$events, levels=c("speciation","duplication"))
levels(dtfr.tau.fish$Event)<-c("Speciation","Duplication")

## Median data of jumps in tau for fishes
median.data.jump.tau.fish<-NULL
median.data.jump.tau.fish<- dtfr.tau.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Tau.abs=median(prop.jump),
            freq = n())
median.data.jump.tau.fish$pic.round<-paste0("Median = ",round(median.data.jump.tau.fish$Tau.abs,4), sep="")
median.data.jump.tau.fish$count<- paste0(median.data.jump.tau.fish$Event,"\n","(n = ",median.data.jump.tau.fish$freq,")")

## Data frame of jumps in average expression levels for fishes
dtfr.fish<-NULL
dtfr.fish<-data.frame(prop.jump=info.jump.trees.final1.fish$spe.jump.prob,events="speciation")
dtfr.fish<-rbind(dtfr.fish,data.frame(prop.jump=info.jump.trees.final1.fish$dup.jump.prob,events="duplication"))
dtfr.fish$Exp.abs <-abs(dtfr.fish$prop.jump)
dtfr.fish$Event<-factor(dtfr.fish$events, levels=c("speciation","duplication"))
levels(dtfr.fish$Event)<-c("Speciation","Duplication")

## Median data of jumps in average expression levels for fishes
median.data.jump.exp.fish<-NULL
median.data.jump.exp.fish<- dtfr.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
#median <- aggregate(pic ~ Event, Frame, median)
median.data.jump.exp.fish$pic.round<-paste0("Median = ",round(median.data.jump.exp.fish$Exp.abs,4), sep="")
median.data.jump.exp.fish$count<- paste0(median.data.jump.exp.fish$Event,"\n","(n = ",median.data.jump.exp.fish$freq,")")


```



```{r Fig_S8, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S8**: Analyses of trait jump models for $\\tau$ and for average gene expression levels in 5 teleosts and in 15 vertebrates. We used a strict posterior probability cutoff of $\\ge$ 0.7  as an evidence of trait jump. (A)-(B) Statistics of calibrated Brownian trees supporting jumps for $\\tau$ and for average gene expression levels, respectively. Proportions of jumps in traits following speciations and duplication events are shown in (C)-(D) for teleosts and in (E)-(F) for vertebrates. 2636 trees supporting rapid jumps in $\\tau$ and 2604 trees supporting jumps in average gene expression levels with the above-mentioned threshold were used for analyses. Since proportion of shifts in traits per branch of events is estimated for each tree, paired Wilcoxon test is used to compare the difference. ‘exp’: Average gene expression levels.")}

# Piechart of tau for vertebrates
plotS8A<-ggplot(tree.jump.tau, aes(x="", y=prop, fill=group)) +
  geom_bar(stat="identity", width=2, color="white") +
  coord_polar("y", start=0) +
  theme_void() + 
  theme(legend.title=element_blank(),legend.text=element_text(size=9, face = "bold"),legend.position=c(1.0, 0.8)) +
  #theme(legend.title=element_blank(),legend.position="top") +
  ggtitle(expression(bold(paste("A. Tree statistics for tissue specificity, ", tau))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  geom_text(aes(y = ypos, label = percent), color = "black", size=4,fontface = "bold")+ 
  scale_fill_brewer(palette = "Oranges")

# Piechart of average expression levels for vertebrates
plotS8B<-ggplot(tree.jump.exp, aes(x="", y=prop, fill=group)) +
  geom_bar(stat="identity", width=2, color="white") +
  coord_polar("y", start=0) +
  theme_void() + 
  #theme(legend.title=element_blank(),legend.position=c(1.1, 0.8)) +
  #guides(fill=guide_legend(ncol=2))+
  theme(legend.title=element_blank(),legend.position="none") +
  ggtitle(expression(bold(paste("B. Tree statistics for exp levels"))))+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  geom_text(aes(y = ypos, label = percent), color = "black", size=4,fontface = "bold") +
  scale_fill_brewer(palette = "Oranges")

## Proportions of jumps in tau for teleosts
plotS8C<-vioplot.new2(dtfr.tau.fish,pval.jump.tau.fish,"Tau",median.data.jump.tau.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("C. Jumps in ", tau, " for teleosts"))))+
  ylab(expression(bold(paste("Proportions of trait jumps")))) +
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax =0.78, alpha=0.8,colour = "blue")+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.tau.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.tau.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.tau.fish,")")), parse=T, size=4.04)+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.tau.fish, parse=T)+
  scale_x_discrete(labels=median.data.jump.tau.fish$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.1, color = "black")  

# Proportions of jumps in average expression levels for teleosts
plotS8D<- vioplot.new2(dtfr.fish,pval.jump.exp.fish,"Exp",median.data.jump.exp.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("D. Jumps in exp levels for teleosts"))))+
  #ylab(expression(bold(paste("Proportions of jump events for teleosts ")))) +
  ylab(" ")+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax =0.78, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.exp.fish, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.exp.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.exp.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.exp.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.exp.fish$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.1, color = "black")  

# Proportions of jumps in tau for vertebrates
plotS8E<-vioplot.new2(dtfr.tau.all,pval.jump.tau.all,"Tau",median.data.jump.tau) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("E. Jumps in ", tau, " for 15 vertebrates"))))+
  ylab(expression(bold(paste("Proportions of trait jumps")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax =0.78, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.tau.all,parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.tau.all,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.tau.all,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.tau.all,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.tau$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.07, color = "black")+
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.82, ysize = 0.1, color = "black")

# Proportions of jumps in average expression levels for vertebrates
plotS8F<-vioplot.new2(dtfr.all,pval.jump.exp.all,"Exp",median.data.jump.exp) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("F. Jumps in exp levels for 15 vertebrates"))))+
  ylab(" ")+
  theme(plot.title = element_text(color="black", size=12, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax =0.78, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.exp.all,parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.exp.all,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.exp.all,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.exp.all,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.tau$count)+ 
  scale_x_discrete(labels=median.data.jump.exp$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.07, color = "black")+
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.82, ysize = 0.1, color = "black")
  
  
gridExtra::grid.arrange(plotS8A,plotS8B,plotS8C,plotS8D,plotS8E,plotS8F,ncol=2, nrow=3)

```


```{r OC_analyses_on_calibrated_trees_teleosts, eval=T,echo=F, message=FALSE, warning=FALSE, cache=T}

##Testing on tau for trees supporting jumps in teleosts
speciation.tau.jump.fish <-speciation.contrast(nodes.contrast.tau.jumps.fish ,"Tau.abs")
duplication.tau.jump.fish <-duplication.contrast(nodes.contrast.tau.jumps.fish,"Tau.abs")

##Stats on tau for for trees supporting jumps in teleosts
Pval.tau.jump.fish <- two.tailed.wilcox(duplication.tau.jump.fish,speciation.tau.jump.fish)

## Median data for trees supporting jump in tau for teleosts
levels(nodes.contrast.tau.jumps.fish$Event)<-c("Speciation", "Duplication")
median.tau.jump.teleosts<-NULL
median.tau.jump.teleosts<-nodes.contrast.tau.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.jump.teleosts$pic.round<-paste0("Median = ",round(median.tau.jump.teleosts$Tau.abs,4), sep="")
median.tau.jump.teleosts$count<- paste0(median.tau.jump.teleosts$Event,"\n","(n = ",median.tau.jump.teleosts$freq,")")

##Testing on tau after removal of nodes for which daughter branches support jumps in teleosts
speciation.tau.jump.fish.removing.jumpnode <-speciation.contrast(nodes.contrast.tau.jumps.fish.removing.jump.node ,"Tau.abs")
duplication.tau.jump.fish.removing.jumpnode <-duplication.contrast(nodes.contrast.tau.jumps.fish.removing.jump.node,"Tau.abs")

##Stats for trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in teleosts
Pval.tau.jump.fish.removing.jumpnode <- two.tailed.wilcox(duplication.tau.jump.fish.removing.jumpnode,speciation.tau.jump.fish.removing.jumpnode)

## Median data of trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in teleosts
levels(nodes.contrast.tau.jumps.fish.removing.jump.node$Event)<-c("Speciation", "Duplication")
median.tau.removing.jumpnode.teleosts<-NULL
median.tau.removing.jumpnode.teleosts<-nodes.contrast.tau.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.removing.jumpnode.teleosts$pic.round<-paste0("Median = ",round(median.tau.removing.jumpnode.teleosts$Tau.abs,4), sep="")
median.tau.removing.jumpnode.teleosts$count<- paste0(median.tau.removing.jumpnode.teleosts$Event,"\n","(n = ",median.tau.removing.jumpnode.teleosts$freq,")")


##Testing on tau for trees not supporting jump in teleosts
speciation.tau.nojump.fish <-speciation.contrast(nodes.contrast.tau.nojumps.fish ,"Tau.abs")
duplication.tau.nojump.fish <-duplication.contrast(nodes.contrast.tau.nojumps.fish,"Tau.abs")

##Stats on tau for for trees not supporting jump in teleosts
Pval.tau.nojump.fish <- two.tailed.wilcox(duplication.tau.nojump.fish,speciation.tau.nojump.fish)

## Median data for trees not supporting jump in teleosts
levels(nodes.contrast.tau.nojumps.fish$Event)<-c("Speciation", "Duplication")
median.tau.nojump.teleosts<-NULL
median.tau.nojump.teleosts<-nodes.contrast.tau.nojumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.nojump.teleosts$pic.round<-paste0("Median = ",round(median.tau.nojump.teleosts$Tau.abs,4), sep="")
median.tau.nojump.teleosts$count<- paste0(median.tau.nojump.teleosts$Event,"\n","(n = ",median.tau.nojump.teleosts$freq,")")


##Testing on average expression levels for trees supporting jumps in teleosts
speciation.exp.jump.fish <-speciation.contrast(nodes.contrast.exp.jumps.fish ,"Exp.abs")
duplication.exp.jump.fish <-duplication.contrast(nodes.contrast.exp.jumps.fish,"Exp.abs")

##Stats on average expression levels for trees supporting jumps in teleosts
Pval.exp.jump.fish <- two.tailed.wilcox(duplication.exp.jump.fish,speciation.exp.jump.fish)

## Median data for trees supporting jump in average expression levels for teleosts
levels(nodes.contrast.exp.jumps.fish$Event)<-c("Speciation", "Duplication")
median.exp.jump.teleosts<-NULL
median.exp.jump.teleosts<-nodes.contrast.exp.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.teleosts$pic.round<-paste0("Median = ",round(median.exp.jump.teleosts$Exp.abs,4), sep="")
median.exp.jump.teleosts$count<- paste0(median.exp.jump.teleosts$Event,"\n","(n = ",median.exp.jump.teleosts$freq,")")

##Testing on average expression levels after removal of nodes for which daughter branches support jumps in teleosts
speciation.exp.jump.fish.removing.jumpnode <-speciation.contrast(nodes.contrast.exp.jumps.fish.removing.jump.node ,"Exp.abs")
duplication.exp.jump.fish.removing.jumpnode <-duplication.contrast(nodes.contrast.exp.jumps.fish.removing.jump.node,"Exp.abs")

##Stats for trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in teleosts
Pval.exp.jump.fish.removing.jumpnode <- two.tailed.wilcox(duplication.exp.jump.fish.removing.jumpnode,speciation.exp.jump.fish.removing.jumpnode)

## Median data of trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in teleosts
levels(nodes.contrast.exp.jumps.fish.removing.jump.node$Event)<-c("Speciation", "Duplication")
median.exp.removing.jumpnode.teleosts<-NULL
median.exp.removing.jumpnode.teleosts<-nodes.contrast.exp.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.removing.jumpnode.teleosts$pic.round<-paste0("Median = ",round(median.exp.removing.jumpnode.teleosts$Exp.abs,4), sep="")
median.exp.removing.jumpnode.teleosts$count<- paste0(median.exp.removing.jumpnode.teleosts$Event,"\n","(n = ",median.exp.removing.jumpnode.teleosts$freq,")")


##Testing on average expression levels for trees not supporting jump in teleosts
speciation.exp.nojump.fish <-speciation.contrast(nodes.contrast.exp.nojumps.fish ,"Exp.abs")
duplication.exp.nojump.fish <-duplication.contrast(nodes.contrast.exp.nojumps.fish,"Exp.abs")

##Stats on average expression levels for for trees not supporting jump in teleosts
Pval.exp.nojump.fish <- two.tailed.wilcox(duplication.exp.nojump.fish,speciation.exp.nojump.fish)

## Median data for trees not supporting jump in average expression levels in teleosts
levels(nodes.contrast.exp.nojumps.fish$Event)<-c("Speciation", "Duplication")
median.exp.nojump.teleosts<-NULL
median.exp.nojump.teleosts<-nodes.contrast.exp.nojumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.nojump.teleosts$pic.round<-paste0("Median = ",round(median.exp.nojump.teleosts$Exp.abs,4), sep="")
median.exp.nojump.teleosts$count<- paste0(median.exp.nojump.teleosts$Event,"\n","(n = ",median.exp.nojump.teleosts$freq,")")

```



```{r OC_analyses_on_calibrated_trees_vertebrates, eval=T,echo=F, message=FALSE, warning=FALSE, cache=T}

##Testing on tau for trees supporting jumps in vertebrates
speciation.tau.jump.15spe <-speciation.contrast(nodes.contrast.tau.jumps.15spe ,"Tau.abs")
duplication.tau.jump.15spe <-duplication.contrast(nodes.contrast.tau.jumps.15spe,"Tau.abs")

##Stats on tau for for trees supporting jumps in vertebrates
Pval.tau.jump.15spe <- two.tailed.wilcox(duplication.tau.jump.15spe,speciation.tau.jump.15spe)

## Median data for trees supporting jump in tau for vertebrates
levels(nodes.contrast.tau.jumps.15spe$Event)<-c("Speciation", "Duplication")
median.tau.jump.15spe<-NULL
median.tau.jump.15spe<-nodes.contrast.tau.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.jump.15spe$pic.round<-paste0("Median = ",round(median.tau.jump.15spe$Tau.abs,4), sep="")
median.tau.jump.15spe$count<- paste0(median.tau.jump.15spe$Event,"\n","(n = ",median.tau.jump.15spe$freq,")")

##Testing on tau after removal of nodes for which daughter branches support jumps in vertebrates
speciation.tau.jump.15spe.removing.jumpnode <-speciation.contrast(nodes.contrast.tau.jumps.15spe.removing.jump.node ,"Tau.abs")
duplication.tau.jump.15spe.removing.jumpnode <-duplication.contrast(nodes.contrast.tau.jumps.15spe.removing.jump.node,"Tau.abs")

##Stats for trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in vertebrates
Pval.tau.jump.15spe.removing.jumpnode <- two.tailed.wilcox(duplication.tau.jump.15spe.removing.jumpnode,speciation.tau.jump.15spe.removing.jumpnode)

## Median data of trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in vertebrates
levels(nodes.contrast.tau.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "Duplication")
median.tau.removing.jumpnode.15spe<-NULL
median.tau.removing.jumpnode.15spe<-nodes.contrast.tau.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.removing.jumpnode.15spe$pic.round<-paste0("Median = ",round(median.tau.removing.jumpnode.15spe$Tau.abs,4), sep="")
median.tau.removing.jumpnode.15spe$count<- paste0(median.tau.removing.jumpnode.15spe$Event,"\n","(n = ",median.tau.removing.jumpnode.15spe$freq,")")


##Testing on tau for trees not supporting jump in vertebrates
speciation.tau.nojump.15spe <-speciation.contrast(nodes.contrast.tau.nojumps.15spe ,"Tau.abs")
duplication.tau.nojump.15spe <-duplication.contrast(nodes.contrast.tau.nojumps.15spe,"Tau.abs")

##Stats on tau for for trees not supporting jump in vertebrates
Pval.tau.nojump.15spe <- two.tailed.wilcox(duplication.tau.nojump.15spe,speciation.tau.nojump.15spe)

## Median data for trees not supporting jump in vertebrates
levels(nodes.contrast.tau.nojumps.15spe$Event)<-c("Speciation", "Duplication")
median.tau.nojump.15spe<-NULL
median.tau.nojump.15spe<-nodes.contrast.tau.nojumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.nojump.15spe$pic.round<-paste0("Median = ",round(median.tau.nojump.15spe$Tau.abs,4), sep="")
median.tau.nojump.15spe$count<- paste0(median.tau.nojump.15spe$Event,"\n","(n = ",median.tau.nojump.15spe$freq,")")


##Testing on average expression levels for trees supporting jumps in vertebrates
speciation.exp.jump.15spe <-speciation.contrast(nodes.contrast.exp.jumps.15spe ,"Exp.abs")
duplication.exp.jump.15spe <-duplication.contrast(nodes.contrast.exp.jumps.15spe,"Exp.abs")

##Stats on average expression levels for trees supporting jumps in vertebrates
Pval.exp.jump.15spe <- two.tailed.wilcox(duplication.exp.jump.15spe,speciation.exp.jump.15spe)

## Median data for trees supporting jump in average expression levels for vertebrates
levels(nodes.contrast.exp.jumps.15spe$Event)<-c("Speciation", "Duplication")
median.exp.jump.15spe<-NULL
median.exp.jump.15spe<-nodes.contrast.exp.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.15spe$Exp.abs,4), sep="")
median.exp.jump.15spe$count<- paste0(median.exp.jump.15spe$Event,"\n","(n = ",median.exp.jump.15spe$freq,")")

##Testing on average expression levels after removal of nodes for which daughter branches support jumps in vertebrates
speciation.exp.jump.15spe.removing.jumpnode <-speciation.contrast(nodes.contrast.exp.jumps.15spe.removing.jump.node ,"Exp.abs")
duplication.exp.jump.15spe.removing.jumpnode <-duplication.contrast(nodes.contrast.exp.jumps.15spe.removing.jump.node,"Exp.abs")

##Stats for trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in vertebrates
Pval.exp.jump.15spe.removing.jumpnode <- two.tailed.wilcox(duplication.exp.jump.15spe.removing.jumpnode,speciation.exp.jump.15spe.removing.jumpnode)

## Median data of trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in vertebrates
levels(nodes.contrast.exp.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "Duplication")
median.exp.removing.jumpnode.15spe<-NULL
median.exp.removing.jumpnode.15spe<-nodes.contrast.exp.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.removing.jumpnode.15spe$pic.round<-paste0("Median = ",round(median.exp.removing.jumpnode.15spe$Exp.abs,4), sep="")
median.exp.removing.jumpnode.15spe$count<- paste0(median.exp.removing.jumpnode.15spe$Event,"\n","(n = ",median.exp.removing.jumpnode.15spe$freq,")")

##Testing on average expression levels for trees not supporting jump in vertebrates
speciation.exp.nojump.15spe <-speciation.contrast(nodes.contrast.exp.nojumps.15spe ,"Exp.abs")
duplication.exp.nojump.15spe <-duplication.contrast(nodes.contrast.exp.nojumps.15spe,"Exp.abs")

##Stats on average expression levels for for trees not supporting jump in vertebrates
Pval.exp.nojump.15spe <- two.tailed.wilcox(duplication.exp.nojump.15spe,speciation.exp.nojump.15spe)

## Median data for trees not supporting jump in average expression levels in vertebrates
levels(nodes.contrast.exp.nojumps.15spe$Event)<-c("Speciation", "Duplication")
median.exp.nojump.15spe<-NULL
median.exp.nojump.15spe<-nodes.contrast.exp.nojumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.nojump.15spe$pic.round<-paste0("Median = ",round(median.exp.nojump.15spe$Exp.abs,4), sep="")
median.exp.nojump.15spe$count<- paste0(median.exp.nojump.15spe$Event,"\n","(n = ",median.exp.nojump.15spe$freq,")")

```



```{r Fig_S9, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S9**: The ortholog conjecture (OC) tests on trees supporting rapid jump in $\\tau$ and in average gene expression levels for 15 vertebrates. We used a strict branch jump posterior cutoff of $\\ge$ 0.7 as an evidence of trait jump. Only teleosts-specific clades were used for the analyses. The ortholog conjecture tests on (A) 2636 calibrated contrasts standardized trees supporting rapid jumps in $\\tau$, and (B) 2604 trees supporting jumps in average gene expression levels with the above-mentioned threshold along a phylogeny for our analyses. (C)-(D) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for 2636 and 2604 trees, respectively. (E)-(F) The ortholog conjecture tests on 1611 and 1999 trees for which the trait jump model is not supported for $\\tau$ and for average gene expression levels, respectively. *P* values are from Wilcoxon two-tailed tests.")}

dodge <- position_dodge(width = 0.31)

# The ortholog conjecture test on tau for trees supporting jump in vertebrates 
plotS9A<- boxplot.new(nodes.contrast.tau.jumps.15spe,Pval.tau.jump.15spe,"Tau",median.tau.jump.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("A. OC test on trees supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.jump.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004)

# The ortholog conjecture test on average expression levels for trees supporting jump in vertebrates 
plotS9B<- boxplot.new(nodes.contrast.exp.jumps.15spe,Pval.exp.jump.15spe,"Exp",median.exp.jump.15spe) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. OC test on trees supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.019)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025) 

# The ortholog conjecture test on tau after removal of node contrasts for which the daughter branches experience jumps in tau in vertebrates 
plotS9C<- boxplot.new(nodes.contrast.tau.jumps.15spe.removing.jump.node,Pval.tau.jump.15spe.removing.jumpnode,"Tau",median.tau.removing.jumpnode.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("C. Test after removing nodes supporting ", tau," jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.removing.jumpnode.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004)


# The ortholog conjecture test on average expression levels after removal of node contrasts for which the daughter branches experience jumps in average expression levels in vertebrates 
plotS9D<- boxplot.new(nodes.contrast.exp.jumps.15spe.removing.jump.node,Pval.exp.jump.15spe.removing.jumpnode,"Exp",median.exp.removing.jumpnode.15spe) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Test after removing nodes supporting exp jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.removing.jumpnode.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.019)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025)

# The ortholog conjecture test on tau for trees do not provide support for jumps in vertebrates
plotS9E<- boxplot.new(nodes.contrast.tau.nojumps.15spe,Pval.tau.nojump.15spe ,"Tau",median.tau.nojump.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("E. OC test on trees not supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.nojump.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004)


# The ortholog conjecture test on average expression levels for trees do not provide support for jumps in vertebrates
plotS9F<- boxplot.new(nodes.contrast.exp.nojumps.15spe,Pval.exp.nojump.15spe ,"Exp",median.exp.nojump.15spe) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. OC test on trees not supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.nojump.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.019)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025) 


gridExtra::grid.arrange(plotS9A,plotS9B,plotS9C,plotS9D,plotS9E,plotS9F,ncol=2,layout_matrix=rbind(c(1,2),c(3,4),c(5,6)))

```


```{r Fig_S10, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S10**: The ortholog conjecture (OC) tests on trees supporting rapid jump in $\\tau$ and in average gene expression levels for 5 teleosts. We used a strict branch jump posterior cutoff of  $\\ge$ 0.7 as an evidence of trait jump. Only teleosts-specific clades were used for the analyses. (A) 2636 calibrated contrasts standardized trees supporting rapid jumps in $\\tau$, and (B) 2604 trees supporting jumps in average gene expression levels were used for the analyses. (C)-(D) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for 2636 and 2604 trees, respectively. (E)-(F) The ortholog conjecture tests on 1611 and 1999 trees for which the trait jump model is not supported for $\\tau$ and for average gene expression levels, respectively.*P* values are from Wilcoxon two-tailed tests.")}

dodge <- position_dodge(width = 0.31)

# The ortholog conjecture test on tau for trees supporting jump in teleosts 
plotS10A<- boxplot.new(nodes.contrast.tau.jumps.fish,Pval.tau.jump.fish,"Tau",median.tau.jump.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("A. OC test on trees supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.jump.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004)


# The ortholog conjecture test on average expression levels for trees supporting jump in teleosts 
plotS10B<- boxplot.new(nodes.contrast.exp.jumps.fish,Pval.exp.jump.fish,"Exp",median.exp.jump.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. OC test on trees supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025) 

# The ortholog conjecture test on tau after removal of node contrasts for which the daughter branches experience jumps in tau in teleosts 
plotS10C<- boxplot.new(nodes.contrast.tau.jumps.fish.removing.jump.node,Pval.tau.jump.fish.removing.jumpnode,"Tau",median.tau.removing.jumpnode.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("C. OC test after removing nodes supporting ", tau," jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.removing.jumpnode.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004)


# The ortholog conjecture test on average expression levels after removal of node contrasts for which the daughter branches experience jumps in average expression levels in teleosts 
plotS10D<- boxplot.new(nodes.contrast.exp.jumps.fish.removing.jump.node,Pval.exp.jump.fish.removing.jumpnode,"Exp",median.exp.removing.jumpnode.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. OC test after removing nodes supporting exp jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.removing.jumpnode.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025) 

# The ortholog conjecture test on tau for trees do not provide support for jumps in teleosts
plotS10E<- boxplot.new(nodes.contrast.tau.nojumps.fish,Pval.tau.nojump.fish ,"Tau",median.tau.nojump.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("E. OC test on trees not supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.nojump.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004)

# The ortholog conjecture test on average expression levels for trees do not provide support for jumps in teleosts
plotS10F<- boxplot.new(nodes.contrast.exp.nojumps.fish,Pval.exp.nojump.fish ,"Exp",median.exp.nojump.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. OC test on trees not supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.nojump.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025) 


gridExtra::grid.arrange(plotS10A,plotS10B,plotS10C,plotS10D,plotS10E,plotS10F,ncol=2, nrow=3)

```



```{r OC_analyses_on_3RWGDs_jump_trees_teleosts, eval=T,echo=F, message=FALSE, warning=FALSE, cache=T}

##Testing on tau for 3RWGD trees supporting jumps in teleosts
speciation.tau.jump.3RWGD.fish <-speciation.contrast(nodes.contrast.3RWGD.tau.jumps.fish ,"Tau.abs")
duplication.tau.jump.3RWGD.fish1 <-duplication.contrast.3R(nodes.contrast.3RWGD.tau.jumps.fish,"Tau.abs")
duplication.tau.jump.3RWGD.fish2 <-duplication.contrast(nodes.contrast.3RWGD.tau.jumps.fish,"Tau.abs")
#duplication.tau.jump.3RWGD.fish2 <-nodes.contrast.3RWGD.tau.jumps.fish$Tau.abs[which(nodes.contrast.3RWGD.tau.jumps.fish$Event=="SSD")]

##Stats on tau for 3RWGD trees supporting jumps in teleosts
Pval.tau.jump.3RWGD.fish1 <- two.tailed.wilcox(duplication.tau.jump.3RWGD.fish1,speciation.tau.jump.3RWGD.fish)
Pval.tau.jump.3RWGD.fish2 <- two.tailed.wilcox(duplication.tau.jump.3RWGD.fish2,speciation.tau.jump.3RWGD.fish)

## Median data for 3RWGD trees supporting jump in tau for teleosts
levels(nodes.contrast.3RWGD.tau.jumps.fish$Event)<-c("Speciation", "SSD", "FishWGD")
median.tau.jump.3RWGD.teleosts<-NULL
median.tau.jump.3RWGD.teleosts<-nodes.contrast.3RWGD.tau.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.jump.3RWGD.teleosts$pic.round<-paste0("Median = ",round(median.tau.jump.3RWGD.teleosts$Tau.abs,4), sep="")
median.tau.jump.3RWGD.teleosts$count<- paste0(median.tau.jump.3RWGD.teleosts$Event,"\n","(n = ",median.tau.jump.3RWGD.teleosts$freq,")")

##Testing on tau after removal of nodes for which daughter branches support jumps in 3RWGD trees in teleosts
speciation.tau.jump.3RWGD.fish.removing.jumpnode <-speciation.contrast(nodes.contrast.3RWGD.tau.jumps.fish.removing.jump.node ,"Tau.abs")
duplication.tau.jump.3RWGD.fish.removing.jumpnode1 <-duplication.contrast.3R(nodes.contrast.3RWGD.tau.jumps.fish.removing.jump.node,"Tau.abs")
duplication.tau.jump.3RWGD.fish.removing.jumpnode2 <-duplication.contrast(nodes.contrast.3RWGD.tau.jumps.fish.removing.jump.node,"Tau.abs")
#duplication.tau.jump.3RWGD.fish.removing.jumpnode2 <-nodes.contrast.3RWGD.tau.jumps.fish.removing.jump.node$Tau.abs[which(nodes.contrast.3RWGD.tau.jumps.fish.removing.jump.node$Event=="SSD")]

##Stats for trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in 3RWGD trees in teleosts
Pval.tau.jump.3RWGD.fish.removing.jumpnode1 <- two.tailed.wilcox(duplication.tau.jump.3RWGD.fish.removing.jumpnode1,speciation.tau.jump.3RWGD.fish.removing.jumpnode)

Pval.tau.jump.3RWGD.fish.removing.jumpnode2 <- two.tailed.wilcox(duplication.tau.jump.3RWGD.fish.removing.jumpnode2,speciation.tau.jump.3RWGD.fish.removing.jumpnode)

## Median data of trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in 3RWGD trees in teleosts
levels(nodes.contrast.3RWGD.tau.jumps.fish.removing.jump.node$Event)<-c("Speciation", "SSD", "FishWGD")
median.tau.removing.3RWGD.jumpnode.teleosts<-NULL
median.tau.removing.3RWGD.jumpnode.teleosts<-nodes.contrast.3RWGD.tau.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.removing.3RWGD.jumpnode.teleosts$pic.round<-paste0("Median = ",round(median.tau.removing.3RWGD.jumpnode.teleosts$Tau.abs,4), sep="")
median.tau.removing.3RWGD.jumpnode.teleosts$count<- paste0(median.tau.removing.3RWGD.jumpnode.teleosts$Event,"\n","(n = ",median.tau.removing.3RWGD.jumpnode.teleosts$freq,")")

##Testing on tau for trees not supporting jump in 3RWGD trees in teleosts
speciation.tau.nojump.3RWGD.fish <-speciation.contrast(nodes.contrast.3RWGD.tau.nojumps.fish ,"Tau.abs")
duplication.tau.nojump.3RWGD.fish1 <-duplication.contrast.3R(nodes.contrast.3RWGD.tau.nojumps.fish,"Tau.abs")
duplication.tau.nojump.3RWGD.fish2 <-duplication.contrast(nodes.contrast.3RWGD.tau.nojumps.fish,"Tau.abs")
#duplication.tau.nojump.3RWGD.fish2 <-nodes.contrast.3RWGD.tau.nojumps.fish$Tau.abs[which(nodes.contrast.3RWGD.tau.nojumps.fish$Event=="SSD")]


##Stats on tau for for trees not supporting jump in 3RWGD trees in teleosts
Pval.tau.nojump.3RWGD.fish1 <- two.tailed.wilcox(duplication.tau.nojump.3RWGD.fish1,speciation.tau.nojump.3RWGD.fish)
Pval.tau.nojump.3RWGD.fish2 <- two.tailed.wilcox(duplication.tau.nojump.3RWGD.fish2,speciation.tau.nojump.3RWGD.fish)


## Median data for trees not supporting jump in 3RWGD trees in teleosts
levels(nodes.contrast.3RWGD.tau.nojumps.fish$Event)<-c("Speciation", "SSD", "FishWGD")
median.tau.nojump.3RWGD.teleosts<-NULL
median.tau.nojump.3RWGD.teleosts<-nodes.contrast.3RWGD.tau.nojumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.nojump.3RWGD.teleosts$pic.round<-paste0("Median = ",round(median.tau.nojump.3RWGD.teleosts$Tau.abs,4), sep="")
median.tau.nojump.3RWGD.teleosts$count<- paste0(median.tau.nojump.3RWGD.teleosts$Event,"\n","(n = ",median.tau.nojump.3RWGD.teleosts$freq,")")


##Testing on average expression levels for trees supporting jumps in 3RWGD trees in teleosts
speciation.exp.jump.3RWGD.fish <-speciation.contrast(nodes.contrast.3RWGD.exp.jumps.fish ,"Exp.abs")
duplication.exp.jump.3RWGD.fish1 <-duplication.contrast.3R(nodes.contrast.3RWGD.exp.jumps.fish,"Exp.abs")
duplication.exp.jump.3RWGD.fish2 <-duplication.contrast(nodes.contrast.3RWGD.exp.jumps.fish,"Exp.abs")
#duplication.exp.jump.3RWGD.fish2 <-nodes.contrast.3RWGD.exp.jumps.fish$Exp.abs[which(nodes.contrast.3RWGD.exp.jumps.fish$Event=="SSD")]


##Stats on average expression levels for trees supporting jumps in 3RWGD trees in teleosts
Pval.exp.jump.3RWGD.fish1 <- two.tailed.wilcox(duplication.exp.jump.3RWGD.fish1,speciation.exp.jump.3RWGD.fish)
Pval.exp.jump.3RWGD.fish2 <- two.tailed.wilcox(duplication.exp.jump.3RWGD.fish2,speciation.exp.jump.3RWGD.fish)

## Median data for trees supporting jump in average expression levels for 3RWGD trees in teleosts
levels(nodes.contrast.3RWGD.exp.jumps.fish$Event)<-c("Speciation", "SSD", "FishWGD")
median.exp.jump.3RWGD.teleosts<-NULL
median.exp.jump.3RWGD.teleosts<-nodes.contrast.3RWGD.exp.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.3RWGD.teleosts$pic.round<-paste0("Median = ",round(median.exp.jump.3RWGD.teleosts$Exp.abs,4), sep="")
median.exp.jump.3RWGD.teleosts$count<- paste0(median.exp.jump.3RWGD.teleosts$Event,"\n","(n = ",median.exp.jump.3RWGD.teleosts$freq,")")

##Testing on average expression levels after removal of nodes for which daughter branches support jumps in 3RWGD trees in teleosts
speciation.exp.3RWGD.jump.fish.removing.jumpnode <-speciation.contrast(nodes.contrast.3RWGD.exp.jumps.fish.removing.jump.node,"Exp.abs")
duplication.exp.3RWGD.jump.fish.removing.jumpnode1 <-duplication.contrast.3R(nodes.contrast.3RWGD.exp.jumps.fish.removing.jump.node,"Exp.abs")
duplication.exp.3RWGD.jump.fish.removing.jumpnode2 <-duplication.contrast(nodes.contrast.3RWGD.exp.jumps.fish.removing.jump.node,"Exp.abs")
#duplication.exp.3RWGD.jump.fish.removing.jumpnode2 <-nodes.contrast.3RWGD.exp.jumps.fish.removing.jump.node$Exp.abs[which(nodes.contrast.3RWGD.exp.jumps.fish.removing.jump.node$Event=="SSD")]

##Stats for trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in 3RWGD trees in teleosts
Pval.exp.3RWGD.jump.fish.removing.jumpnode1 <- two.tailed.wilcox(duplication.exp.3RWGD.jump.fish.removing.jumpnode1,speciation.exp.3RWGD.jump.fish.removing.jumpnode)
Pval.exp.3RWGD.jump.fish.removing.jumpnode2 <- two.tailed.wilcox(duplication.exp.3RWGD.jump.fish.removing.jumpnode2,speciation.exp.3RWGD.jump.fish.removing.jumpnode)

## Median data of trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in 3RWGD trees in teleosts
levels(nodes.contrast.3RWGD.exp.jumps.fish.removing.jump.node$Event)<-c("Speciation", "SSD", "FishWGD")
median.exp.3RWGD.removing.jumpnode.teleosts<-NULL
median.exp.3RWGD.removing.jumpnode.teleosts<-nodes.contrast.3RWGD.exp.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.3RWGD.removing.jumpnode.teleosts$pic.round<-paste0("Median = ",round(median.exp.3RWGD.removing.jumpnode.teleosts$Exp.abs,4), sep="")
median.exp.3RWGD.removing.jumpnode.teleosts$count<- paste0(median.exp.3RWGD.removing.jumpnode.teleosts$Event,"\n","(n = ",median.exp.3RWGD.removing.jumpnode.teleosts$freq,")")

##Testing on average expression levels for trees not supporting jump in 3RWGD trees in teleosts
speciation.exp.3RWGD.nojump.fish <-speciation.contrast(nodes.contrast.3RWGD.exp.nojumps.fish ,"Exp.abs")
duplication.exp.3RWGD.nojump.fish1 <-duplication.contrast.3R(nodes.contrast.3RWGD.exp.nojumps.fish,"Exp.abs")
duplication.exp.3RWGD.nojump.fish2 <-duplication.contrast(nodes.contrast.3RWGD.exp.nojumps.fish,"Exp.abs")
#duplication.exp.3RWGD.nojump.fish2 <-nodes.contrast.3RWGD.exp.nojumps.fish$Exp.abs[which(nodes.contrast.3RWGD.exp.nojumps.fish$Event=="SSD")]

##Stats on average expression levels for for trees not supporting jump in 3RWGD trees in teleosts
Pval.exp.3RWGD.nojump.fish1 <- two.tailed.wilcox(duplication.exp.3RWGD.nojump.fish1,speciation.exp.3RWGD.nojump.fish)
Pval.exp.3RWGD.nojump.fish2 <- two.tailed.wilcox(duplication.exp.3RWGD.nojump.fish2,speciation.exp.3RWGD.nojump.fish)


## Median data for trees not supporting jump in average expression levels in 3RWGD trees in teleosts
levels(nodes.contrast.3RWGD.exp.nojumps.fish$Event)<-c("Speciation", "SSD", "FishWGD")
median.exp.3RWGD.nojump.teleosts<-NULL
median.exp.3RWGD.nojump.teleosts<-nodes.contrast.3RWGD.exp.nojumps.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.3RWGD.nojump.teleosts$pic.round<-paste0("Median = ",round(median.exp.3RWGD.nojump.teleosts$Exp.abs,4), sep="")
median.exp.3RWGD.nojump.teleosts$count<- paste0(median.exp.3RWGD.nojump.teleosts$Event,"\n","(n = ",median.exp.3RWGD.nojump.teleosts$freq,")")


```



```{r Fig_S11, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S11**: The ortholog conjecture (OC) tests on trusted 3RWGD trees to analyze the jump results in $\\tau$ and in average gene expression levels for teleosts. We used a strict branch jump posterior cutoff of $\\ge$ 0.7 as an evidence of trait jump. Only teleosts-specific clades were used for the analyses. The ortholog conjecture tests on (A) ", length(Calibrated.3RWGD.tau.jump), " calibrated contrasts standardized trusted 3RWGD trees supporting rapid jumps in $\\tau$, and on (B) ", length(Calibrated.3RWGD.exp.jump)," trusted 3RWGD trees supporting jumps in average gene expression levels. (C)-(D) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for ", length(Calibrated.3RWGD.tau.jump), " and ", length(Calibrated.3RWGD.exp.jump), " trusted 3RWGD trees, respectively. (E)-(F) The ortholog conjecture tests on ",length(Calibrated.3RWGD.tau.nojump)," and ", length(Calibrated.3RWGD.exp.nojump)," trusted 3RWGD trees for which the trait jump model is not supported for $\\tau$ and for average gene expression levels, respectively. FishWGD: ohnologs, other duplicates refer to small scale duplicates. P values are from Wilcoxon two-tailed tests. ")}

dodge <- position_dodge(width = 0.31)

# The ortholog conjecture test on tau for trees supporting jump in 3RWGD teleosts 
plotS11A<- boxplot.new.3RWGD(nodes.contrast.3RWGD.tau.jumps.fish,Pval.tau.jump.3RWGD.fish1,Pval.tau.jump.3RWGD.fish2,"Tau.3RWGD",median.tau.jump.3RWGD.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("A. OC test on trees supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.jump.3RWGD.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.043, ysize = 0.005)

# The ortholog conjecture test on average expression levels for trees supporting jump in 3RWGD teleosts 
plotS11B<- boxplot.new.3RWGD(nodes.contrast.3RWGD.exp.jumps.fish,Pval.exp.jump.3RWGD.fish1,Pval.exp.jump.3RWGD.fish2,"Exp.3RWGD",median.exp.jump.3RWGD.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. OC test on trees supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.3RWGD.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.25, ysize = 0.027)

# The ortholog conjecture test on tau after removal of node contrasts for which the daughter branches experience jumps in tau in 3RWGD teleosts 
plotS11C<- boxplot.new.3RWGD(nodes.contrast.3RWGD.tau.jumps.fish.removing.jump.node,Pval.tau.jump.3RWGD.fish.removing.jumpnode1,Pval.tau.jump.3RWGD.fish.removing.jumpnode2,"Tau.3RWGD",median.tau.removing.3RWGD.jumpnode.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("C. OC test after removing nodes supporting ", tau," jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.removing.3RWGD.jumpnode.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.043, ysize = 0.005)


# The ortholog conjecture test on average expression levels after removal of node contrasts for which the daughter branches experience jumps in average expression levels in 3RWGD teleosts 
plotS11D<- boxplot.new.3RWGD(nodes.contrast.3RWGD.exp.jumps.fish.removing.jump.node,Pval.exp.3RWGD.jump.fish.removing.jumpnode1,Pval.exp.3RWGD.jump.fish.removing.jumpnode2,"Exp.3RWGD",median.exp.3RWGD.removing.jumpnode.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. OC test after removing nodes supporting exp jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.3RWGD.removing.jumpnode.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.25, ysize = 0.027)

# The ortholog conjecture test on tau for trees do not provide support for jumps in 3RWGD teleosts
plotS11E<- boxplot.new.3RWGD(nodes.contrast.3RWGD.tau.nojumps.fish,Pval.tau.nojump.3RWGD.fish1,Pval.tau.nojump.3RWGD.fish2,"Tau.3RWGD",median.tau.nojump.3RWGD.teleosts) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("E. OC test on trees not supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.nojump.3RWGD.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.043, ysize = 0.005)

# The ortholog conjecture test on average expression levels for trees do not provide support for jumps in 3RWGD teleosts
plotS11F<- boxplot.new.3RWGD(nodes.contrast.3RWGD.exp.nojumps.fish,Pval.exp.3RWGD.nojump.fish1,Pval.exp.3RWGD.nojump.fish2,"Exp.3RWGD",median.exp.3RWGD.nojump.teleosts) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. OC test on trees not supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.3RWGD.nojump.teleosts$count)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.25, ysize = 0.027)

gridExtra::grid.arrange(plotS11A,plotS11B,plotS11C,plotS11D,plotS11E,plotS11F,ncol=2, nrow=3)

```



```{r OC_analyses_on_3RWGDs_jump_trees_15spe, eval=T,echo=F, message=FALSE, warning=FALSE, cache=T}

##Testing on tau for 3RWGD trees supporting jumps in trusted 3RWGD vertebrates
speciation.tau.jump.3RWGD.15spe <-speciation.contrast(nodes.contrast.3RWGD.tau.jumps.15spe ,"Tau.abs")
duplication.tau.jump.3RWGD.15spe1 <-duplication.contrast.3R(nodes.contrast.3RWGD.tau.jumps.15spe,"Tau.abs")
duplication.tau.jump.3RWGD.15spe2 <-duplication.contrast(nodes.contrast.3RWGD.tau.jumps.15spe,"Tau.abs")
#duplication.tau.jump.3RWGD.15spe2 <-nodes.contrast.3RWGD.tau.jumps.15spe$Tau.abs[which(nodes.contrast.3RWGD.tau.jumps.15spe$Event=="SSD")]

##Stats on tau for 3RWGD trees supporting jumps in trusted 3RWGD vertebrates
Pval.tau.jump.3RWGD.15spe1 <- two.tailed.wilcox(duplication.tau.jump.3RWGD.15spe1,speciation.tau.jump.3RWGD.15spe)
Pval.tau.jump.3RWGD.15spe2 <- two.tailed.wilcox(duplication.tau.jump.3RWGD.15spe2,speciation.tau.jump.3RWGD.15spe)

## Median data for 3RWGD trees supporting jump in tau for trusted 3RWGD vertebrates
levels(nodes.contrast.3RWGD.tau.jumps.15spe$Event)<-c("Speciation", "SSD", "FishWGD")
median.tau.jump.3RWGD.15spe<-NULL
median.tau.jump.3RWGD.15spe<-nodes.contrast.3RWGD.tau.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.jump.3RWGD.15spe$pic.round<-paste0("Median = ",round(median.tau.jump.3RWGD.15spe$Tau.abs,4), sep="")
median.tau.jump.3RWGD.15spe$count<- paste0(median.tau.jump.3RWGD.15spe$Event,"\n","(n = ",median.tau.jump.3RWGD.15spe$freq,")")

##Testing on tau after removal of nodes for which daughter branches support jumps in 3RWGD trees in trusted 3RWGD vertebrates
speciation.tau.jump.3RWGD.15spe.removing.jumpnode <-speciation.contrast(nodes.contrast.3RWGD.tau.jumps.15spe.removing.jump.node ,"Tau.abs")
duplication.tau.jump.3RWGD.15spe.removing.jumpnode1 <-duplication.contrast.3R(nodes.contrast.3RWGD.tau.jumps.15spe.removing.jump.node,"Tau.abs")
duplication.tau.jump.3RWGD.15spe.removing.jumpnode2 <-duplication.contrast(nodes.contrast.3RWGD.tau.jumps.15spe.removing.jump.node,"Tau.abs")
#duplication.tau.jump.3RWGD.15spe.removing.jumpnode2 <-nodes.contrast.3RWGD.tau.jumps.15spe.removing.jump.node$Tau.abs[which(nodes.contrast.3RWGD.tau.jumps.15spe.removing.jump.nod$Event=="SSD")]

##Stats for trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in trusted 3RWGD vertebrates
Pval.tau.jump.3RWGD.15spe.removing.jumpnode1 <- two.tailed.wilcox(duplication.tau.jump.3RWGD.15spe.removing.jumpnode1,speciation.tau.jump.3RWGD.15spe.removing.jumpnode)

Pval.tau.jump.3RWGD.15spe.removing.jumpnode2 <- two.tailed.wilcox(duplication.tau.jump.3RWGD.15spe.removing.jumpnode2,speciation.tau.jump.3RWGD.15spe.removing.jumpnode)

## Median data of trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in trusted 3RWGD vertebrates
levels(nodes.contrast.3RWGD.tau.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD", "FishWGD")
median.tau.removing.3RWGD.jumpnode.15spe<-NULL
median.tau.removing.3RWGD.jumpnode.15spe<-nodes.contrast.3RWGD.tau.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.removing.3RWGD.jumpnode.15spe$pic.round<-paste0("Median = ",round(median.tau.removing.3RWGD.jumpnode.15spe$Tau.abs,4), sep="")
median.tau.removing.3RWGD.jumpnode.15spe$count<- paste0(median.tau.removing.3RWGD.jumpnode.15spe$Event,"\n","(n = ",median.tau.removing.3RWGD.jumpnode.15spe$freq,")")

##Testing on tau for trees not supporting jump in 3RWGD trees in trusted 3RWGD vertebrates
speciation.tau.nojump.3RWGD.15spe <-speciation.contrast(nodes.contrast.3RWGD.tau.nojumps.15spe ,"Tau.abs")
duplication.tau.nojump.3RWGD.15spe1 <-duplication.contrast.3R(nodes.contrast.3RWGD.tau.nojumps.15spe,"Tau.abs")
duplication.tau.nojump.3RWGD.15spe2 <-duplication.contrast(nodes.contrast.3RWGD.tau.nojumps.15spe,"Tau.abs")
#duplication.tau.nojump.3RWGD.15spe2 <-nodes.contrast.3RWGD.tau.nojumps.15spe$Tau.abs[which(nodes.contrast.3RWGD.tau.nojumps.15spe$Event=="SSD")]


##Stats on tau for for trees not supporting jump in trusted 3RWGD vertebrates
Pval.tau.nojump.3RWGD.15spe1 <- two.tailed.wilcox(duplication.tau.nojump.3RWGD.15spe1,speciation.tau.nojump.3RWGD.15spe)
Pval.tau.nojump.3RWGD.15spe2 <- two.tailed.wilcox(duplication.tau.nojump.3RWGD.15spe2,speciation.tau.nojump.3RWGD.15spe)

## Median data for trees not supporting jump in 3RWGD trees in trusted 3RWGD vertebrates
levels(nodes.contrast.3RWGD.tau.nojumps.15spe$Event)<-c("Speciation", "SSD", "FishWGD")
median.tau.nojump.3RWGD.15spe<-NULL
median.tau.nojump.3RWGD.15spe<-nodes.contrast.3RWGD.tau.nojumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.nojump.3RWGD.15spe$pic.round<-paste0("Median = ",round(median.tau.nojump.3RWGD.15spe$Tau.abs,4), sep="")
median.tau.nojump.3RWGD.15spe$count<- paste0(median.tau.nojump.3RWGD.15spe$Event,"\n","(n = ",median.tau.nojump.3RWGD.15spe$freq,")")


##Testing on average expression levels for trees supporting jumps in trusted 3RWGD vertebrates
speciation.exp.jump.3RWGD.15spe <-speciation.contrast(nodes.contrast.3RWGD.exp.jumps.15spe ,"Exp.abs")
duplication.exp.jump.3RWGD.15spe1 <-duplication.contrast.3R(nodes.contrast.3RWGD.exp.jumps.15spe,"Exp.abs")
duplication.exp.jump.3RWGD.15spe2 <-duplication.contrast(nodes.contrast.3RWGD.exp.jumps.15spe,"Exp.abs")
#duplication.exp.jump.3RWGD.15spe2 <-nodes.contrast.3RWGD.exp.jumps.15spe$Exp.abs[which(nodes.contrast.3RWGD.exp.jumps.15spe$Event=="SSD")]

##Stats on average expression levels for trees supporting jumps in trusted 3RWGD vertebrates
Pval.exp.jump.3RWGD.15spe1 <- two.tailed.wilcox(duplication.exp.jump.3RWGD.15spe1,speciation.exp.jump.3RWGD.15spe)
Pval.exp.jump.3RWGD.15spe2 <- two.tailed.wilcox(duplication.exp.jump.3RWGD.15spe2,speciation.exp.jump.3RWGD.15spe)

## Median data for trees supporting jump in average expression levels for 3RWGD trees in trusted 3RWGD vertebrates
levels(nodes.contrast.3RWGD.exp.jumps.15spe$Event)<-c("Speciation", "SSD", "FishWGD")
median.exp.jump.3RWGD.15spe<-NULL
median.exp.jump.3RWGD.15spe<-nodes.contrast.3RWGD.exp.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.3RWGD.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.3RWGD.15spe$Exp.abs,4), sep="")
median.exp.jump.3RWGD.15spe$count<- paste0(median.exp.jump.3RWGD.15spe$Event,"\n","(n = ",median.exp.jump.3RWGD.15spe$freq,")")

##Testing on average expression levels after removal of nodes for which daughter branches support jumps in trusted 3RWGD vertebrates
speciation.exp.3RWGD.jump.15spe.removing.jumpnode <-speciation.contrast(nodes.contrast.3RWGD.exp.jumps.15spe.removing.jump.node,"Exp.abs")
duplication.exp.3RWGD.jump.15spe.removing.jumpnode1 <-duplication.contrast.3R(nodes.contrast.3RWGD.exp.jumps.15spe.removing.jump.node,"Exp.abs")
duplication.exp.3RWGD.jump.15spe.removing.jumpnode2 <-duplication.contrast(nodes.contrast.3RWGD.exp.jumps.15spe.removing.jump.node,"Exp.abs")
#duplication.exp.3RWGD.jump.15spe.removing.jumpnode2 <-nodes.contrast.3RWGD.exp.jumps.15spe.removing.jump.node$Exp.abs[which(nodes.contrast.3RWGD.exp.jumps.15spe.removing.jump.node$Event=="SSD")]


##Stats for trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in trusted 3RWGD vertebrates
Pval.exp.3RWGD.jump.15spe.removing.jumpnode1 <- two.tailed.wilcox(duplication.exp.3RWGD.jump.15spe.removing.jumpnode1,speciation.exp.3RWGD.jump.15spe.removing.jumpnode)
Pval.exp.3RWGD.jump.15spe.removing.jumpnode2 <- two.tailed.wilcox(duplication.exp.3RWGD.jump.15spe.removing.jumpnode2,speciation.exp.3RWGD.jump.15spe.removing.jumpnode)

## Median data of trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in trusted 3RWGD vertebrates
levels(nodes.contrast.3RWGD.exp.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD", "FishWGD")
median.exp.3RWGD.removing.jumpnode.15spe<-NULL
median.exp.3RWGD.removing.jumpnode.15spe<-nodes.contrast.3RWGD.exp.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.3RWGD.removing.jumpnode.15spe$pic.round<-paste0("Median = ",round(median.exp.3RWGD.removing.jumpnode.15spe$Exp.abs,4), sep="")
median.exp.3RWGD.removing.jumpnode.15spe$count<- paste0(median.exp.3RWGD.removing.jumpnode.15spe$Event,"\n","(n = ",median.exp.3RWGD.removing.jumpnode.15spe$freq,")")

##Testing on average expression levels for trees not supporting jump in trusted 3RWGD vertebrates
speciation.exp.3RWGD.nojump.15spe <-speciation.contrast(nodes.contrast.3RWGD.exp.nojumps.15spe ,"Exp.abs")
duplication.exp.3RWGD.nojump.15spe1 <-duplication.contrast.3R(nodes.contrast.3RWGD.exp.nojumps.15spe,"Exp.abs")
duplication.exp.3RWGD.nojump.15spe2 <-duplication.contrast(nodes.contrast.3RWGD.exp.nojumps.15spe,"Exp.abs")
#duplication.exp.3RWGD.nojump.15spe2 <-nodes.contrast.3RWGD.exp.nojumps.15spe$Exp.abs[which(nodes.contrast.3RWGD.exp.nojumps.15spe$Event=="SSD")]


##Stats on average expression levels for for trees not supporting jump in trusted 3RWGD vertebrates

Pval.exp.3RWGD.nojump.15spe1 <- two.tailed.wilcox(duplication.exp.3RWGD.nojump.15spe1,speciation.exp.3RWGD.nojump.15spe)
Pval.exp.3RWGD.nojump.15spe2 <- two.tailed.wilcox(duplication.exp.3RWGD.nojump.15spe2,speciation.exp.3RWGD.nojump.15spe)


## Median data for trees not supporting jump in average expression levels in trusted 3RWGD vertebrates
levels(nodes.contrast.3RWGD.exp.nojumps.15spe$Event)<-c("Speciation", "SSD", "FishWGD")
median.exp.3RWGD.nojump.15spe<-NULL
median.exp.3RWGD.nojump.15spe<-nodes.contrast.3RWGD.exp.nojumps.15spe%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.3RWGD.nojump.15spe$pic.round<-paste0("Median = ",round(median.exp.3RWGD.nojump.15spe$Exp.abs,4), sep="")
median.exp.3RWGD.nojump.15spe$count<- paste0(median.exp.3RWGD.nojump.15spe$Event,"\n","(n = ",median.exp.3RWGD.nojump.15spe$freq,")")


```



```{r Fig_S12, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S12**: The ortholog conjecture (OC) tests on trusted 3RWGD trees to analyze the jump results in $\\tau$ and in average gene expression levels for 15 vertebrates. We used a strict branch jump posterior cutoff of $\\ge$ 0.7 as an evidence of trait jump. The ortholog conjecture tests on (A)  ", length(Calibrated.3RWGD.tau.jump), " calibrated contrasts standardized trusted 3RWGD trees supporting rapid jumps in $\\tau$, and on (B) ", length(Calibrated.3RWGD.exp.jump)," trusted 3RWGD trees supporting jumps in average gene expression levels. (C)-(D) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for ", length(Calibrated.3RWGD.tau.jump), " and ", length(Calibrated.3RWGD.exp.jump), " trusted 3RWGD trees, respectively. (E)-(F) The ortholog conjecture tests on ",length(Calibrated.3RWGD.tau.nojump)," and ", length(Calibrated.3RWGD.exp.nojump)," trusted 3RWGD trees for which the trait jump model is not supported for $\\tau$ and for average gene expression levels, respectively. FishWGD: ohnologs, other duplicates refer to small scale duplicates. P values are from Wilcoxon two-tailed tests. ")}

dodge <- position_dodge(width = 0.31)

# The ortholog conjecture test on tau for trees supporting jump in 3RWGD vertebrates 
plotS12A<- boxplot.new.3RWGD(nodes.contrast.3RWGD.tau.jumps.15spe,Pval.tau.jump.3RWGD.15spe1,Pval.tau.jump.3RWGD.15spe2,"Tau.3RWGD",median.tau.jump.3RWGD.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("A. OC test on trees supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.jump.3RWGD.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.037, ysize = 0.004)

# The ortholog conjecture test on average expression levels for trees supporting jump in 3RWGD vertebrates  
plotS12B<- boxplot.new.3RWGD(nodes.contrast.3RWGD.exp.jumps.15spe,Pval.exp.jump.3RWGD.15spe1,Pval.exp.jump.3RWGD.15spe2,"Exp.3RWGD",median.exp.jump.3RWGD.15spe) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. OC test on trees supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.3RWGD.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.26, ysize = 0.02,color = "black") +
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.027)

# The ortholog conjecture test on tau after removal of node contrasts for which the daughter branches experience jumps in tau in 3RWGD vertebrates  
plotS12C<- boxplot.new.3RWGD(nodes.contrast.3RWGD.tau.jumps.15spe.removing.jump.node,Pval.tau.jump.3RWGD.15spe.removing.jumpnode1,Pval.tau.jump.3RWGD.15spe.removing.jumpnode2,"Tau.3RWGD",median.tau.removing.3RWGD.jumpnode.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("C. Test after removing nodes supporting ", tau," jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.removing.3RWGD.jumpnode.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.037, ysize = 0.004)


# The ortholog conjecture test on average expression levels after removal of node contrasts for which the daughter branches experience jumps in average expression levels in 3RWGD vertebrates  
plotS12D<- boxplot.new.3RWGD(nodes.contrast.3RWGD.exp.jumps.15spe.removing.jump.node,Pval.exp.3RWGD.jump.15spe.removing.jumpnode1,Pval.exp.3RWGD.jump.15spe.removing.jumpnode2,"Exp.3RWGD",median.exp.3RWGD.removing.jumpnode.15spe) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Test after removing nodes supporting exp jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.3RWGD.removing.jumpnode.15spe$count)+       add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.26, ysize = 0.02,color = "black") +
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.027)

# The ortholog conjecture test on tau for trees do not provide support for jumps in 3RWGD vertebrates 
plotS12E<- boxplot.new.3RWGD(nodes.contrast.3RWGD.tau.nojumps.15spe,Pval.tau.nojump.3RWGD.15spe1,Pval.tau.nojump.3RWGD.15spe2,"Tau.3RWGD",median.tau.nojump.3RWGD.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("E. OC test on trees not supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.nojump.3RWGD.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.037, ysize = 0.004)

# The ortholog conjecture test on average expression levels for trees do not provide support for jumps in 3RWGD vertebrates
plotS12F<- boxplot.new.3RWGD(nodes.contrast.3RWGD.exp.nojumps.15spe,Pval.exp.3RWGD.nojump.15spe1,Pval.exp.3RWGD.nojump.15spe2,"Exp.3RWGD",median.exp.3RWGD.nojump.15spe) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. OC test on trees not supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.3RWGD.nojump.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 3.3, y = 0.26, ysize = 0.02,color = "black") +
  add_phylopic(fish_pic, alpha = 1,x = 3.3, y = 0.22, ysize = 0.027)


gridExtra::grid.arrange(plotS12A,plotS12B,plotS12C,plotS12D,plotS12E,plotS12F,ncol=2, nrow=3)

```



```{r OC_analyses_on_SSDs_jump_trees_15species, eval=T,echo=F, message=FALSE, warning=FALSE, cache=T}

##Testing on tau for SSD trees supporting jumps in 15 vertebrates
speciation.tau.jump.SSD.15spe <-speciation.contrast(nodes.contrast.SSD.tau.jumps.15spe ,"Tau.abs")
duplication.tau.jump.SSD.15spe <-duplication.contrast(nodes.contrast.SSD.tau.jumps.15spe,"Tau.abs")

##Stats on tau for SSD trees supporting jumps in 15 vertebrates
Pval.tau.jump.SSD.15spe <- two.tailed.wilcox(duplication.tau.jump.SSD.15spe,speciation.tau.jump.SSD.15spe)

## Median data for SSD trees supporting jump in tau for 15 vertebrates
levels(nodes.contrast.SSD.tau.jumps.15spe$Event)<-c("Speciation", "SSD")
median.tau.jump.SSD.15spe<-NULL
median.tau.jump.SSD.15spe<-nodes.contrast.SSD.tau.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.jump.SSD.15spe$pic.round<-paste0("Median = ",round(median.tau.jump.SSD.15spe$Tau.abs,4), sep="")
median.tau.jump.SSD.15spe$count<- paste0(median.tau.jump.SSD.15spe$Event,"\n","(n = ",median.tau.jump.SSD.15spe$freq,")")

##Testing on tau after removal of nodes for which daughter branches support jumps in SSD trees in 15 vertebrates
speciation.tau.jump.SSD.15spe.removing.jumpnode <-speciation.contrast(nodes.contrast.SSD.tau.jumps.15spe.removing.jump.node ,"Tau.abs")
duplication.tau.jump.SSD.15spe.removing.jumpnode <-duplication.contrast(nodes.contrast.SSD.tau.jumps.15spe.removing.jump.node,"Tau.abs")

##Stats for trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in SSD trees in 15 vertebrates
Pval.tau.jump.SSD.15spe.removing.jumpnode <- two.tailed.wilcox(duplication.tau.jump.SSD.15spe.removing.jumpnode,speciation.tau.jump.SSD.15spe.removing.jumpnode)

## Median data of trees supporting jumps in tau after removal of nodes for which daughter branches support jumps in SSD trees in 15 vertebrates
levels(nodes.contrast.SSD.tau.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD")
median.tau.removing.SSD.jumpnode.15spe<-NULL
median.tau.removing.SSD.jumpnode.15spe<-nodes.contrast.SSD.tau.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.removing.SSD.jumpnode.15spe$pic.round<-paste0("Median = ",round(median.tau.removing.SSD.jumpnode.15spe$Tau.abs,4), sep="")
median.tau.removing.SSD.jumpnode.15spe$count<- paste0(median.tau.removing.SSD.jumpnode.15spe$Event,"\n","(n = ",median.tau.removing.SSD.jumpnode.15spe$freq,")")

##Testing on tau for trees not supporting jump in SSD trees in 15 vertebrates
speciation.tau.nojump.SSD.15spe <-speciation.contrast(nodes.contrast.SSD.tau.nojumps.15spe ,"Tau.abs")
duplication.tau.nojump.SSD.15spe <-duplication.contrast(nodes.contrast.SSD.tau.nojumps.15spe,"Tau.abs")

##Stats on tau for for trees not supporting jump in SSD trees in 15 vertebrates
Pval.tau.nojump.SSD.15spe <- two.tailed.wilcox(duplication.tau.nojump.SSD.15spe,speciation.tau.nojump.SSD.15spe)

## Median data for trees not supporting jump in SSD trees in 15 vertebrates
levels(nodes.contrast.SSD.tau.nojumps.15spe$Event)<-c("Speciation", "SSD")
median.tau.nojump.SSD.15spe<-NULL
median.tau.nojump.SSD.15spe<-nodes.contrast.SSD.tau.nojumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Tau.abs),
            Tau.abs=median(Tau.abs),
            freq = n())
median.tau.nojump.SSD.15spe$pic.round<-paste0("Median = ",round(median.tau.nojump.SSD.15spe$Tau.abs,4), sep="")
median.tau.nojump.SSD.15spe$count<- paste0(median.tau.nojump.SSD.15spe$Event,"\n","(n = ",median.tau.nojump.SSD.15spe$freq,")")


##Testing on average expression levels for trees supporting jumps in SSD trees in 15 vertebrates
speciation.exp.jump.SSD.15spe <-speciation.contrast(nodes.contrast.SSD.exp.jumps.15spe ,"Exp.abs")
duplication.exp.jump.SSD.15spe <-duplication.contrast(nodes.contrast.SSD.exp.jumps.15spe,"Exp.abs")

##Stats on average expression levels for trees supporting jumps in SSD trees in 15 vertebrates
Pval.exp.jump.SSD.15spe <- two.tailed.wilcox(duplication.exp.jump.SSD.15spe,speciation.exp.jump.SSD.15spe)

## Median data for trees supporting jump in average expression levels for SSD trees in 15 vertebrates
levels(nodes.contrast.SSD.exp.jumps.15spe$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.15spe<-NULL
median.exp.jump.SSD.15spe<-nodes.contrast.SSD.exp.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.15spe$Exp.abs,4), sep="")
median.exp.jump.SSD.15spe$count<- paste0(median.exp.jump.SSD.15spe$Event,"\n","(n = ",median.exp.jump.SSD.15spe$freq,")")

##Testing on average expression levels after removal of nodes for which daughter branches support jumps in SSD trees in 15 vertebrates
speciation.exp.SSD.jump.15spe.removing.jumpnode <-speciation.contrast(nodes.contrast.SSD.exp.jumps.15spe.removing.jump.node,"Exp.abs")
duplication.exp.SSD.jump.15spe.removing.jumpnode <-duplication.contrast(nodes.contrast.SSD.exp.jumps.15spe.removing.jump.node,"Exp.abs")

##Stats for trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in SSD trees in 15 vertebrates
Pval.exp.SSD.jump.15spe.removing.jumpnode <- two.tailed.wilcox(duplication.exp.SSD.jump.15spe.removing.jumpnode,speciation.exp.SSD.jump.15spe.removing.jumpnode)

## Median data of trees supporting jumps in average expression levels after removal of nodes for which daughter branches support jumps in SSD trees in 15 vertebrates
levels(nodes.contrast.SSD.exp.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.SSD.removing.jumpnode.15spe<-NULL
median.exp.SSD.removing.jumpnode.15spe<-nodes.contrast.SSD.exp.jumps.15spe.removing.jump.node%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.SSD.removing.jumpnode.15spe$pic.round<-paste0("Median = ",round(median.exp.SSD.removing.jumpnode.15spe$Exp.abs,4), sep="")
median.exp.SSD.removing.jumpnode.15spe$count<- paste0(median.exp.SSD.removing.jumpnode.15spe$Event,"\n","(n = ",median.exp.SSD.removing.jumpnode.15spe$freq,")")

##Testing on average expression levels for trees not supporting jump in SSD trees in 15 vertebrates
speciation.exp.SSD.nojump.15spe <-speciation.contrast(nodes.contrast.SSD.exp.nojumps.15spe ,"Exp.abs")
duplication.exp.SSD.nojump.15spe <-duplication.contrast(nodes.contrast.SSD.exp.nojumps.15spe,"Exp.abs")

##Stats on average expression levels for for trees not supporting jump in SSD trees in 15 vertebrates
Pval.exp.SSD.nojump.15spe <- two.tailed.wilcox(duplication.exp.SSD.nojump.15spe,speciation.exp.SSD.nojump.15spe)

## Median data for trees not supporting jump in average expression levels in SSD trees in 15 vertebrates
levels(nodes.contrast.SSD.exp.nojumps.15spe$Event)<-c("Speciation", "SSD")
median.exp.SSD.nojump.15spe<-NULL
median.exp.SSD.nojump.15spe<-nodes.contrast.SSD.exp.nojumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.SSD.nojump.15spe$pic.round<-paste0("Median = ",round(median.exp.SSD.nojump.15spe$Exp.abs,4), sep="")
median.exp.SSD.nojump.15spe$count<- paste0(median.exp.SSD.nojump.15spe$Event,"\n","(n = ",median.exp.SSD.nojump.15spe$freq,")")

```



```{r Fig_S13, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S13**: The ortholog conjecture (OC) tests on trusted SSD trees to analyze the jump results in $\\tau$ and in average gene expression levels for 15 vertebrates. We used a strict branch jump posterior cutoff of $\\ge$ 0.7 as an evidence of trait jump. The ortholog conjecture tests on (A) ", length(Calibrated.SSD.tau.jump), " calibrated contrasts standardized trusted SSD trees supporting rapid jumps in $\\tau$, and on (B) ", length(Calibrated.SSD.exp.jump)," trusted SSD trees supporting jumps in average gene expression levels. (C)-(D) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for ", length(Calibrated.SSD.tau.jump), " and ", length(Calibrated.SSD.exp.jump), " trusted SSD trees, respectively. (E)-(F) The ortholog conjecture tests on ",length(Calibrated.SSD.tau.nojump)," and ", length(Calibrated.SSD.exp.nojump)," trusted SSD trees for which the trait jump model is not supported for $\\tau$ and for average gene expression levels, respectively. *P* values are from Wilcoxon two-tailed tests. ")}

dodge <- position_dodge(width = 0.31)

# The ortholog conjecture test on tau for trees supporting jump in trusted SSD vertebrates
plotS13A<- boxplot.new(nodes.contrast.SSD.tau.jumps.15spe,Pval.tau.jump.SSD.15spe,"Tau",median.tau.jump.SSD.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("A. OC test on trees supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.jump.SSD.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004) 

# The ortholog conjecture test on average expression levels for trees supporting jump in trusted SSD vertebrates
plotS13B<- boxplot.new(nodes.contrast.SSD.exp.jumps.15spe,Pval.exp.jump.SSD.15spe,"Exp",median.exp.jump.SSD.15spe)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("B. OC test on trees supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.019)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025)

# The ortholog conjecture test on tau after removal of node contrasts for which the daughter branches experience jumps in tau in trusted SSD vertebrates
plotS13C<- boxplot.new(nodes.contrast.SSD.tau.jumps.15spe.removing.jump.node,Pval.tau.jump.SSD.15spe.removing.jumpnode,"Tau",median.tau.removing.SSD.jumpnode.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("C. Test after removing nodes supporting ", tau," jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.removing.SSD.jumpnode.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004) 



# The ortholog conjecture test on average expression levels after removal of node contrasts for which the daughter branches experience jumps in average expression levels in trusted SSD vertebrates 
plotS13D<- boxplot.new(nodes.contrast.SSD.exp.jumps.15spe.removing.jump.node,Pval.exp.SSD.jump.15spe.removing.jumpnode,"Exp",median.exp.SSD.removing.jumpnode.15spe) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Test after removing nodes supporting exp jump"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.SSD.removing.jumpnode.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.019)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025)

# The ortholog conjecture test on tau for trees do not provide support for jumps in trusted SSD vertebrates
plotS13E<- boxplot.new(nodes.contrast.SSD.tau.nojumps.15spe,Pval.tau.nojump.SSD.15spe ,"Tau",median.tau.nojump.SSD.15spe) +
  coord_cartesian(ylim=c(0, 0.05)) +
  ggtitle(expression(bold(paste("E. OC test on trees not supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.nojump.SSD.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.043, ysize = 0.003)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.037, ysize = 0.004) 


# The ortholog conjecture test on average expression levels for trees do not provide support for jumps in trusted SSD vertebrates
plotS13F<- boxplot.new(nodes.contrast.SSD.exp.nojumps.15spe,Pval.exp.SSD.nojump.15spe ,"Exp",median.exp.SSD.nojump.15spe) +
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. OC test on trees not supporting jumps in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.SSD.nojump.15spe$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.019)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.025) 


gridExtra::grid.arrange(plotS13A,plotS13B,plotS13C,plotS13D,plotS13E,plotS13F,ncol=2, nrow=3)

```



```{r Fig_S14, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S14**: Analyses of rapid jumps in brain expression for trusted SSD trees. We used a strict posterior probability cutoff of $\\ge$ 0.7 as an evidence of trait jump. (A)-(B) Proportions of jumps in brain expressions following speciation and duplication events for teleosts and for vertebrates using 1752 trait jump supporting trees of Fig. 7A. Since proportion of shifts in traits per branch of events is estimated for each tree, paired Wilcoxon test is used to compare the difference for (A)-(B). (C)-(D) The ortholog conjecture (OC) tests on 1752 trusted SSD trees to analyze the jump results in brain expressions for teleosts and for 15 vertebrates, respectively. (E)-(F) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for 1752 trusted SSD trees for 5 teleosts and for 15 vertebrates, respectively. *P* values are from Wilcoxon two-tailed tests for (C)-(F). ‘exp’: expression levels. ")}

## Proportions of jumps in brain for teleosts
pval.jump.SSD.brain.fish <- paired.wilcox(info.jump.trees.SSD.brain.final1.fish$spe.jump.prob,info.jump.trees.SSD.brain.final1.fish$dup.jump.prob) #p-value < 2.2e-16

## For median data
dtfr.SSD.brain.fish<-NULL
dtfr.SSD.brain.fish<-data.frame(prop.jump=info.jump.trees.SSD.brain.final1.fish$spe.jump.prob,events="speciation")
dtfr.SSD.brain.fish<-rbind(dtfr.SSD.brain.fish,data.frame(prop.jump=info.jump.trees.SSD.brain.final1.fish$dup.jump.prob,events="duplication"))
dtfr.SSD.brain.fish$Exp.abs <-abs(dtfr.SSD.brain.fish$prop.jump)
dtfr.SSD.brain.fish$Event<-factor(dtfr.SSD.brain.fish$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.brain.fish$Event)<-c("Speciation","SSD")

median.data.jump.SSD.brain.fish<-NULL
median.data.jump.SSD.brain.fish<- dtfr.SSD.brain.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.brain.fish$pic.round<-paste0("Median = ",round(median.data.jump.SSD.brain.fish$Exp.abs,4), sep="")
median.data.jump.SSD.brain.fish$count<- paste0(median.data.jump.SSD.brain.fish$Event,"\n","(n = ",median.data.jump.SSD.brain.fish$freq,")")

plotS14A<-vioplot.new2(dtfr.SSD.brain.fish,pval.jump.SSD.brain.fish,"Exp",median.data.jump.SSD.brain.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("A. Jump in brain exp for 5 teleosts"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
 # annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.brain.fish, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.brain.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.brain.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.brain.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.brain.fish$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.1, color = "black")  
 

## Proportions of jumps in brain for 15 species
pval.jump.SSD.brain <- paired.wilcox(info.jump.brain.final1.SSD.all$spe.jump.prob,info.jump.brain.final1.SSD.all$dup.jump.prob) #p-value = 2.2e-16

## For median data
dtfr.SSD.brain.15spe<-NULL
dtfr.SSD.brain.15spe<-data.frame(prop.jump=info.jump.brain.final1.SSD.all$spe.jump.prob,events="speciation")
dtfr.SSD.brain.15spe<-rbind(dtfr.SSD.brain.15spe,data.frame(prop.jump=info.jump.brain.final1.SSD.all$dup.jump.prob,events="duplication"))
dtfr.SSD.brain.15spe$Exp.abs <-abs(dtfr.SSD.brain.15spe$prop.jump)
dtfr.SSD.brain.15spe$Event<-factor(dtfr.SSD.brain.15spe$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.brain.15spe$Event)<-c("Speciation","SSD")

library(tidyverse)
median.data.jump.SSD.brain<-NULL
median.data.jump.SSD.brain<- dtfr.SSD.brain.15spe%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.brain$pic.round<-paste0("Median = ",round(median.data.jump.SSD.brain$Exp.abs,4), sep="")
median.data.jump.SSD.brain$count<- paste0(median.data.jump.SSD.brain$Event,"\n","(n = ",median.data.jump.SSD.brain$freq,")")

plotS14B<-vioplot.new2(dtfr.SSD.brain.15spe,pval.jump.SSD.brain,"Exp",median.data.jump.SSD.brain) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("B. Jump in brain exp 15 species"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.brain, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.brain,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.brain,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.brain,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.brain.fish$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.07, color = "black")+
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.82, ysize = 0.1, color = "black") 


# The ortholog conjecture test on brain expressions for trees supporting jump in trusted SSD teleosts

brain.speciation.fish<-speciation.contrast(nodes.contrast.SSD.brain.jumps.fish,"pic")
brain.duplication.fish<-duplication.contrast(nodes.contrast.SSD.brain.jumps.fish,"pic")
Pval.exp.jump.SSD.brain.fish<-two.tailed.wilcox(brain.speciation.fish,brain.duplication.fish)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.brain.jumps.fish$Event)<-c("Speciation","SSD")
median.exp.jump.SSD.brain.fish<-NULL
median.exp.jump.SSD.brain.fish<-nodes.contrast.SSD.brain.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.brain.fish$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.brain.fish$Exp.abs,4), sep="")
median.exp.jump.SSD.brain.fish$count<- paste0(median.exp.jump.SSD.brain.fish$Event,"\n","(n = ",median.exp.jump.SSD.brain.fish$freq,")")


plotS14C<- boxplot.new(nodes.contrast.SSD.brain.jumps.fish,Pval.exp.jump.SSD.brain.fish,"Exp",median.exp.jump.SSD.brain.fish)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("C. Jump in brain exp for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.brain.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test on brain expressions for trees supporting jump in trusted SSD teleosts

brain.speciation.15spe<-speciation.contrast(nodes.contrast.SSD.brain.jumps.15spe,"pic")
brain.duplication.15spe<-duplication.contrast(nodes.contrast.SSD.brain.jumps.15spe,"pic")
Pval.exp.jump.SSD.brain.15spe<-two.tailed.wilcox(brain.speciation.15spe,brain.duplication.15spe)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.brain.jumps.15spe$Event)<-c("Speciation","SSD")
median.exp.jump.SSD.brain.15spe<-NULL
median.exp.jump.SSD.brain.15spe<-nodes.contrast.SSD.brain.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.brain.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.brain.15spe$Exp.abs,4), sep="")
median.exp.jump.SSD.brain.15spe$count<- paste0(median.exp.jump.SSD.brain.15spe$Event,"\n","(n = ",median.exp.jump.SSD.brain.15spe$freq,")")


plotS14D<- boxplot.new(nodes.contrast.SSD.brain.jumps.15spe,Pval.exp.jump.SSD.brain.15spe,"Exp",median.exp.jump.SSD.brain.15spe)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Jump in brain exp for 15 species"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.brain.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)


# The ortholog conjecture test after removing nodes supporting jump for brain in trusted SSD teleosts

brain.speciation.fish.after.node.removal<-speciation.contrast(nodes.contrast.SSD.brain.jumps.fish.removing.jump.node,"pic")
brain.duplication.fish.after.node.removal<-duplication.contrast(nodes.contrast.SSD.brain.jumps.fish.removing.jump.node,"pic")
Pval.exp.jump.SSD.brain.fish.after.node.removal<-two.tailed.wilcox(brain.speciation.fish.after.node.removal,brain.duplication.fish.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.brain.jumps.fish.removing.jump.node$Event)<-c("Speciation","SSD")
median.exp.jump.SSD.brain.fish.after.node.removal<-NULL
median.exp.jump.SSD.brain.fish.after.node.removal<-nodes.contrast.SSD.brain.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.brain.fish.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.brain.fish.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.brain.fish.after.node.removal$count<- paste0(median.exp.jump.SSD.brain.fish.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.brain.fish.after.node.removal$freq,")")


plotS14E<- boxplot.new(nodes.contrast.SSD.brain.jumps.fish.removing.jump.node,Pval.exp.jump.SSD.brain.fish.after.node.removal,"Exp",median.exp.jump.SSD.brain.fish.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("E. Removing jump node contrasts for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.brain.fish.after.node.removal$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test after removing nodes supporting jump for brain in trusted SSD teleosts

brain.speciation.15spe.after.node.removal<-speciation.contrast(nodes.contrast.SSD.brain.jumps.15spe.removing.jump.node,"pic")
brain.duplication.15spe.after.node.removal<-duplication.contrast(nodes.contrast.SSD.brain.jumps.15spe.removing.jump.node,"pic")
Pval.exp.jump.SSD.brain.15spe.after.node.removal<-two.tailed.wilcox(brain.speciation.15spe.after.node.removal,brain.duplication.15spe.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.brain.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.brain.15spe.after.node.removal<-NULL
median.exp.jump.SSD.brain.15spe.after.node.removal<-nodes.contrast.SSD.brain.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.brain.15spe.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.brain.15spe.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.brain.15spe.after.node.removal$count<- paste0(median.exp.jump.SSD.brain.15spe.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.brain.15spe.after.node.removal$freq,")")


plotS14F<- boxplot.new(nodes.contrast.SSD.brain.jumps.15spe.removing.jump.node,Pval.exp.jump.SSD.brain.15spe.after.node.removal,"Exp",median.exp.jump.SSD.brain.15spe.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. Removing jump node contrasts in 15 species"))))+
  theme(plot.title = element_text(color="black", size=11, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.brain.15spe.after.node.removal$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)

  
gridExtra::grid.arrange(plotS14A,plotS14B,plotS14C,plotS14D,plotS14E,plotS14F,ncol=2, nrow=3)

```


```{r Fig_S15, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S15**: Analyses of rapid jumps in heart expression for trusted SSD trees. We used a strict posterior probability cutoff of $\\ge$ 0.7 as an evidence of trait jump. (A)-(B) Proportions of jumps in heart expressions following speciation and duplication events for teleosts and for vertebrates using 2135 trait jump supporting trees of Fig. 7B. Since proportion of shifts in traits per branch of events is estimated for each tree, paired Wilcoxon test is used to compare the difference for (A)-(B). (C)-(D) The ortholog conjecture (OC) tests on 2135 trusted SSD trees to analyze the jump results in heart expressions for teleosts and for 15 vertebrates, respectively. (E)-(F) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for 2135 trusted SSD trees for 5 teleosts and for 15 vertebrates, respectively. *P* values are from Wilcoxon two-tailed tests for (C)-(F). ‘exp’: expression levels. ")}

## Proportions of jumps in heart for teleosts
pval.jump.SSD.heart.fish <- paired.wilcox(info.jump.trees.SSD.heart.final1.fish$spe.jump.prob,info.jump.trees.SSD.heart.final1.fish$dup.jump.prob) #p-value < 2.2e-16

## For median data
dtfr.SSD.heart.fish<-NULL
dtfr.SSD.heart.fish<-data.frame(prop.jump=info.jump.trees.SSD.heart.final1.fish$spe.jump.prob,events="speciation")
dtfr.SSD.heart.fish<-rbind(dtfr.SSD.heart.fish,data.frame(prop.jump=info.jump.trees.SSD.heart.final1.fish$dup.jump.prob,events="duplication"))
dtfr.SSD.heart.fish$Exp.abs <-abs(dtfr.SSD.heart.fish$prop.jump)
dtfr.SSD.heart.fish$Event<-factor(dtfr.SSD.heart.fish$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.heart.fish$Event)<-c("Speciation","SSD")

median.data.jump.SSD.heart.fish<-NULL
median.data.jump.SSD.heart.fish<- dtfr.SSD.heart.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.heart.fish$pic.round<-paste0("Median = ",round(median.data.jump.SSD.heart.fish$Exp.abs,4), sep="")
median.data.jump.SSD.heart.fish$count<- paste0(median.data.jump.SSD.heart.fish$Event,"\n","(n = ",median.data.jump.SSD.heart.fish$freq,")")

plotS15A<-vioplot.new2(dtfr.SSD.heart.fish,pval.jump.SSD.heart.fish,"Exp",median.data.jump.SSD.heart.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("A. Jump in heart exp for 5 teleosts"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.heart.fish, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.heart.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.heart.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.heart.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.heart.fish$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.1, color = "black")  
 

## Proportions of jumps in heart for 15 species
pval.jump.SSD.heart <- paired.wilcox(info.jump.heart.final1.SSD.all$spe.jump.prob,info.jump.heart.final1.SSD.all$dup.jump.prob) #p-value = 2.2e-16

## For median data
dtfr.SSD.heart.15spe<-NULL
dtfr.SSD.heart.15spe<-data.frame(prop.jump=info.jump.heart.final1.SSD.all$spe.jump.prob,events="speciation")
dtfr.SSD.heart.15spe<-rbind(dtfr.SSD.heart.15spe,data.frame(prop.jump=info.jump.heart.final1.SSD.all$dup.jump.prob,events="duplication"))
dtfr.SSD.heart.15spe$Exp.abs <-abs(dtfr.SSD.heart.15spe$prop.jump)
dtfr.SSD.heart.15spe$Event<-factor(dtfr.SSD.heart.15spe$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.heart.15spe$Event)<-c("Speciation","SSD")

library(tidyverse)
median.data.jump.SSD.heart<-NULL
median.data.jump.SSD.heart<- dtfr.SSD.heart.15spe%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.heart$pic.round<-paste0("Median = ",round(median.data.jump.SSD.heart$Exp.abs,4), sep="")
median.data.jump.SSD.heart$count<- paste0(median.data.jump.SSD.heart$Event,"\n","(n = ",median.data.jump.SSD.heart$freq,")")

plotS15B<-vioplot.new2(dtfr.SSD.heart.15spe,pval.jump.SSD.heart,"Exp",median.data.jump.SSD.heart) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("B. Jump in heart exp 15 species"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
 # annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.heart, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.heart,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.heart,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.heart,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.heart$count) +       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.07, color = "black")+
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.82, ysize = 0.1, color = "black")


# The ortholog conjecture test on heart expressions for trees supporting jump in trusted SSD teleosts

heart.speciation.fish<-speciation.contrast(nodes.contrast.SSD.heart.jumps.fish,"pic")
heart.duplication.fish<-duplication.contrast(nodes.contrast.SSD.heart.jumps.fish,"pic")
Pval.exp.jump.SSD.heart.fish<-two.tailed.wilcox(heart.speciation.fish,heart.duplication.fish)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.heart.jumps.fish$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.heart.fish<-NULL
median.exp.jump.SSD.heart.fish<-nodes.contrast.SSD.heart.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.heart.fish$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.heart.fish$Exp.abs,4), sep="")
median.exp.jump.SSD.heart.fish$count<- paste0(median.exp.jump.SSD.heart.fish$Event,"\n","(n = ",median.exp.jump.SSD.heart.fish$freq,")")


plotS15C<- boxplot.new(nodes.contrast.SSD.heart.jumps.fish,Pval.exp.jump.SSD.heart.fish,"Exp",median.exp.jump.SSD.heart.fish)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("C. Jump in heart exp for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.heart.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test on heart expressions for trees supporting jump in trusted SSD teleosts

heart.speciation.15spe<-speciation.contrast(nodes.contrast.SSD.heart.jumps.15spe,"pic")
heart.duplication.15spe<-duplication.contrast(nodes.contrast.SSD.heart.jumps.15spe,"pic")
Pval.exp.jump.SSD.heart.15spe<-two.tailed.wilcox(heart.speciation.15spe,heart.duplication.15spe)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.heart.jumps.15spe$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.heart.15spe<-NULL
median.exp.jump.SSD.heart.15spe<-nodes.contrast.SSD.heart.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.heart.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.heart.15spe$Exp.abs,4), sep="")
median.exp.jump.SSD.heart.15spe$count<- paste0(median.exp.jump.SSD.heart.15spe$Event,"\n","(n = ",median.exp.jump.SSD.heart.15spe$freq,")")


plotS15D<- boxplot.new(nodes.contrast.SSD.heart.jumps.15spe,Pval.exp.jump.SSD.heart.15spe,"Exp",median.exp.jump.SSD.heart.15spe)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Jump in heart exp for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.heart.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)


# The ortholog conjecture test after removing nodes supporting jump for heart in trusted SSD teleosts

heart.speciation.fish.after.node.removal<-speciation.contrast(nodes.contrast.SSD.heart.jumps.fish.removing.jump.node,"pic")
heart.duplication.fish.after.node.removal<-duplication.contrast(nodes.contrast.SSD.heart.jumps.fish.removing.jump.node,"pic")
Pval.exp.jump.SSD.heart.fish.after.node.removal<-two.tailed.wilcox(heart.speciation.fish.after.node.removal,heart.duplication.fish.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.heart.jumps.fish.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.heart.fish.after.node.removal<-NULL
median.exp.jump.SSD.heart.fish.after.node.removal<-nodes.contrast.SSD.heart.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.heart.fish.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.heart.fish.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.heart.fish.after.node.removal$count<- paste0(median.exp.jump.SSD.heart.fish.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.heart.fish.after.node.removal$freq,")")


plotS15E<- boxplot.new(nodes.contrast.SSD.heart.jumps.fish.removing.jump.node,Pval.exp.jump.SSD.heart.fish.after.node.removal,"Exp",median.exp.jump.SSD.heart.fish.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("E. Removing jump node contrasts for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.heart.fish.after.node.removal$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test after removing nodes supporting jump for heart in trusted SSD teleosts

heart.speciation.15spe.after.node.removal<-speciation.contrast(nodes.contrast.SSD.heart.jumps.15spe.removing.jump.node,"pic")
heart.duplication.15spe.after.node.removal<-duplication.contrast(nodes.contrast.SSD.heart.jumps.15spe.removing.jump.node,"pic")
Pval.exp.jump.SSD.heart.15spe.after.node.removal<-two.tailed.wilcox(heart.speciation.15spe.after.node.removal,heart.duplication.15spe.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.heart.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.heart.15spe.after.node.removal<-NULL
median.exp.jump.SSD.heart.15spe.after.node.removal<-nodes.contrast.SSD.heart.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.heart.15spe.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.heart.15spe.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.heart.15spe.after.node.removal$count<- paste0(median.exp.jump.SSD.heart.15spe.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.heart.15spe.after.node.removal$freq,")")


plotS15F<- boxplot.new(nodes.contrast.SSD.heart.jumps.15spe.removing.jump.node,Pval.exp.jump.SSD.heart.15spe.after.node.removal,"Exp",median.exp.jump.SSD.heart.15spe.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. Removing jump node contrasts for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.heart.15spe.after.node.removal$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)

  
gridExtra::grid.arrange(plotS15A,plotS15B,plotS15C,plotS15D,plotS15E,plotS15F,ncol=2, nrow=3)

```


```{r Fig_S16, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S16**: Analyses of rapid jumps in kidney expression for trusted SSD trees. We used a strict posterior probability cutoff of $\\ge$ 0.7 as an evidence of trait jump. (A)-(B) Proportions of jumps in kidney expressions following speciation and duplication events for teleosts and for vertebrates using 1788 trait jump supporting trees of Fig. 7C. Since proportion of shifts in traits per branch of events is estimated for each tree, paired Wilcoxon test is used to compare the difference for (A)-(B). (C)-(D) The ortholog conjecture (OC) tests on 1788 trusted SSD trees to analyze the jump results in kidney expressions for teleosts and for 15 vertebrates, respectively. (E)-(F) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for 1788 trusted SSD trees for 5 teleosts and for 15 vertebrates, respectively. *P* values are from Wilcoxon two-tailed tests for (C)-(F). ‘exp’: expression levels. ")}

## Proportions of jumps in kidney for teleosts
pval.jump.SSD.kidney.fish <- paired.wilcox(info.jump.trees.SSD.kidney.final1.fish$spe.jump.prob,info.jump.trees.SSD.kidney.final1.fish$dup.jump.prob) #p-value < 2.2e-16

## For median data
dtfr.SSD.kidney.fish<-NULL
dtfr.SSD.kidney.fish<-data.frame(prop.jump=info.jump.trees.SSD.kidney.final1.fish$spe.jump.prob,events="speciation")
dtfr.SSD.kidney.fish<-rbind(dtfr.SSD.kidney.fish,data.frame(prop.jump=info.jump.trees.SSD.kidney.final1.fish$dup.jump.prob,events="duplication"))
dtfr.SSD.kidney.fish$Exp.abs <-abs(dtfr.SSD.kidney.fish$prop.jump)
dtfr.SSD.kidney.fish$Event<-factor(dtfr.SSD.kidney.fish$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.kidney.fish$Event)<-c("Speciation","SSD")

median.data.jump.SSD.kidney.fish<-NULL
median.data.jump.SSD.kidney.fish<- dtfr.SSD.kidney.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.kidney.fish$pic.round<-paste0("Median = ",round(median.data.jump.SSD.kidney.fish$Exp.abs,4), sep="")
median.data.jump.SSD.kidney.fish$count<- paste0(median.data.jump.SSD.kidney.fish$Event,"\n","(n = ",median.data.jump.SSD.kidney.fish$freq,")")

plotS16A<-vioplot.new2(dtfr.SSD.kidney.fish,pval.jump.SSD.kidney.fish,"Exp",median.data.jump.SSD.kidney.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("A. Jump in kidney exp for 5 teleosts"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.kidney.fish, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.kidney.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.kidney.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.kidney.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.kidney.fish$count) +       
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.1, color = "black")  


## Proportions of jumps in kidney for 15 species
pval.jump.SSD.kidney <- paired.wilcox(info.jump.kidney.final1.SSD.all$spe.jump.prob,info.jump.kidney.final1.SSD.all$dup.jump.prob) #p-value = 2.2e-16

## For median data
dtfr.SSD.kidney.15spe<-NULL
dtfr.SSD.kidney.15spe<-data.frame(prop.jump=info.jump.kidney.final1.SSD.all$spe.jump.prob,events="speciation")
dtfr.SSD.kidney.15spe<-rbind(dtfr.SSD.kidney.15spe,data.frame(prop.jump=info.jump.kidney.final1.SSD.all$dup.jump.prob,events="duplication"))
dtfr.SSD.kidney.15spe$Exp.abs <-abs(dtfr.SSD.kidney.15spe$prop.jump)
dtfr.SSD.kidney.15spe$Event<-factor(dtfr.SSD.kidney.15spe$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.kidney.15spe$Event)<-c("Speciation","SSD")

library(tidyverse)
median.data.jump.SSD.kidney<-NULL
median.data.jump.SSD.kidney<- dtfr.SSD.kidney.15spe%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.kidney$pic.round<-paste0("Median = ",round(median.data.jump.SSD.kidney$Exp.abs,4), sep="")
median.data.jump.SSD.kidney$count<- paste0(median.data.jump.SSD.kidney$Event,"\n","(n = ",median.data.jump.SSD.kidney$freq,")")

plotS16B<-vioplot.new2(dtfr.SSD.kidney.15spe,pval.jump.SSD.kidney,"Exp",median.data.jump.SSD.kidney) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("B. Jump in kidney exp 15 species"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.kidney, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.kidney,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.kidney,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.kidney,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.kidney$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.07, color = "black")+
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.82, ysize = 0.1, color = "black") 


# The ortholog conjecture test on kidney expressions for trees supporting jump in trusted SSD teleosts

kidney.speciation.fish<-speciation.contrast(nodes.contrast.SSD.kidney.jumps.fish,"pic")
kidney.duplication.fish<-duplication.contrast(nodes.contrast.SSD.kidney.jumps.fish,"pic")
Pval.exp.jump.SSD.kidney.fish<-two.tailed.wilcox(kidney.speciation.fish,kidney.duplication.fish)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.kidney.jumps.fish$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.kidney.fish<-NULL
median.exp.jump.SSD.kidney.fish<-nodes.contrast.SSD.kidney.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.kidney.fish$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.kidney.fish$Exp.abs,4), sep="")
median.exp.jump.SSD.kidney.fish$count<- paste0(median.exp.jump.SSD.kidney.fish$Event,"\n","(n = ",median.exp.jump.SSD.kidney.fish$freq,")")


plotS16C<- boxplot.new(nodes.contrast.SSD.kidney.jumps.fish,Pval.exp.jump.SSD.kidney.fish,"Exp",median.exp.jump.SSD.kidney.fish)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("C. Jump in kidney exp for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.kidney.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test on kidney expressions for trees supporting jump in trusted SSD teleosts

kidney.speciation.15spe<-speciation.contrast(nodes.contrast.SSD.kidney.jumps.15spe,"pic")
kidney.duplication.15spe<-duplication.contrast(nodes.contrast.SSD.kidney.jumps.15spe,"pic")
Pval.exp.jump.SSD.kidney.15spe<-two.tailed.wilcox(kidney.speciation.15spe,kidney.duplication.15spe)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.kidney.jumps.15spe$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.kidney.15spe<-NULL
median.exp.jump.SSD.kidney.15spe<-nodes.contrast.SSD.kidney.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.kidney.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.kidney.15spe$Exp.abs,4), sep="")
median.exp.jump.SSD.kidney.15spe$count<- paste0(median.exp.jump.SSD.kidney.15spe$Event,"\n","(n = ",median.exp.jump.SSD.kidney.15spe$freq,")")


plotS16D<- boxplot.new(nodes.contrast.SSD.kidney.jumps.15spe,Pval.exp.jump.SSD.kidney.15spe,"Exp",median.exp.jump.SSD.kidney.15spe)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Jump in kidney exp for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.kidney.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)


# The ortholog conjecture test after removing nodes supporting jump for kidney in trusted SSD teleosts

kidney.speciation.fish.after.node.removal<-speciation.contrast(nodes.contrast.SSD.kidney.jumps.fish.removing.jump.node,"pic")
kidney.duplication.fish.after.node.removal<-duplication.contrast(nodes.contrast.SSD.kidney.jumps.fish.removing.jump.node,"pic")
Pval.exp.jump.SSD.kidney.fish.after.node.removal<-two.tailed.wilcox(kidney.speciation.fish.after.node.removal,kidney.duplication.fish.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.kidney.jumps.fish.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.kidney.fish.after.node.removal<-NULL
median.exp.jump.SSD.kidney.fish.after.node.removal<-nodes.contrast.SSD.kidney.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.kidney.fish.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.kidney.fish.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.kidney.fish.after.node.removal$count<- paste0(median.exp.jump.SSD.kidney.fish.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.kidney.fish.after.node.removal$freq,")")


plotS16E<- boxplot.new(nodes.contrast.SSD.kidney.jumps.fish.removing.jump.node,Pval.exp.jump.SSD.kidney.fish.after.node.removal,"Exp",median.exp.jump.SSD.kidney.fish.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("E. Removing jump node contrasts for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.kidney.fish.after.node.removal$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test after removing nodes supporting jump for kidney in trusted SSD teleosts

kidney.speciation.15spe.after.node.removal<-speciation.contrast(nodes.contrast.SSD.kidney.jumps.15spe.removing.jump.node,"pic")
kidney.duplication.15spe.after.node.removal<-duplication.contrast(nodes.contrast.SSD.kidney.jumps.15spe.removing.jump.node,"pic")
Pval.exp.jump.SSD.kidney.15spe.after.node.removal<-two.tailed.wilcox(kidney.speciation.15spe.after.node.removal,kidney.duplication.15spe.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.kidney.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.kidney.15spe.after.node.removal<-NULL
median.exp.jump.SSD.kidney.15spe.after.node.removal<-nodes.contrast.SSD.kidney.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.kidney.15spe.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.kidney.15spe.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.kidney.15spe.after.node.removal$count<- paste0(median.exp.jump.SSD.kidney.15spe.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.kidney.15spe.after.node.removal$freq,")")


plotS16F<- boxplot.new(nodes.contrast.SSD.kidney.jumps.15spe.removing.jump.node,Pval.exp.jump.SSD.kidney.15spe.after.node.removal,"Exp",median.exp.jump.SSD.kidney.15spe.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. Removing jump node contrasts for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.kidney.15spe.after.node.removal$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)

  
gridExtra::grid.arrange(plotS16A,plotS16B,plotS16C,plotS16D,plotS16E,plotS16F,ncol=2, nrow=3)

```


```{r Fig_S17, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S17**: Analyses of rapid jumps in muscle expression for trusted SSD trees. We used a strict posterior probability cutoff of $\\ge$ 0.7 as an evidence of trait jump. (A)-(B) Proportions of jumps in muscle expressions following speciation and duplication events for teleosts and for vertebrates using 2319 trait jump supporting trees of Fig. 7D. Since proportion of shifts in traits per branch of events is estimated for each tree, paired Wilcoxon test is used to compare the difference for (A)-(B). (C)-(D) The ortholog conjecture (OC) tests on 2319 trusted SSD trees to analyze the jump results in muscle expressions for teleosts and for 15 vertebrates, respectively. (E)-(F) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for 2319 trusted SSD trees for 5 teleosts and for 15 vertebrates, respectively. *P* values are from Wilcoxon two-tailed tests for (C)-(F). ‘exp’: expression levels. ")}

## Proportions of jumps in muscle for teleosts
pval.jump.SSD.muscle.fish <- paired.wilcox(info.jump.trees.SSD.muscle.final1.fish$spe.jump.prob,info.jump.trees.SSD.muscle.final1.fish$dup.jump.prob) #p-value < 2.2e-16

## For median data
dtfr.SSD.muscle.fish<-NULL
dtfr.SSD.muscle.fish<-data.frame(prop.jump=info.jump.trees.SSD.muscle.final1.fish$spe.jump.prob,events="speciation")
dtfr.SSD.muscle.fish<-rbind(dtfr.SSD.muscle.fish,data.frame(prop.jump=info.jump.trees.SSD.muscle.final1.fish$dup.jump.prob,events="duplication"))
dtfr.SSD.muscle.fish$Exp.abs <-abs(dtfr.SSD.muscle.fish$prop.jump)
dtfr.SSD.muscle.fish$Event<-factor(dtfr.SSD.muscle.fish$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.muscle.fish$Event)<-c("Speciation","SSD")

median.data.jump.SSD.muscle.fish<-NULL
median.data.jump.SSD.muscle.fish<- dtfr.SSD.muscle.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.muscle.fish$pic.round<-paste0("Median = ",round(median.data.jump.SSD.muscle.fish$Exp.abs,4), sep="")
median.data.jump.SSD.muscle.fish$count<- paste0(median.data.jump.SSD.muscle.fish$Event,"\n","(n = ",median.data.jump.SSD.muscle.fish$freq,")")

plotS17A<-vioplot.new2(dtfr.SSD.muscle.fish,pval.jump.SSD.muscle.fish,"Exp",median.data.jump.SSD.muscle.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("A. Jump in muscle exp for 5 teleosts"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.muscle.fish, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.muscle.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.muscle.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.muscle.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.muscle.fish$count) +       
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.1, color = "black")  


## Proportions of jumps in muscle for 15 species
pval.jump.SSD.muscle <- paired.wilcox(info.jump.muscle.final1.SSD.all$spe.jump.prob,info.jump.muscle.final1.SSD.all$dup.jump.prob) #p-value = 2.2e-16

## For median data
dtfr.SSD.muscle.15spe<-NULL
dtfr.SSD.muscle.15spe<-data.frame(prop.jump=info.jump.muscle.final1.SSD.all$spe.jump.prob,events="speciation")
dtfr.SSD.muscle.15spe<-rbind(dtfr.SSD.muscle.15spe,data.frame(prop.jump=info.jump.muscle.final1.SSD.all$dup.jump.prob,events="duplication"))
dtfr.SSD.muscle.15spe$Exp.abs <-abs(dtfr.SSD.muscle.15spe$prop.jump)
dtfr.SSD.muscle.15spe$Event<-factor(dtfr.SSD.muscle.15spe$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.muscle.15spe$Event)<-c("Speciation","SSD")

library(tidyverse)
median.data.jump.SSD.muscle<-NULL
median.data.jump.SSD.muscle<- dtfr.SSD.muscle.15spe%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.muscle$pic.round<-paste0("Median = ",round(median.data.jump.SSD.muscle$Exp.abs,4), sep="")
median.data.jump.SSD.muscle$count<- paste0(median.data.jump.SSD.muscle$Event,"\n","(n = ",median.data.jump.SSD.muscle$freq,")")

plotS17B<-vioplot.new2(dtfr.SSD.muscle.15spe,pval.jump.SSD.muscle,"Exp",median.data.jump.SSD.muscle) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("B. Jump in muscle exp 15 species"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.muscle, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.muscle,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.muscle,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.muscle,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.muscle$count) +       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.07, color = "black")+
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.82, ysize = 0.1, color = "black")


# The ortholog conjecture test on muscle expressions for trees supporting jump in trusted SSD teleosts

muscle.speciation.fish<-speciation.contrast(nodes.contrast.SSD.muscle.jumps.fish,"pic")
muscle.duplication.fish<-duplication.contrast(nodes.contrast.SSD.muscle.jumps.fish,"pic")
Pval.exp.jump.SSD.muscle.fish<-two.tailed.wilcox(muscle.speciation.fish,muscle.duplication.fish)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.muscle.jumps.fish$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.muscle.fish<-NULL
median.exp.jump.SSD.muscle.fish<-nodes.contrast.SSD.muscle.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.muscle.fish$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.muscle.fish$Exp.abs,4), sep="")
median.exp.jump.SSD.muscle.fish$count<- paste0(median.exp.jump.SSD.muscle.fish$Event,"\n","(n = ",median.exp.jump.SSD.muscle.fish$freq,")")


plotS17C<- boxplot.new(nodes.contrast.SSD.muscle.jumps.fish,Pval.exp.jump.SSD.muscle.fish,"Exp",median.exp.jump.SSD.muscle.fish)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("C. Jump in muscle exp for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.muscle.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test on muscle expressions for trees supporting jump in trusted SSD teleosts

muscle.speciation.15spe<-speciation.contrast(nodes.contrast.SSD.muscle.jumps.15spe,"pic")
muscle.duplication.15spe<-duplication.contrast(nodes.contrast.SSD.muscle.jumps.15spe,"pic")
Pval.exp.jump.SSD.muscle.15spe<-two.tailed.wilcox(muscle.speciation.15spe,muscle.duplication.15spe)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.muscle.jumps.15spe$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.muscle.15spe<-NULL
median.exp.jump.SSD.muscle.15spe<-nodes.contrast.SSD.muscle.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.muscle.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.muscle.15spe$Exp.abs,4), sep="")
median.exp.jump.SSD.muscle.15spe$count<- paste0(median.exp.jump.SSD.muscle.15spe$Event,"\n","(n = ",median.exp.jump.SSD.muscle.15spe$freq,")")


plotS17D<- boxplot.new(nodes.contrast.SSD.muscle.jumps.15spe,Pval.exp.jump.SSD.muscle.15spe,"Exp",median.exp.jump.SSD.muscle.15spe)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Jump in muscle exp for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.muscle.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)


# The ortholog conjecture test after removing nodes supporting jump for muscle in trusted SSD teleosts

muscle.speciation.fish.after.node.removal<-speciation.contrast(nodes.contrast.SSD.muscle.jumps.fish.removing.jump.node,"pic")
muscle.duplication.fish.after.node.removal<-duplication.contrast(nodes.contrast.SSD.muscle.jumps.fish.removing.jump.node,"pic")
Pval.exp.jump.SSD.muscle.fish.after.node.removal<-two.tailed.wilcox(muscle.speciation.fish.after.node.removal,muscle.duplication.fish.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.muscle.jumps.fish.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.muscle.fish.after.node.removal<-NULL
median.exp.jump.SSD.muscle.fish.after.node.removal<-nodes.contrast.SSD.muscle.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.muscle.fish.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.muscle.fish.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.muscle.fish.after.node.removal$count<- paste0(median.exp.jump.SSD.muscle.fish.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.muscle.fish.after.node.removal$freq,")")


plotS17E<- boxplot.new(nodes.contrast.SSD.muscle.jumps.fish.removing.jump.node,Pval.exp.jump.SSD.muscle.fish.after.node.removal,"Exp",median.exp.jump.SSD.muscle.fish.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("E. Removing jump node contrasts for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.muscle.fish.after.node.removal$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test after removing nodes supporting jump for muscle in trusted SSD teleosts

muscle.speciation.15spe.after.node.removal<-speciation.contrast(nodes.contrast.SSD.muscle.jumps.15spe.removing.jump.node,"pic")
muscle.duplication.15spe.after.node.removal<-duplication.contrast(nodes.contrast.SSD.muscle.jumps.15spe.removing.jump.node,"pic")
Pval.exp.jump.SSD.muscle.15spe.after.node.removal<-two.tailed.wilcox(muscle.speciation.15spe.after.node.removal,muscle.duplication.15spe.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.muscle.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.muscle.15spe.after.node.removal<-NULL
median.exp.jump.SSD.muscle.15spe.after.node.removal<-nodes.contrast.SSD.muscle.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.muscle.15spe.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.muscle.15spe.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.muscle.15spe.after.node.removal$count<- paste0(median.exp.jump.SSD.muscle.15spe.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.muscle.15spe.after.node.removal$freq,")")


plotS17F<- boxplot.new(nodes.contrast.SSD.muscle.jumps.15spe.removing.jump.node,Pval.exp.jump.SSD.muscle.15spe.after.node.removal,"Exp",median.exp.jump.SSD.muscle.15spe.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. Removing jump node contrasts for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.muscle.15spe.after.node.removal$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)

  
gridExtra::grid.arrange(plotS17A,plotS17B,plotS17C,plotS17D,plotS17E,plotS17F,ncol=2, nrow=3)

```


```{r Fig_S18, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S18**: Analyses of rapid jumps in liver expression for trusted SSD trees. We used a strict posterior probability cutoff of $\\ge$ 0.7 as an evidence of trait jump. (A)-(B) Proportions of jumps in liver expressions following speciation and duplication events for teleosts and for vertebrates using 2641 trait jump supporting trees of Fig. 7E. Since proportion of shifts in traits per branch of events is estimated for each tree, paired Wilcoxon test is used to compare the difference for (A)-(B). (C)-(D) The ortholog conjecture (OC) tests on 2641 trusted SSD trees to analyze the jump results in liver expressions for teleosts and for 15 vertebrates, respectively. (E)-(F) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for 2641 trusted SSD trees for 5 teleosts and for 15 vertebrates, respectively. *P* values are from Wilcoxon two-tailed tests for (C)-(F). ‘exp’: expression levels. ")}

## Proportions of jumps in liver for teleosts
pval.jump.SSD.liver.fish <- paired.wilcox(info.jump.trees.SSD.liver.final1.fish$spe.jump.prob,info.jump.trees.SSD.liver.final1.fish$dup.jump.prob) #p-value < 2.2e-16

## For median data
dtfr.SSD.liver.fish<-NULL
dtfr.SSD.liver.fish<-data.frame(prop.jump=info.jump.trees.SSD.liver.final1.fish$spe.jump.prob,events="speciation")
dtfr.SSD.liver.fish<-rbind(dtfr.SSD.liver.fish,data.frame(prop.jump=info.jump.trees.SSD.liver.final1.fish$dup.jump.prob,events="duplication"))
dtfr.SSD.liver.fish$Exp.abs <-abs(dtfr.SSD.liver.fish$prop.jump)
dtfr.SSD.liver.fish$Event<-factor(dtfr.SSD.liver.fish$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.liver.fish$Event)<-c("Speciation","SSD")

median.data.jump.SSD.liver.fish<-NULL
median.data.jump.SSD.liver.fish<- dtfr.SSD.liver.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.liver.fish$pic.round<-paste0("Median = ",round(median.data.jump.SSD.liver.fish$Exp.abs,4), sep="")
median.data.jump.SSD.liver.fish$count<- paste0(median.data.jump.SSD.liver.fish$Event,"\n","(n = ",median.data.jump.SSD.liver.fish$freq,")")

plotS18A<-vioplot.new2(dtfr.SSD.liver.fish,pval.jump.SSD.liver.fish,"Exp",median.data.jump.SSD.liver.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("A. Jump in liver exp for 5 teleosts"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.liver.fish, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.liver.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.liver.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.liver.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.liver.fish$count)+       
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.1, color = "black")  
 

## Proportions of jumps in liver for 15 species
pval.jump.SSD.liver <- paired.wilcox(info.jump.liver.final1.SSD.all$spe.jump.prob,info.jump.liver.final1.SSD.all$dup.jump.prob) #p-value = 2.2e-16

## For median data
dtfr.SSD.liver.15spe<-NULL
dtfr.SSD.liver.15spe<-data.frame(prop.jump=info.jump.liver.final1.SSD.all$spe.jump.prob,events="speciation")
dtfr.SSD.liver.15spe<-rbind(dtfr.SSD.liver.15spe,data.frame(prop.jump=info.jump.liver.final1.SSD.all$dup.jump.prob,events="duplication"))
dtfr.SSD.liver.15spe$Exp.abs <-abs(dtfr.SSD.liver.15spe$prop.jump)
dtfr.SSD.liver.15spe$Event<-factor(dtfr.SSD.liver.15spe$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.liver.15spe$Event)<-c("Speciation","SSD")

library(tidyverse)
median.data.jump.SSD.liver<-NULL
median.data.jump.SSD.liver<- dtfr.SSD.liver.15spe%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.liver$pic.round<-paste0("Median = ",round(median.data.jump.SSD.liver$Exp.abs,4), sep="")
median.data.jump.SSD.liver$count<- paste0(median.data.jump.SSD.liver$Event,"\n","(n = ",median.data.jump.SSD.liver$freq,")")

plotS18B<-vioplot.new2(dtfr.SSD.liver.15spe,pval.jump.SSD.liver,"Exp",median.data.jump.SSD.liver) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("B. Jump in liver exp 15 species"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
 # annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.liver, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.liver,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.liver,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.liver,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.liver$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.07, color = "black")+
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.82, ysize = 0.1, color = "black") 


# The ortholog conjecture test on liver expressions for trees supporting jump in trusted SSD teleosts

liver.speciation.fish<-speciation.contrast(nodes.contrast.SSD.liver.jumps.fish,"pic")
liver.duplication.fish<-duplication.contrast(nodes.contrast.SSD.liver.jumps.fish,"pic")
Pval.exp.jump.SSD.liver.fish<-two.tailed.wilcox(liver.speciation.fish,liver.duplication.fish)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.liver.jumps.fish$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.liver.fish<-NULL
median.exp.jump.SSD.liver.fish<-nodes.contrast.SSD.liver.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.liver.fish$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.liver.fish$Exp.abs,4), sep="")
median.exp.jump.SSD.liver.fish$count<- paste0(median.exp.jump.SSD.liver.fish$Event,"\n","(n = ",median.exp.jump.SSD.liver.fish$freq,")")


plotS18C<- boxplot.new(nodes.contrast.SSD.liver.jumps.fish,Pval.exp.jump.SSD.liver.fish,"Exp",median.exp.jump.SSD.liver.fish)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("C. Jump in liver exp for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.liver.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test on liver expressions for trees supporting jump in trusted SSD teleosts

liver.speciation.15spe<-speciation.contrast(nodes.contrast.SSD.liver.jumps.15spe,"pic")
liver.duplication.15spe<-duplication.contrast(nodes.contrast.SSD.liver.jumps.15spe,"pic")
Pval.exp.jump.SSD.liver.15spe<-two.tailed.wilcox(liver.speciation.15spe,liver.duplication.15spe)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.liver.jumps.15spe$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.liver.15spe<-NULL
median.exp.jump.SSD.liver.15spe<-nodes.contrast.SSD.liver.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.liver.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.liver.15spe$Exp.abs,4), sep="")
median.exp.jump.SSD.liver.15spe$count<- paste0(median.exp.jump.SSD.liver.15spe$Event,"\n","(n = ",median.exp.jump.SSD.liver.15spe$freq,")")


plotS18D<- boxplot.new(nodes.contrast.SSD.liver.jumps.15spe,Pval.exp.jump.SSD.liver.15spe,"Exp",median.exp.jump.SSD.liver.15spe)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Jump in liver exp for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.liver.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)


# The ortholog conjecture test after removing nodes supporting jump for liver in trusted SSD teleosts

liver.speciation.fish.after.node.removal<-speciation.contrast(nodes.contrast.SSD.liver.jumps.fish.removing.jump.node,"pic")
liver.duplication.fish.after.node.removal<-duplication.contrast(nodes.contrast.SSD.liver.jumps.fish.removing.jump.node,"pic")
Pval.exp.jump.SSD.liver.fish.after.node.removal<-two.tailed.wilcox(liver.speciation.fish.after.node.removal,liver.duplication.fish.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.liver.jumps.fish.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.liver.fish.after.node.removal<-NULL
median.exp.jump.SSD.liver.fish.after.node.removal<-nodes.contrast.SSD.liver.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.liver.fish.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.liver.fish.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.liver.fish.after.node.removal$count<- paste0(median.exp.jump.SSD.liver.fish.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.liver.fish.after.node.removal$freq,")")


plotS18E<- boxplot.new(nodes.contrast.SSD.liver.jumps.fish.removing.jump.node,Pval.exp.jump.SSD.liver.fish.after.node.removal,"Exp",median.exp.jump.SSD.liver.fish.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("E. Removing jump node contrasts for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.liver.fish.after.node.removal$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test after removing nodes supporting jump for liver in trusted SSD teleosts

liver.speciation.15spe.after.node.removal<-speciation.contrast(nodes.contrast.SSD.liver.jumps.15spe.removing.jump.node,"pic")
liver.duplication.15spe.after.node.removal<-duplication.contrast(nodes.contrast.SSD.liver.jumps.15spe.removing.jump.node,"pic")
Pval.exp.jump.SSD.liver.15spe.after.node.removal<-two.tailed.wilcox(liver.speciation.15spe.after.node.removal,liver.duplication.15spe.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.liver.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.liver.15spe.after.node.removal<-NULL
median.exp.jump.SSD.liver.15spe.after.node.removal<-nodes.contrast.SSD.liver.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.liver.15spe.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.liver.15spe.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.liver.15spe.after.node.removal$count<- paste0(median.exp.jump.SSD.liver.15spe.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.liver.15spe.after.node.removal$freq,")")


plotS18F<- boxplot.new(nodes.contrast.SSD.liver.jumps.15spe.removing.jump.node,Pval.exp.jump.SSD.liver.15spe.after.node.removal,"Exp",median.exp.jump.SSD.liver.15spe.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. Removing jump node contrasts for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.liver.15spe.after.node.removal$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)

  
gridExtra::grid.arrange(plotS18A,plotS18B,plotS18C,plotS18D,plotS18E,plotS18F,ncol=2, nrow=3)

```



```{r Fig_S19, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S19**: Analyses of rapid jumps in testis expression for trusted SSD trees. We used a strict posterior probability cutoff of $\\ge$ 0.7 as an evidence of trait jump. (A)-(B) Proportions of jumps in testis expressions following speciation and duplication events for teleosts and for vertebrates using 1441 trait jump supporting trees of Fig. 7F. Since proportion of shifts in traits per branch of events is estimated for each tree, paired Wilcoxon test is used to compare the difference for (A)-(B). (C)-(D) The ortholog conjecture (OC) tests on 1441 trusted SSD trees to analyse the jump results in testis expressions for teleosts and for 15 vertebrates, respectively. (E)-(F) The ortholog conjecture tests after removing contrasts of nodes whose daughter branch(es) experienced jumps in the corresponding trait for 1441 trusted SSD trees for 5 teleosts and for 15 vertebrates, respectively. *P* values are from Wilcoxon two-tailed tests for (C)-(F). ‘exp’: expression levels. ")}

## Proportions of jumps in testis for teleosts
pval.jump.SSD.testis.fish <- paired.wilcox(info.jump.trees.SSD.testis.final1.fish$spe.jump.prob,info.jump.trees.SSD.testis.final1.fish$dup.jump.prob) #p-value < 2.2e-16

## For median data
dtfr.SSD.testis.fish<-NULL
dtfr.SSD.testis.fish<-data.frame(prop.jump=info.jump.trees.SSD.testis.final1.fish$spe.jump.prob,events="speciation")
dtfr.SSD.testis.fish<-rbind(dtfr.SSD.testis.fish,data.frame(prop.jump=info.jump.trees.SSD.testis.final1.fish$dup.jump.prob,events="duplication"))
dtfr.SSD.testis.fish$Exp.abs <-abs(dtfr.SSD.testis.fish$prop.jump)
dtfr.SSD.testis.fish$Event<-factor(dtfr.SSD.testis.fish$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.testis.fish$Event)<-c("Speciation","SSD")

median.data.jump.SSD.testis.fish<-NULL
median.data.jump.SSD.testis.fish<- dtfr.SSD.testis.fish%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.testis.fish$pic.round<-paste0("Median = ",round(median.data.jump.SSD.testis.fish$Exp.abs,4), sep="")
median.data.jump.SSD.testis.fish$count<- paste0(median.data.jump.SSD.testis.fish$Event,"\n","(n = ",median.data.jump.SSD.testis.fish$freq,")")

plotS19A<-vioplot.new2(dtfr.SSD.testis.fish,pval.jump.SSD.testis.fish,"Exp",median.data.jump.SSD.testis.fish) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("A. Jump in testis exp for 5 teleosts"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.testis.fish, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.testis.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.testis.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.testis.fish,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.testis.fish$count) +       
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.1, color = "black")  

## Proportions of jumps in testis for 15 species
pval.jump.SSD.testis <- paired.wilcox(info.jump.testis.final1.SSD.all$spe.jump.prob,info.jump.testis.final1.SSD.all$dup.jump.prob) #p-value = 2.2e-16

## For median data
dtfr.SSD.testis.15spe<-NULL
dtfr.SSD.testis.15spe<-data.frame(prop.jump=info.jump.testis.final1.SSD.all$spe.jump.prob,events="speciation")
dtfr.SSD.testis.15spe<-rbind(dtfr.SSD.testis.15spe,data.frame(prop.jump=info.jump.testis.final1.SSD.all$dup.jump.prob,events="duplication"))
dtfr.SSD.testis.15spe$Exp.abs <-abs(dtfr.SSD.testis.15spe$prop.jump)
dtfr.SSD.testis.15spe$Event<-factor(dtfr.SSD.testis.15spe$events, levels=c("speciation","duplication"))
levels(dtfr.SSD.testis.15spe$Event)<-c("Speciation","SSD")

library(tidyverse)
median.data.jump.SSD.testis<-NULL
median.data.jump.SSD.testis<- dtfr.SSD.testis.15spe%>% 
  group_by(Event) %>% 
  summarise(Mean=mean(prop.jump),
            Exp.abs=median(prop.jump),
            freq = n())
median.data.jump.SSD.testis$pic.round<-paste0("Median = ",round(median.data.jump.SSD.testis$Exp.abs,4), sep="")
median.data.jump.SSD.testis$count<- paste0(median.data.jump.SSD.testis$Event,"\n","(n = ",median.data.jump.SSD.testis$freq,")")

plotS19B<-vioplot.new2(dtfr.SSD.testis.15spe,pval.jump.SSD.testis,"Exp",median.data.jump.SSD.testis) +  
  coord_cartesian(ylim=c(0, 1)) +
  ggtitle(expression(bold(paste("B. Jump in testis exp 15 species"))))+
  ylab(expression(bold(paste("Proportions of jump events")))) +
  #ylab(" ")+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.7, ymin = 0.78, ymax = 0.78, alpha = 0.8,colour = "blue")+
  #annotate("text", x = 1.45, y = 0.85, label= pval.jump.SSD.testis, parse=T)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.testis,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.testis,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 0.85, label= as.character(paste0("italic(",pval.jump.SSD.testis,")")), parse=T, size=4.04)+
  scale_x_discrete(labels=median.data.jump.SSD.testis$count)+       
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.33, y = 0.97, ysize = 0.07, color = "black")+
  add_phylopic(fish_pic, alpha = 1,x = 2.33, y = 0.82, ysize = 0.1, color = "black")  


# The ortholog conjecture test on testis expressions for trees supporting jump in trusted SSD teleosts

testis.speciation.fish<-speciation.contrast(nodes.contrast.SSD.testis.jumps.fish,"pic")
testis.duplication.fish<-duplication.contrast(nodes.contrast.SSD.testis.jumps.fish,"pic")
Pval.exp.jump.SSD.testis.fish<-two.tailed.wilcox(testis.speciation.fish,testis.duplication.fish)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.testis.jumps.fish$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.testis.fish<-NULL
median.exp.jump.SSD.testis.fish<-nodes.contrast.SSD.testis.jumps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.testis.fish$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.testis.fish$Exp.abs,4), sep="")
median.exp.jump.SSD.testis.fish$count<- paste0(median.exp.jump.SSD.testis.fish$Event,"\n","(n = ",median.exp.jump.SSD.testis.fish$freq,")")


plotS19C<- boxplot.new(nodes.contrast.SSD.testis.jumps.fish,Pval.exp.jump.SSD.testis.fish,"Exp",median.exp.jump.SSD.testis.fish)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("C. Jump in testis exp for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.testis.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test on testis expressions for trees supporting jump in trusted SSD teleosts

testis.speciation.15spe<-speciation.contrast(nodes.contrast.SSD.testis.jumps.15spe,"pic")
testis.duplication.15spe<-duplication.contrast(nodes.contrast.SSD.testis.jumps.15spe,"pic")
Pval.exp.jump.SSD.testis.15spe<-two.tailed.wilcox(testis.speciation.15spe,testis.duplication.15spe)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.testis.jumps.15spe$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.testis.15spe<-NULL
median.exp.jump.SSD.testis.15spe<-nodes.contrast.SSD.testis.jumps.15spe %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.testis.15spe$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.testis.15spe$Exp.abs,4), sep="")
median.exp.jump.SSD.testis.15spe$count<- paste0(median.exp.jump.SSD.testis.15spe$Event,"\n","(n = ",median.exp.jump.SSD.testis.15spe$freq,")")


plotS19D<- boxplot.new(nodes.contrast.SSD.testis.jumps.15spe,Pval.exp.jump.SSD.testis.15spe,"Exp",median.exp.jump.SSD.testis.15spe)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("D. Jump in testis exp for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.testis.15spe$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)


# The ortholog conjecture test after removing nodes supporting jump for testis in trusted SSD teleosts

testis.speciation.fish.after.node.removal<-speciation.contrast(nodes.contrast.SSD.testis.jumps.fish.removing.jump.node,"pic")
testis.duplication.fish.after.node.removal<-duplication.contrast(nodes.contrast.SSD.testis.jumps.fish.removing.jump.node,"pic")
Pval.exp.jump.SSD.testis.fish.after.node.removal<-two.tailed.wilcox(testis.speciation.fish.after.node.removal,testis.duplication.fish.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.testis.jumps.fish.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.testis.fish.after.node.removal<-NULL
median.exp.jump.SSD.testis.fish.after.node.removal<-nodes.contrast.SSD.testis.jumps.fish.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.testis.fish.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.testis.fish.after.node.removal$Exp.abs,5), sep="")
median.exp.jump.SSD.testis.fish.after.node.removal$count<- paste0(median.exp.jump.SSD.testis.fish.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.testis.fish.after.node.removal$freq,")")


plotS19E<- boxplot.new(nodes.contrast.SSD.testis.jumps.fish.removing.jump.node,Pval.exp.jump.SSD.testis.fish.after.node.removal,"Exp",median.exp.jump.SSD.testis.fish.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("E. Removing jump node contrasts for 5 teleosts"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.testis.fish.after.node.removal$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.03)


# The ortholog conjecture test after removing nodes supporting jump for testis in trusted SSD teleosts

testis.speciation.15spe.after.node.removal<-speciation.contrast(nodes.contrast.SSD.testis.jumps.15spe.removing.jump.node,"pic")
testis.duplication.15spe.after.node.removal<-duplication.contrast(nodes.contrast.SSD.testis.jumps.15spe.removing.jump.node,"pic")
Pval.exp.jump.SSD.testis.15spe.after.node.removal<-two.tailed.wilcox(testis.speciation.15spe.after.node.removal,testis.duplication.15spe.after.node.removal)

## Median data for trees supporting jump in average expression levels for SSD trees in teleosts
levels(nodes.contrast.SSD.testis.jumps.15spe.removing.jump.node$Event)<-c("Speciation", "SSD")
median.exp.jump.SSD.testis.15spe.after.node.removal<-NULL
median.exp.jump.SSD.testis.15spe.after.node.removal<-nodes.contrast.SSD.testis.jumps.15spe.removing.jump.node %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(Exp.abs),
            Exp.abs=median(Exp.abs),
            freq = n())
median.exp.jump.SSD.testis.15spe.after.node.removal$pic.round<-paste0("Median = ",round(median.exp.jump.SSD.testis.15spe.after.node.removal$Exp.abs,4), sep="")
median.exp.jump.SSD.testis.15spe.after.node.removal$count<- paste0(median.exp.jump.SSD.testis.15spe.after.node.removal$Event,"\n","(n = ",median.exp.jump.SSD.testis.15spe.after.node.removal$freq,")")


plotS19F<- boxplot.new(nodes.contrast.SSD.testis.jumps.15spe.removing.jump.node,Pval.exp.jump.SSD.testis.15spe.after.node.removal,"Exp",median.exp.jump.SSD.testis.15spe.after.node.removal)+
  coord_cartesian(ylim=c(0, 0.3)) +
  ggtitle(expression(bold(paste("F. Removing jump node contrasts for 15 species"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.exp.jump.SSD.testis.15spe.after.node.removal$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.26, ysize = 0.022)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.22, ysize = 0.027)

  
gridExtra::grid.arrange(plotS19A,plotS19B,plotS19C,plotS19D,plotS19E,plotS19F,ncol=2, nrow=3)

```



```{r Analyses_predupspe_postdupspe_calibrated_trees_our_method, eval=T, echo=F, message=FALSE, warning=FALSE, cache=T}

###Testing result for tau and plotting data
paired.pval.tau.our.predupspe.potdupspe<-paired.wilcox(effect.dup.measure$predup.spe.pic.x,effect.dup.measure$postspe.spenode.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.predupspe.postdupspe.our.method<-data.frame(pic=NA,label=NA)
Frame.tau.predupspe.postdupspe.our.method<-rbind(Frame.tau.predupspe.postdupspe.our.method,data.frame(pic=effect.dup.measure$predup.spe.pic.x,label="Predup_speciation"))
Frame.tau.predupspe.postdupspe.our.method<-rbind(Frame.tau.predupspe.postdupspe.our.method,data.frame(pic=effect.dup.measure$postspe.spenode.pic,label="Post_predup_speciation"))
Frame.tau.predupspe.postdupspe.our.method<-Frame.tau.predupspe.postdupspe.our.method[-1,]
Frame.tau.predupspe.postdupspe.our.method$Event <- factor(Frame.tau.predupspe.postdupspe.our.method$label, levels=c("Predup_speciation", "Post_predup_speciation"))

## For median data
median.data.predupspe.postdupspe.our.method<-NULL
median.data.predupspe.postdupspe.our.method<-Frame.tau.predupspe.postdupspe.our.method %>%   group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.predupspe.postdupspe.our.method$pic.round<-round(median.data.predupspe.postdupspe.our.method$Median,4)
median.data.predupspe.postdupspe.our.method$count<- paste0(median.data.predupspe.postdupspe.our.method$Event," \n(n = ",median.data.predupspe.postdupspe.our.method$freq,")")

###Testing result for exp levels and plotting data
paired.pval.exp.our.predupspe.potdupspe<-paired.wilcox(effect.dup.measure.exp$predup.spe.pic.x,effect.dup.measure.exp$postspe.spenode.pic)

##Getting dataframe to plot the result of exp data
Frame.exp.predupspe.postdupspe.our.method<-data.frame(pic=NA,label=NA)
Frame.exp.predupspe.postdupspe.our.method<-rbind(Frame.exp.predupspe.postdupspe.our.method,data.frame(pic=effect.dup.measure.exp$predup.spe.pic.x,label="Predup_speciation"))
Frame.exp.predupspe.postdupspe.our.method<-rbind(Frame.exp.predupspe.postdupspe.our.method,data.frame(pic=effect.dup.measure.exp$postspe.spenode.pic,label="Post_predup_speciation"))
Frame.exp.predupspe.postdupspe.our.method<-Frame.exp.predupspe.postdupspe.our.method[-1,]
Frame.exp.predupspe.postdupspe.our.method$Event <- factor(Frame.exp.predupspe.postdupspe.our.method$label, levels=c("Predup_speciation", "Post_predup_speciation"))

## For median data
median.data.exp.predupspe.postdupspe.our.method<-NULL
median.data.exp.predupspe.postdupspe.our.method<-Frame.exp.predupspe.postdupspe.our.method %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.exp.predupspe.postdupspe.our.method$pic.round<-round(median.data.exp.predupspe.postdupspe.our.method$Median,4)
median.data.exp.predupspe.postdupspe.our.method$count<- paste0(median.data.exp.predupspe.postdupspe.our.method$Event,"\n(n = ",median.data.exp.predupspe.postdupspe.our.method$freq,")")

```


```{r Analyses_predupspe_potdupspe_calibrated_SSDs_our_method, eval=T, echo=F, message=FALSE, warning=FALSE, cache=T}

###Testing result for tau in SSDs and plotting data
paired.pval.tau.predupspe.potdupspe.SSDs<-paired.wilcox(effect.dup.measure.SSD$predup.spe.pic.x,effect.dup.measure.SSD$postspe.spenode.pic)

##Getting dataframe to plot the result of tau data
Frame.tau.predupspe.postdupspe.SSDs.our.method<-data.frame(pic=NA,label=NA)
Frame.tau.predupspe.postdupspe.SSDs.our.method<-rbind(Frame.tau.predupspe.postdupspe.SSDs.our.method,data.frame(pic=effect.dup.measure.SSD$predup.spe.pic.x,label="Predup_speciation"))
Frame.tau.predupspe.postdupspe.SSDs.our.method<-rbind(Frame.tau.predupspe.postdupspe.SSDs.our.method,data.frame(pic=effect.dup.measure.SSD$postspe.spenode.pic,label="Post_predup_speciation"))
Frame.tau.predupspe.postdupspe.SSDs.our.method<-Frame.tau.predupspe.postdupspe.SSDs.our.method[-1,]
Frame.tau.predupspe.postdupspe.SSDs.our.method$Event <- factor(Frame.tau.predupspe.postdupspe.SSDs.our.method$label, levels=c("Predup_speciation", "Post_predup_speciation"))

## For median data
median.data.predupspe.postdupspe.SSDs.our.method<-NULL
median.data.predupspe.postdupspe.SSDs.our.method<-Frame.tau.predupspe.postdupspe.SSDs.our.method %>%   group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.predupspe.postdupspe.SSDs.our.method$pic.round<-round(median.data.predupspe.postdupspe.SSDs.our.method$Median,4)
median.data.predupspe.postdupspe.SSDs.our.method$count<- paste0(median.data.predupspe.postdupspe.SSDs.our.method$Event," \n(n = ",median.data.predupspe.postdupspe.SSDs.our.method$freq,")")

###Testing result for exp levels in SSDs and plotting data
paired.pval.exp.predupspe.potdupspe.SSDs<-paired.wilcox(effect.dup.measure.SSD.exp$predup.spe.pic.x,effect.dup.measure.SSD.exp$postspe.spenode.pic)

##Getting dataframe to plot the result of exp data
Frame.exp.predupspe.postdupspe.SSDs.our.method<-data.frame(pic=NA,label=NA)
Frame.exp.predupspe.postdupspe.SSDs.our.method<-rbind(Frame.exp.predupspe.postdupspe.SSDs.our.method,data.frame(pic=effect.dup.measure.SSD.exp$predup.spe.pic.x,label="Predup_speciation"))
Frame.exp.predupspe.postdupspe.SSDs.our.method<-rbind(Frame.exp.predupspe.postdupspe.SSDs.our.method,data.frame(pic=effect.dup.measure.SSD.exp$postspe.spenode.pic,label="Post_predup_speciation"))
Frame.exp.predupspe.postdupspe.SSDs.our.method<-Frame.exp.predupspe.postdupspe.SSDs.our.method[-1,]
Frame.exp.predupspe.postdupspe.SSDs.our.method$Event <- factor(Frame.exp.predupspe.postdupspe.SSDs.our.method$label, levels=c("Predup_speciation", "Post_predup_speciation"))

## For median data
median.data.exp.predupspe.postdupspe.SSDs.our.method<-NULL
median.data.exp.predupspe.postdupspe.SSDs.our.method<-Frame.exp.predupspe.postdupspe.SSDs.our.method %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            Median=median(pic),
            freq = n())
median.data.exp.predupspe.postdupspe.SSDs.our.method$pic.round<-round(median.data.exp.predupspe.postdupspe.SSDs.our.method$Median,4)
median.data.exp.predupspe.postdupspe.SSDs.our.method$count<- paste0(median.data.exp.predupspe.postdupspe.SSDs.our.method$Event,"\n(n = ",median.data.exp.predupspe.postdupspe.SSDs.our.method$freq,")")

```



```{r Fig_S20, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S20**: Positive selection and trait divergence analyses for contrast standardized trusted SSD trees in all 15 vertebrates. We only used phylogenies supporting an evolutionary trait jump model for these analyses. We used a strict posterior probability cutoff of $\\ge$ 0.7 to identify branches with rapid jumps in traits. exp: Gene expression, PICs: phylogenetic independent contrasts, LRT: likelihood ratio test.")}


## Median data 
levels(selectome.output.jump.tau$Event)<-c("Speciation","SSD")
median.SSD.tau.jump.ps<-NULL
median.SSD.tau.jump.ps<-selectome.output.jump.tau %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(lrt),
            lrt=as.numeric(format(median(lrt), digits= 3, scientific = TRUE)),
            freq = n())
median.SSD.tau.jump.ps$lrt.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.tau.jump.ps$lrt,8))), sep="")
median.SSD.tau.jump.ps$count<- paste0(median.SSD.tau.jump.ps$Event,"\n","(n = ",median.SSD.tau.jump.ps$freq,")")

plotS20A <- ggplot(selectome.output.jump.tau,aes(x=Event, y=lrt, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.tau.jump.ps, aes(label = lrt.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste(Delta, "lnL")))) +
  ylim(c(0,2))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.1, xmax = 1.8, ymin = 1.0, ymax =1.0, alpha=0.8,colour = "blue")+
 # annotate("text", x = 1.4, y = 1.2, label= pval.jump.lrt.tau, fontface = 4,parse=T)+
  annotate("text", x = 1.45, y = 1.2, label= as.character(paste0("italic(",pval.jump.lrt.tau,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 1.2, label= as.character(paste0("italic(",pval.jump.lrt.tau,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 1.2, label= as.character(paste0("italic(",pval.jump.lrt.tau,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("A. LR test for ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.tau.jump.ps$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 1.5, ysize = 0.15)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 1.2, ysize = 0.19)
  

## Median data 
levels(selectome.output.jump.avg.exp$Event)<-c("Speciation","SSD")
median.SSD.avgexp.jump.ps<-NULL
median.SSD.avgexp.jump.ps<-selectome.output.jump.avg.exp %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(lrt),
            lrt=as.numeric(format(median(lrt), digits= 3, scientific = TRUE)),
            freq = n())
median.SSD.avgexp.jump.ps$lrt.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.avgexp.jump.ps$lrt,8))), sep="")
median.SSD.avgexp.jump.ps$count<- paste0(median.SSD.avgexp.jump.ps$Event,"\n","(n = ",median.SSD.avgexp.jump.ps$freq,")")

plotS20B <- ggplot(selectome.output.jump.avg.exp,aes(x=Event, y=lrt, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.avgexp.jump.ps, aes(label = lrt.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste(Delta, "lnL")))) +
  ylim(c(0,10))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.1, xmax = 1.8, ymin = 5.0, ymax =5.0, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.5, y = 6, label= pval.jump.lrt.avg.exp, fontface = 4)+
  annotate("text", x = 1.45, y = 6, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp,")")), parse=T, size=4)+
  annotate("text", x = 1.45, y = 6, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp,")")), parse=T, size=4.02)+
  annotate("text", x = 1.45, y = 6, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("B. LR test for average exp levels"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.avgexp.jump.ps$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 7.5, ysize = 0.75)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 6, ysize = 0.9)
  

## Median data 
levels(selectome.output.tau.jump.trees.only.ps$Event)<-c("Speciation","SSD")
median.tau.SSD.jump.trees.only.ps<-NULL
median.tau.SSD.jump.trees.only.ps<-selectome.output.tau.jump.trees.only.ps %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(lrt),
            lrt=as.numeric(format(median(lrt), digits= 3, scientific = TRUE)),
            freq = n())
median.tau.SSD.jump.trees.only.ps$lrt.new<-paste0("Median = ",sub("e", " × 10",(round(median.tau.SSD.jump.trees.only.ps$lrt,8))), sep="")
median.tau.SSD.jump.trees.only.ps$count<- paste0(median.tau.SSD.jump.trees.only.ps$Event,"\n","(n = ",median.tau.SSD.jump.trees.only.ps$freq,")")

plotS20C <- ggplot(selectome.output.tau.jump.trees.only.ps,aes(x=Event, y=lrt, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.tau.SSD.jump.trees.only.ps, aes(label = lrt.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste(Delta, "lnL")))) +
  ylim(c(0,0.02))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.1, xmax = 1.8, ymin = 0.01, ymax =0.01, alpha=0.8,colour = "blue")+
 # annotate("text", x = 1.5, y = 0.012, label= pval.lrt.tau.otherthanjump, fontface = 4)+
  annotate("text", x = 1.5, y = 0.012, label= as.character(paste0("italic(",pval.lrt.tau.otherthanjump,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.012, label= as.character(paste0("italic(",pval.lrt.tau.otherthanjump,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.012, label= as.character(paste0("italic(",pval.lrt.tau.otherthanjump,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("C. Removing branches supporting jump in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.SSD.jump.trees.only.ps$count) +
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.015, ysize = 0.0015)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.012, ysize = 0.002)
  
## Median data 
levels(selectome.output.avgexp.jump.trees.only.ps$Event)<-c("Speciation","SSD")
median.SSD.avgexp.jump.trees.only.ps<-NULL
median.SSD.avgexp.jump.trees.only.ps<-selectome.output.avgexp.jump.trees.only.ps %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(lrt),
            lrt=median(lrt),
            freq = n())
median.SSD.avgexp.jump.trees.only.ps$lrt.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.avgexp.jump.trees.only.ps$lrt,8))), sep="")
median.SSD.avgexp.jump.trees.only.ps$count<- paste0(median.SSD.avgexp.jump.trees.only.ps$Event,"\n","(n = ",median.SSD.avgexp.jump.trees.only.ps$freq,")")

plotS20D <- ggplot(selectome.output.avgexp.jump.trees.only.ps,aes(x=Event, y=lrt, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.avgexp.jump.trees.only.ps, aes(label = lrt.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste(Delta, "lnL")))) +
  ylim(c(0,10))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.1, xmax = 1.8, ymin = 5.0, ymax =5.0, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.5, y = 6, label= pval.lrt.avg.exp.otherthanjump, fontface = 4)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.lrt.avg.exp.otherthanjump,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.lrt.avg.exp.otherthanjump,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.lrt.avg.exp.otherthanjump,")")), parse=T, size=4.04)+
  #scale_fill_manual(values=c("#00BFC4","#F8766D"))+
  ggtitle(expression(bold(paste("D. Removing branches supporting jump in exp"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.avgexp.jump.trees.only.ps$count) +
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 7.5, ysize = 0.8)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 6, ysize = 0.9)

## Median data 
levels(selectome.output.jump.tau1$Event)<-c("Speciation","SSD")
median.SSD.tau.jump.ps.pic<-NULL
median.SSD.tau.jump.ps.pic<-selectome.output.jump.tau1  %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic_Tau),
            pic_Tau=as.numeric(format(median(abs(pic_Tau)), digits= 3, scientific = FALSE)),
            freq = n())
median.SSD.tau.jump.ps.pic$pic.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.tau.jump.ps.pic$pic_Tau,8))), sep="")
median.SSD.tau.jump.ps.pic$count<- paste0(median.SSD.tau.jump.ps.pic$Event,"\n","(n = ",median.SSD.tau.jump.ps.pic$freq,")")

plotS20E <- ggplot(selectome.output.jump.tau1,aes(x=Event, y=pic_Tau, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.tau.jump.ps.pic, aes(label = pic.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab(NULL) +
  ylab(expression(bold(paste("PICs of ", tau)))) +
  ylim(c(0,0.03))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.9, ymin = 0.015, ymax =0.015, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.5, y = 0.018, label= pval.jump.lrt.tau.pic, fontface = 4)+
  annotate("text", x = 1.5, y = 0.018, label= as.character(paste0("italic(",pval.jump.lrt.tau.pic,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.018, label= as.character(paste0("italic(",pval.jump.lrt.tau.pic,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.018, label= as.character(paste0("italic(",pval.jump.lrt.tau.pic,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("E. PIC analyses for ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.tau.jump.ps.pic$count) +
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.023, ysize = 0.002)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.018, ysize = 0.003)

## Median data 
levels(selectome.output.jump.avg.exp1$Event)<-c("Speciation","SSD")
median.SSD.avgexp.jump.ps.pic<-NULL
median.SSD.avgexp.jump.ps.pic<-selectome.output.jump.avg.exp1 %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            pic_Exp=as.numeric(format(median(pic), digits= 3, scientific = TRUE)),
            freq = n())
median.SSD.avgexp.jump.ps.pic$pic.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.avgexp.jump.ps.pic$pic_Exp,8))), sep="")
median.SSD.avgexp.jump.ps.pic$count<- paste0(median.SSD.avgexp.jump.ps.pic$Event,"\n","(n = ",median.SSD.avgexp.jump.ps.pic$freq,")")

plotS20F <- ggplot(selectome.output.jump.avg.exp1,aes(x=Event, y=pic_Exp, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.avgexp.jump.ps.pic, aes(label = pic.new),fontface = 2,hjust=0.5,vjust =-1)+
  xlab( NULL ) +
  ylab(expression(bold(paste("PICs of exp levels")))) +
  ylim(c(0,0.2))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.9, ymin = 0.1, ymax =0.1, alpha=0.8,colour = "blue")+
 # annotate("text", x = 1.5, y = 0.12, label= pval.jump.lrt.avg.exp.pic, fontface = 4)+
  #scale_fill_manual(values=c("#00BFC4","#F8766D"))+
  annotate("text", x = 1.5, y = 0.12, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.pic,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.12, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.pic,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.12, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.pic,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("F. PIC analyses for exp levels"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.avgexp.jump.ps.pic$count)+
  add_phylopic(tetrapods_pic, alpha = 1,x = 2.3, y = 0.15, ysize = 0.015)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.12, ysize = 0.019) 

##Plotting results together
  grid.arrange(plotS20A,plotS20B,plotS20C,plotS20D,plotS20E,plotS20F,nrow=3,ncol=2) 
  
```



```{r Fig_S21, eval=T, dpi=500,fig.width=8, fig.height=8, fig.cap= paste0("**Fig S21**: Positive selection and trait divergence analyses for contrast standardized trusted SSD trees in 5 teleost fishes. We only used phylogenies supporting an evolutionary trait jump, and then considered teleosts-specific branches for these analyses. We used a strict posterior probability cutoff of $\\ge$ 0.7 to identify branches with rapid jumps in traits. exps: Gene expression levels, PICs: phylogenetic independent contrasts, LRT: likelihood ratio test. ")}


## Median data 
levels(selectome.output.jump.tau.fish$Event)<-c("Speciation","SSD")
median.SSD.tau.jump.ps.fish<-NULL
median.SSD.tau.jump.ps.fish<-selectome.output.jump.tau.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(lrt),
            lrt=as.numeric(format(median(lrt), digits= 3, scientific = TRUE)),
            freq = n())
median.SSD.tau.jump.ps.fish$lrt.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.tau.jump.ps.fish$lrt,8))), sep="")
median.SSD.tau.jump.ps.fish$count<- paste0(median.SSD.tau.jump.ps.fish$Event,"\n","(n = ",median.SSD.tau.jump.ps.fish$freq,")")

plotS21A <- ggplot(selectome.output.jump.tau.fish,aes(x=Event, y=lrt, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.tau.jump.ps.fish, aes(label = lrt.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste(Delta, "lnL")))) +
  ylim(c(0,2))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.9, ymin = 1.0, ymax =1.0, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.5, y = 1.2, label= pval.jump.lrt.tau.fish, fontface = 4)+
  annotate("text", x = 1.5, y = 1.2, label= as.character(paste0("italic(",pval.jump.lrt.tau.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 1.2, label= as.character(paste0("italic(",pval.jump.lrt.tau.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 1.2, label= as.character(paste0("italic(",pval.jump.lrt.tau.fish,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("A. LR test for ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.tau.jump.ps.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 1.5, ysize = 0.25)

## Median data 
levels(selectome.output.jump.avg.exp.fish$Event)<-c("Speciation","SSD")
median.SSD.avgexp.jump.ps.fish<-NULL
median.SSD.avgexp.jump.ps.fish<-selectome.output.jump.avg.exp.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(lrt),
            lrt=as.numeric(format(median(lrt), digits= 3, scientific = TRUE)),
            freq = n())
median.SSD.avgexp.jump.ps.fish$lrt.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.avgexp.jump.ps.fish$lrt,8))), sep="")
median.SSD.avgexp.jump.ps.fish$count<- paste0(median.SSD.avgexp.jump.ps.fish$Event,"\n","(n = ",median.SSD.avgexp.jump.ps.fish$freq,")")

plotS21B <- ggplot(selectome.output.jump.avg.exp.fish,aes(x=Event, y=lrt, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.avgexp.jump.ps.fish, aes(label = lrt.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste(Delta, "lnL")))) +
  ylim(c(0,10))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.9, ymin = 5.0, ymax =5.0, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.5, y = 6, label= pval.jump.lrt.avg.exp.fish, fontface = 4)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.fish,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("B. LR test for average exp levels"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.avgexp.jump.ps.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 7.5, ysize = 1.3) 
  

## Median data 
levels(selectome.output.tau.jump.trees.only.ps.fish$Event)<-c("Speciation","SSD")
median.tau.SSD.jump.trees.only.ps.fish<-NULL
median.tau.SSD.jump.trees.only.ps.fish<-selectome.output.tau.jump.trees.only.ps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(lrt),
            lrt=as.numeric(format(median(lrt), digits= 3, scientific = F)),
            freq = n())
median.tau.SSD.jump.trees.only.ps.fish$lrt.new<-paste0("Median = ",sub("e", " × 10",(round(median.tau.SSD.jump.trees.only.ps.fish$lrt,8))), sep="")
median.tau.SSD.jump.trees.only.ps.fish$count<- paste0(median.tau.SSD.jump.trees.only.ps.fish$Event,"\n","(n = ",median.tau.SSD.jump.trees.only.ps.fish$freq,")")

plotS21C <- ggplot(selectome.output.tau.jump.trees.only.ps.fish,aes(x=Event, y=lrt, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.tau.SSD.jump.trees.only.ps.fish, aes(label = lrt.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste(Delta, "lnL")))) +
  ylim(c(0,0.02))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.9, ymin = 0.01, ymax =0.01, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.5, y = 0.012, label= pval.lrt.tau.otherthanjump.fish, fontface = 4)+
  annotate("text", x = 1.5, y = 0.012, label= as.character(paste0("italic(",pval.lrt.tau.otherthanjump.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.012, label= as.character(paste0("italic(",pval.lrt.tau.otherthanjump.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.012, label= as.character(paste0("italic(",pval.lrt.tau.otherthanjump.fish,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("C. Removing branches supporting jumps in ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.tau.SSD.jump.trees.only.ps.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.015, ysize = 0.0027) 
  

## Median data 
levels(selectome.output.avgexp.jump.trees.only.ps.fish$Event)<-c("Speciation","SSD")
median.SSD.avgexp.jump.trees.only.ps.fish<-NULL
median.SSD.avgexp.jump.trees.only.ps.fish<-selectome.output.avgexp.jump.trees.only.ps.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(lrt),
            lrt=median(lrt),
            freq = n())
median.SSD.avgexp.jump.trees.only.ps.fish$lrt.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.avgexp.jump.trees.only.ps.fish$lrt,8))), sep="")
median.SSD.avgexp.jump.trees.only.ps.fish$count<- paste0(median.SSD.avgexp.jump.trees.only.ps.fish$Event,"\n","(n = ",median.SSD.avgexp.jump.trees.only.ps.fish$freq,")")

plotS21D <- ggplot(selectome.output.avgexp.jump.trees.only.ps.fish,aes(x=Event, y=lrt, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.avgexp.jump.trees.only.ps.fish, aes(label = lrt.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste(Delta, "lnL")))) +
  ylim(c(0,10))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.9, ymin = 5.0, ymax =5.0, alpha=0.8,colour = "blue")+
  #annotate("text", x = 1.5, y = 6, label= pval.lrt.avg.exp.otherthanjump.fish, fontface = 4)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.lrt.avg.exp.otherthanjump.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.lrt.avg.exp.otherthanjump.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 6, label= as.character(paste0("italic(",pval.lrt.avg.exp.otherthanjump.fish,")")), parse=T, size=4.04)+
  #scale_fill_manual(values=c("#00BFC4","#F8766D"))+
  ggtitle(expression(bold(paste("D. Removing branches supporting jump in exps"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.avgexp.jump.trees.only.ps.fish$count) +
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 7.5, ysize = 1.3)

## Median data 
levels(selectome.output.jump.tau.fish$Event)<-c("Speciation","SSD")
median.SSD.tau.jump.ps.pic.fish<-NULL
median.SSD.tau.jump.ps.pic.fish<-selectome.output.jump.tau.fish  %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            pic_Tau=as.numeric(format(median(pic_Tau), digits= 3, scientific = TRUE)),
            freq = n())
median.SSD.tau.jump.ps.pic.fish$pic.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.tau.jump.ps.pic.fish$pic_Tau,8))), sep="")
median.SSD.tau.jump.ps.pic.fish$count<- paste0(median.SSD.tau.jump.ps.pic.fish$Event,"\n","(n = ",median.SSD.tau.jump.ps.pic.fish$freq,")")

plotS21E <- ggplot(selectome.output.jump.tau.fish,aes(x=Event, y=pic_Tau, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.tau.jump.ps.pic.fish, aes(label = pic.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste("PICs of ", tau)))) +
  ylim(c(0,0.03))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.9, ymin = 0.015, ymax =0.015, alpha=0.8,colour = "blue")+
 # annotate("text", x = 1.5, y = 0.018, label= pval.jump.lrt.tau.pic.fish, fontface = 4)+
  annotate("text", x = 1.5, y = 0.018, label= as.character(paste0("italic(",pval.jump.lrt.tau.pic.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.018, label= as.character(paste0("italic(",pval.jump.lrt.tau.pic.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.018, label= as.character(paste0("italic(",pval.jump.lrt.tau.pic.fish,")")), parse=T, size=4.04)+
  ggtitle(expression(bold(paste("E. PIC analyses for ", tau))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.tau.jump.ps.pic.fish$count) +
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.023, ysize = 0.004)


## Median data 
levels(selectome.output.jump.avg.exp.fish$Event)<-c("Speciation","SSD")
median.SSD.avgexp.jump.ps.pic.fish<-NULL
median.SSD.avgexp.jump.ps.pic.fish<-selectome.output.jump.avg.exp.fish %>% 
  group_by(Event) %>% 
  summarise(Mean=mean(pic),
            pic=as.numeric(format(median(pic), digits= 3, scientific = TRUE)),
            freq = n())
median.SSD.avgexp.jump.ps.pic.fish$pic.new<-paste0("Median = ",sub("e", " × 10",(round(median.SSD.avgexp.jump.ps.pic.fish$pic,8))), sep="")
median.SSD.avgexp.jump.ps.pic.fish$count<- paste0(median.SSD.avgexp.jump.ps.pic.fish$Event,"\n","(n = ",median.SSD.avgexp.jump.ps.pic.fish$freq,")")

plotS21F <- ggplot(selectome.output.jump.avg.exp.fish,aes(x=Event, y=pic, fill=Event)) + 
  guides(colour = guide_legend(override.aes = list(shape = 16))) +
  geom_boxplot(width=0.5,position = dodge, outlier.colour=NA,notch = T) +
  geom_text(data = median.SSD.avgexp.jump.ps.pic.fish, aes(label = pic.new),fontface = 2,hjust=0.5,vjust =-1,size=3.5)+
  xlab( NULL ) +
  ylab(expression(bold(paste("PICs of exp levels")))) +
  ylim(c(0,0.2))+
  theme_classic()+
  theme(legend.title=element_blank(),legend.position="none") +
  theme(axis.text=element_text(size=10,face="bold")) +
  theme(legend.text=element_text(size=10,face="bold"))+
  annotate("rect", xmin = 1.2, xmax = 1.9, ymin = 0.1, ymax =0.1, alpha=0.8,colour = "blue")+
 # annotate("text", x = 1.5, y = 0.12, label= pval.jump.lrt.avg.exp.pic.fish, fontface = 4)+
  annotate("text", x = 1.5, y = 0.12, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.pic.fish,")")), parse=T, size=4)+
  annotate("text", x = 1.5, y = 0.12, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.pic.fish,")")), parse=T, size=4.02)+
  annotate("text", x = 1.5, y = 0.12, label= as.character(paste0("italic(",pval.jump.lrt.avg.exp.pic.fish,")")), parse=T, size=4.04)+
  #scale_fill_manual(values=c("#00BFC4","#F8766D"))+
  ggtitle(expression(bold(paste("F. PIC analyses for exp levels"))))+
  theme(plot.title = element_text(color="black", size=10, face="bold"))+
  scale_x_discrete(labels=median.SSD.avgexp.jump.ps.pic.fish$count)+
  add_phylopic(fish_pic, alpha = 1,x = 2.3, y = 0.15, ysize = 0.025)  

##Plotting results together
  grid.arrange(plotS21A,plotS21B,plotS21C,plotS21D,plotS21E,plotS21F,nrow=3,ncol=2) 
  
```


\pagebreak

### References  

